{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#default_exp reader.spectronaut_reader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import pandas as pd\n",
    "\n",
    "from alphadeep.reader.psm_reader import \\\n",
    "    psm_reader_provider\n",
    "\n",
    "from alphadeep.reader.maxquant_reader import \\\n",
    "    MaxQuantReader\n",
    "\n",
    "class SpectronautReader(MaxQuantReader):\n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "        self.mod_sep = '[]'\n",
    "        self.tsv_sep = '\\t'\n",
    "\n",
    "        self.column_mapping = {\n",
    "            'sequence': 'StrippedPeptide',\n",
    "            'charge': 'PrecursorCharge',\n",
    "            'RT': ['RT','iRT','Tr_recalibrated','RetentionTime'],\n",
    "            'CCS': 'CCS',\n",
    "            'mobility': ['Mobility','IonMobility'],\n",
    "            'proteins': 'Protein Name',\n",
    "            'uniprot_ids': 'UniProtIds',\n",
    "            'genes': 'Genes',\n",
    "        } \n",
    "        self.modseq_col = 'ModifiedPeptide'\n",
    "    \n",
    "    def _load_file(self, filename):\n",
    "        df = pd.read_csv(filename, sep=self.tsv_sep)\n",
    "        df.drop_duplicates([\n",
    "            'ReferenceRun',self.modseq_col, 'PrecursorCharge'\n",
    "        ], inplace=True)\n",
    "        df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "        for rt_col in self.column_mapping['RT']:\n",
    "            if rt_col not in df.columns: continue\n",
    "            min_rt = df[rt_col].min()\n",
    "            df[rt_col] = (\n",
    "                df[rt_col] - min_rt\n",
    "            )/(df[rt_col].max() - min_rt)\n",
    "            break\n",
    "        return df\n",
    "\n",
    "class OpenSwathReader(SpectronautReader):\n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "        self.mod_sep = '()'\n",
    "        self.underscore_for_ncterm = False\n",
    "        self.fixed_C=False\n",
    "\n",
    "        self.column_mapping = {\n",
    "            'sequence': 'PeptideSequence',\n",
    "            'charge': 'PrecursorCharge',\n",
    "            'RT': ['RT','iRT','Tr_recalibrated','RetentionTime'],\n",
    "            'CCS': 'CCS',\n",
    "            'mobility': ['Mobility','IonMobility'],\n",
    "            'proteins': 'Protein Name',\n",
    "            'uniprot_ids': 'UniProtID',\n",
    "            'genes': 'Genes',\n",
    "        }\n",
    "        self._modseq_columns = [\n",
    "            'ModifiedPeptide',\n",
    "            'ModifiedSequence',\n",
    "            'FullUniModPeptideName',\n",
    "        ]\n",
    "        self.modseq_col = 'FullUniModPeptideName'\n",
    "\n",
    "    def _find_modseq_column(self, df):\n",
    "        for modseq_col in self._modseq_columns:\n",
    "            if modseq_col in df.columns:\n",
    "                self.modseq_col = modseq_col\n",
    "                break\n",
    "    \n",
    "    def _load_file(self, filename):\n",
    "        df = pd.read_csv(filename, sep=self.tsv_sep)\n",
    "        self._find_modseq_column(df)\n",
    "        df.drop_duplicates([\n",
    "            self.modseq_col, 'PrecursorCharge'\n",
    "        ], inplace=True)\n",
    "        df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "        for rt_col in self.column_mapping['RT']:\n",
    "            if rt_col not in df.columns: continue\n",
    "            min_rt = df[rt_col].min()\n",
    "            df[rt_col] = (\n",
    "                df[rt_col] - min_rt\n",
    "            )/(df[rt_col].max() - min_rt)\n",
    "            break\n",
    "        return df\n",
    "\n",
    "psm_reader_provider.register_reader(\n",
    "    'spectronaut', SpectronautReader\n",
    ")\n",
    "psm_reader_provider.register_reader(\n",
    "    'openswath', OpenSwathReader\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>charge</th>\n",
       "      <th>RT</th>\n",
       "      <th>CCS</th>\n",
       "      <th>mobility</th>\n",
       "      <th>proteins</th>\n",
       "      <th>uniprot_ids</th>\n",
       "      <th>genes</th>\n",
       "      <th>nAA</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ALVATPGK</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.758</td>\n",
       "      <td>NUCL_HUMAN</td>\n",
       "      <td>P19338</td>\n",
       "      <td>NCL</td>\n",
       "      <td>8</td>\n",
       "      <td>Phospho@T</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TLTPPLR</td>\n",
       "      <td>2</td>\n",
       "      <td>0.760144</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.818</td>\n",
       "      <td>ZC3HD_HUMAN</td>\n",
       "      <td>Q5T200</td>\n",
       "      <td>ZC3H13</td>\n",
       "      <td>7</td>\n",
       "      <td>Phospho@T</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LFVSPPEGSSR</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.917</td>\n",
       "      <td>MAP7_HUMAN</td>\n",
       "      <td>Q14244;Q14244-6;Q14244-7</td>\n",
       "      <td>MAP7</td>\n",
       "      <td>11</td>\n",
       "      <td>Acetyl@Protein N-term;Phospho@S</td>\n",
       "      <td>0;4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sequence  charge        RT   CCS  mobility     proteins  \\\n",
       "0     ALVATPGK       2  0.000000  <NA>     0.758   NUCL_HUMAN   \n",
       "1      TLTPPLR       2  0.760144  <NA>     0.818  ZC3HD_HUMAN   \n",
       "2  LFVSPPEGSSR       2  1.000000  <NA>     0.917   MAP7_HUMAN   \n",
       "\n",
       "                uniprot_ids   genes  nAA                             mods  \\\n",
       "0                    P19338     NCL    8                        Phospho@T   \n",
       "1                    Q5T200  ZC3H13    7                        Phospho@T   \n",
       "2  Q14244;Q14244-6;Q14244-7    MAP7   11  Acetyl@Protein N-term;Phospho@S   \n",
       "\n",
       "  mod_sites  \n",
       "0         5  \n",
       "1         3  \n",
       "2       0;4  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from io import StringIO\n",
    "tsv = StringIO('''ReferenceRun\tPrecursorCharge\tWorkflow\tIntModifiedPeptide\tCV\tAllowForNormalization\tModifiedPeptide\tStrippedPeptide\tiRT\tIonMobility\tiRTSourceSpecific\tBGSInferenceId\tIsProteotypic\tIntLabeledPeptide\tLabeledPeptide\tPrecursorMz\tReferenceRunQvalue\tReferenceRunMS1Response\tFragmentLossType\tFragmentNumber\tFragmentType\tFragmentCharge\tFragmentMz\tRelativeIntensity\tExcludeFromAssay\tDatabase\tProteinGroups\tUniProtIds\tProtein Name\tProteinDescription\tOrganisms\tOrganismId\tGenes\tProtein Existence\tSequence Version\tFASTAName\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_100ug_test_S4-A1_1_25843\t2\t\t_ALVAT[+80]PGK_\t\tTrue\t_ALVAT[Phospho (STY)]PGK_\tALVATPGK\t-5.032703\t0.758\t-5.032703\tP19338\tFalse\t_ALVAT[+80]PGK_\t_ALVAT[Phospho (STY)]PGK_\t418.717511324722\t0\t10352\tnoloss\t3\ty\t1\t301.187031733932\t53.1991\tFalse\tsp\tP19338\tP19338\tNUCL_HUMAN\tNucleolin\tHomo sapiens\t\tNCL\t1\t3\tMCT_human_UP000005640_9606\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_100ug_test_S4-A1_1_25843\t2\t\t_ALVAT[+80]PGK_\t\tTrue\t_ALVAT[Phospho (STY)]PGK_\tALVATPGK\t-5.032703\t0.758\t-5.032703\tP19338\tFalse\t_ALVAT[+80]PGK_\t_ALVAT[Phospho (STY)]PGK_\t418.717511324722\t0\t10352\tH3PO4\t4\ty\t1\t384.224142529733\t26.31595\tFalse\tsp\tP19338\tP19338\tNUCL_HUMAN\tNucleolin\tHomo sapiens\t\tNCL\t1\t3\tMCT_human_UP000005640_9606\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_100ug_test_S4-A1_1_25843\t2\t\t_TLT[+80]PPLR_\t\tTrue\t_TLT[Phospho (STY)]PPLR_\tTLTPPLR\t27.71659\t0.818\t27.71659\tQ5T200\tFalse\t_TLT[+80]PPLR_\t_TLT[Phospho (STY)]PPLR_\t439.230785875227\t0.000138389150379226\t23117\tnoloss\t3\tb\t1\t396.153027901512\t6.3264\tFalse\tsp\tQ5T200\tQ5T200\tZC3HD_HUMAN\tZinc finger CCCH domain-containing protein 13\tHomo sapiens\t\tZC3H13\t1\t1\tMCT_human_UP000005640_9606\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_100ug_test_S4-A1_1_25843\t2\t\t_TLT[+80]PPLR_\t\tTrue\t_TLT[Phospho (STY)]PPLR_\tTLTPPLR\t27.71659\t0.818\t27.71659\tQ5T200\tFalse\t_TLT[+80]PPLR_\t_TLT[Phospho (STY)]PPLR_\t439.230785875227\t0.000138389150379226\t23117\tnoloss\t3\ty\t1\t385.255780000092\t29.70625\tFalse\tsp\tQ5T200\tQ5T200\tZC3HD_HUMAN\tZinc finger CCCH domain-containing protein 13\tHomo sapiens\t\tZC3H13\t1\t1\tMCT_human_UP000005640_9606\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_library25_S4-C1_1_25867\t2\t\t_LFVT[+80]PPEGSSR_\t\tTrue\t_[Acetyl (Protein N-term)]LFVS[Phospho (STY)]PPEGSSR_\tLFVSPPEGSSR\t38.05031\t0.917\t38.05031\tQ14244;Q14244-6;Q14244-7\tFalse\t_LFVT[+80]PPEGSSR_\t_LFVT[Phospho (STY)]PPEGSSR_\t635.297385373987\t0\t14164\tH3PO4\t4\tb\t1\t443.265279065723\t12.24525\tFalse\tsp\tQ14244;Q14244-6;Q14244-7\tQ14244;Q14244-6;Q14244-7\tMAP7_HUMAN\tEnsconsin;Isoform of Q14244, Isoform 6 of Ensconsin;Isoform of Q14244, Isoform 7 of Ensconsin\tHomo sapiens\t\tMAP7\t1;;\t1;;\tMCT_human_UP000005640_9606;MCT_human2_UP000005640_9606_additional;MCT_human2_UP000005640_9606_additional\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_library25_S4-C1_1_25867\t2\t\t_LFVT[+80]PPEGSSR_\t\tTrue\t_[Acetyl (Protein N-term)]LFVS[Phospho (STY)]PPEGSSR_\tLFVSPPEGSSR\t38.05031\t0.917\t38.05031\tQ14244;Q14244-6;Q14244-7\tFalse\t_LFVT[+80]PPEGSSR_\t_LFVT[Phospho (STY)]PPEGSSR_\t635.297385373987\t0\t14164\tnoloss\t6\ty\t1\t632.299829640042\t46.07855\tFalse\tsp\tQ14244;Q14244-6;Q14244-7\tQ14244;Q14244-6;Q14244-7\tMAP7_HUMAN\tEnsconsin;Isoform of Q14244, Isoform 6 of Ensconsin;Isoform of Q14244, Isoform 7 of Ensconsin\tHomo sapiens\t\tMAP7\t1;;\t1;;\tMCT_human_UP000005640_9606;MCT_human2_UP000005640_9606_additional;MCT_human2_UP000005640_9606_additional\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_library25_S4-C1_1_25867\t2\t\t_LFVT[+80]PPEGSSR_\t\tTrue\t_[Acetyl (Protein N-term)]LFVS[Phospho (STY)]PPEGSSR_\tLFVSPPEGSSR\t38.05031\t0.917\t38.05031\tQ14244;Q14244-6;Q14244-7\tFalse\t_LFVT[+80]PPEGSSR_\t_LFVT[Phospho (STY)]PPEGSSR_\t635.297385373987\t0\t14164\tnoloss\t7\ty\t1\t729.352593488892\t100\tFalse\tsp\tQ14244;Q14244-6;Q14244-7\tQ14244;Q14244-6;Q14244-7\tMAP7_HUMAN\tEnsconsin;Isoform of Q14244, Isoform 6 of Ensconsin;Isoform of Q14244, Isoform 7 of Ensconsin\tHomo sapiens\t\tMAP7\t1;;\t1;;\tMCT_human_UP000005640_9606;MCT_human2_UP000005640_9606_additional;MCT_human2_UP000005640_9606_additional\n",
    "''')\n",
    "\n",
    "\n",
    "reader = psm_reader_provider.get_reader('spectronaut')\n",
    "reader.load(tsv)\n",
    "reader.psm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "assert len(reader.psm_df) == 3\n",
    "assert reader.psm_df.mods.values[0] == 'Phospho@T'\n",
    "assert reader.psm_df.mod_sites.values[0] == '5'\n",
    "assert reader.psm_df.mods.values[1] == 'Phospho@T'\n",
    "assert reader.psm_df.mod_sites.values[1] == '3'\n",
    "assert reader.psm_df.mods.values[2] == 'Acetyl@Protein N-term;Phospho@S'\n",
    "assert reader.psm_df.mod_sites.values[2] == '0;4'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>charge</th>\n",
       "      <th>RT</th>\n",
       "      <th>CCS</th>\n",
       "      <th>mobility</th>\n",
       "      <th>proteins</th>\n",
       "      <th>uniprot_ids</th>\n",
       "      <th>genes</th>\n",
       "      <th>nAA</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAAAAASGAAIPPLIPPRR</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>23</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAAAAAASGAAIPPLIPPRR</td>\n",
       "      <td>4</td>\n",
       "      <td>0.004673</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>23</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAAAAASGAAIPPLIPPRRVITLYQCFSVSQR</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>36</td>\n",
       "      <td>Carbamidomethyl@C</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               sequence  charge        RT   CCS mobility  \\\n",
       "0               AAAAAAAAAASGAAIPPLIPPRR       3  0.000000  <NA>     <NA>   \n",
       "1               AAAAAAAAAASGAAIPPLIPPRR       4  0.004673  <NA>     <NA>   \n",
       "2  AAAAAAAAAASGAAIPPLIPPRRVITLYQCFSVSQR       5  1.000000  <NA>     <NA>   \n",
       "\n",
       "  proteins uniprot_ids genes  nAA               mods mod_sites  \n",
       "0     <NA>        <NA>  <NA>   23                               \n",
       "1     <NA>        <NA>  <NA>   23                               \n",
       "2     <NA>        <NA>  <NA>   36  Carbamidomethyl@C        30  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from io import StringIO\n",
    "tsv = StringIO('''PrecursorMz\tProductMz\tTr_recalibrated\ttransition_name\tCE\tLibraryIntensity\ttransition_group_id\tdecoy\tPeptideSequence\tProteinName\tAnnotation\tFullUniModPeptideName\tPrecursorCharge\tGroupLabel\tUniprotID\tFragmentType\tFragmentCharge\tFragmentSeriesNumber\n",
    "685.732240417\t886.020494795\t59.0\t255_AAAAAAAAAASGAAIPPLIPPRR_3\t-1\t5257.9\t13_AAAAAAAAAASGAAIPPLIPPRR_3\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\ty19^2/0.002\tAAAAAAAAAASGAAIPPLIPPRR\t3\tlight\t1/O14654\ty\t2\t19\n",
    "514.550999438\t473.303261576\t59.2\t268_AAAAAAAAAASGAAIPPLIPPRR_4\t-1\t10000.0\t14_AAAAAAAAAASGAAIPPLIPPRR_4\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\ty8^2/0.002\tAAAAAAAAAASGAAIPPLIPPRR\t4\tlight\t1/O14654\ty\t2\t8\n",
    "514.550999438\t629.39313922\t59.2\t276_AAAAAAAAAASGAAIPPLIPPRR_4\t-1\t5923.1\t14_AAAAAAAAAASGAAIPPLIPPRR_4\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\ty12^2/0.001\tAAAAAAAAAASGAAIPPLIPPRR\t4\tlight\t1/O14654\ty\t2\t12\n",
    "514.550999438\t672.909153425\t59.2\t279_AAAAAAAAAASGAAIPPLIPPRR_4\t-1\t5249.8\t14_AAAAAAAAAASGAAIPPLIPPRR_4\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\ty13^2/0.001\tAAAAAAAAAASGAAIPPLIPPRR\t4\tlight\t1/O14654\ty\t2\t13\n",
    "514.550999438\t356.19284545\t59.2\t262_AAAAAAAAAASGAAIPPLIPPRR_4\t-1\t5233.6\t14_AAAAAAAAAASGAAIPPLIPPRR_4\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\tb5/0.001,b10^2/0.001,m6:10/0.001\tAAAAAAAAAASGAAIPPLIPPRR\t4\tlight\t1/O14654\tb\t1\t5\n",
    "514.550999438\t498.26707303\t59.2\t269_AAAAAAAAAASGAAIPPLIPPRR_4\t-1\t4976.0\t14_AAAAAAAAAASGAAIPPLIPPRR_4\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\tb7/0.001,m4:10/0.001\tAAAAAAAAAASGAAIPPLIPPRR\t4\tlight\t1/O14654\tb\t1\t7\n",
    "514.550999438\t427.22995924\t59.2\t265_AAAAAAAAAASGAAIPPLIPPRR_4\t-1\t4859.4\t14_AAAAAAAAAASGAAIPPLIPPRR_4\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\tb6/0.002,m5:10/0.002\tAAAAAAAAAASGAAIPPLIPPRR\t4\tlight\t1/O14654\tb\t1\t6\n",
    "728.201724416\t356.19284545\t101.8\t292_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t-1\t10000.0\t15_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t0\tAAAAAAAAAASGAAIPPLIPPRRVITLYQCFSVSQR\t1/O14654\tb5/0.003,b10^2/0.003,m6:10/0.003\tAAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR\t5\tlight\t1/O14654\tb\t1\t5\n",
    "728.201724416\t576.310000482\t101.8\t297_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t-1\t7611.0\t15_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t0\tAAAAAAAAAASGAAIPPLIPPRRVITLYQCFSVSQR\t1/O14654\ty5/0.002\tAAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR\t5\tlight\t1/O14654\ty\t1\t5\n",
    "728.201724416\t427.22995924\t101.8\t293_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t-1\t6805.1\t15_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t0\tAAAAAAAAAASGAAIPPLIPPRRVITLYQCFSVSQR\t1/O14654\tb6/-0.002,m5:10/-0.002\tAAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR\t5\tlight\t1/O14654\tb\t1\t6\n",
    "728.201724416\t569.30418682\t101.8\t296_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t-1\t6312.7\t15_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t0\tAAAAAAAAAASGAAIPPLIPPRRVITLYQCFSVSQR\t1/O14654\tb8/0.009,m3:10/0.009\tAAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR\t5\tlight\t1/O14654\tb\t1\t8\n",
    "''')\n",
    "\n",
    "\n",
    "reader = psm_reader_provider.get_reader('openswath')\n",
    "reader.load(tsv)\n",
    "reader.psm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "8a3b27e141e49c996c9b863f8707e97aabd49c4a7e8445b9b783b34e4a21a9b2"
  },
  "kernelspec": {
   "display_name": "Python 3.8.3 64-bit ('base': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
