{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#default_exp reader.spectronaut_reader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from alphadeep.reader.psm_reader import \\\n",
    "    psm_reader_provider\n",
    "\n",
    "from alphadeep.reader.maxquant_reader import parse_mq, \\\n",
    "    MaxQuantReader\n",
    "\n",
    "class SpectronautReader(MaxQuantReader):\n",
    "    def __init__(self, fragment_reader=None):\n",
    "        super().__init__(None)\n",
    "        self.mod_sep = '[]'\n",
    "    \n",
    "    def _load_file(self, filename):\n",
    "        df = pd.read_csv(filename, sep='\\t')\n",
    "        df.drop_duplicates([\n",
    "            'ReferenceRun','ModifiedPeptide', 'PrecursorCharge'\n",
    "        ], inplace=True)\n",
    "        df.reset_index(drop=True, inplace=True)\n",
    "        psm_df = pd.DataFrame()\n",
    "        psm_df['sequence'] = df['StrippedPeptide']\n",
    "        df['nAA'] = df['StrippedPeptide'].str.len() # place holder for future\n",
    "        psm_df['nAA'] = df['nAA']\n",
    "        psm_df['mods'], psm_df['mod_sites'] = zip(\n",
    "            *df['ModifiedPeptide'].apply(\n",
    "                parse_mq, mod_sep=self.mod_sep\n",
    "            )\n",
    "        )\n",
    "        psm_df['charge'] = df['PrecursorCharge']\n",
    "        \n",
    "        psm_df['RT'] = df['iRT']\n",
    "        min_rt = psm_df.RT.min()\n",
    "        psm_df.RT = (\n",
    "            psm_df.RT - min_rt\n",
    "        )/(psm_df.RT.max() - min_rt)\n",
    "\n",
    "        if 'K0' in df.columns:\n",
    "            psm_df['mobility'] = 1/df['K0']\n",
    "        elif 'IonMobility' in df.columns:\n",
    "            psm_df['mobility'] = df['IonMobility']\n",
    "        else:\n",
    "            psm_df['mobility'] = pd.NA\n",
    "\n",
    "        if 'CCS' in df.columns:\n",
    "            psm_df['CCS'] = df['CCS']\n",
    "        else:\n",
    "            psm_df['CCS'] = pd.NA\n",
    "\n",
    "        psm_df['proteins'] = df['Protein Name']\n",
    "        if 'UniProtIds' in df.columns:\n",
    "            psm_df['uniprot_id'] = df['UniProtIds']\n",
    "        if 'Genes' in df.columns:\n",
    "            psm_df['genes'] = df['Genes']\n",
    "        else:\n",
    "            psm_df['genes'] = ''\n",
    "        self._psm_df = psm_df\n",
    "\n",
    "psm_reader_provider.register_reader(\n",
    "    'spectronaut', SpectronautReader\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>nAA</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>charge</th>\n",
       "      <th>RT</th>\n",
       "      <th>mobility</th>\n",
       "      <th>CCS</th>\n",
       "      <th>proteins</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ALVATPGK</td>\n",
       "      <td>8</td>\n",
       "      <td>Phospho@T</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.758</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NUCL_HUMAN</td>\n",
       "      <td>P19338</td>\n",
       "      <td>NCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TLTPPLR</td>\n",
       "      <td>7</td>\n",
       "      <td>Phospho@T</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.760144</td>\n",
       "      <td>0.818</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ZC3HD_HUMAN</td>\n",
       "      <td>Q5T200</td>\n",
       "      <td>ZC3H13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LFVSPPEGSSR</td>\n",
       "      <td>11</td>\n",
       "      <td>Acetyl@Protein N-term;Phospho@S</td>\n",
       "      <td>0;4</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.917</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MAP7_HUMAN</td>\n",
       "      <td>Q14244;Q14244-6;Q14244-7</td>\n",
       "      <td>MAP7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sequence  nAA                             mods mod_sites  charge  \\\n",
       "0     ALVATPGK    8                        Phospho@T         5       2   \n",
       "1      TLTPPLR    7                        Phospho@T         3       2   \n",
       "2  LFVSPPEGSSR   11  Acetyl@Protein N-term;Phospho@S       0;4       2   \n",
       "\n",
       "         RT  mobility   CCS     proteins                uniprot_id   genes  \n",
       "0  0.000000     0.758  <NA>   NUCL_HUMAN                    P19338     NCL  \n",
       "1  0.760144     0.818  <NA>  ZC3HD_HUMAN                    Q5T200  ZC3H13  \n",
       "2  1.000000     0.917  <NA>   MAP7_HUMAN  Q14244;Q14244-6;Q14244-7    MAP7  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from io import StringIO\n",
    "tsv = StringIO('''ReferenceRun\tPrecursorCharge\tWorkflow\tIntModifiedPeptide\tCV\tAllowForNormalization\tModifiedPeptide\tStrippedPeptide\tiRT\tIonMobility\tiRTSourceSpecific\tBGSInferenceId\tIsProteotypic\tIntLabeledPeptide\tLabeledPeptide\tPrecursorMz\tReferenceRunQvalue\tReferenceRunMS1Response\tFragmentLossType\tFragmentNumber\tFragmentType\tFragmentCharge\tFragmentMz\tRelativeIntensity\tExcludeFromAssay\tDatabase\tProteinGroups\tUniProtIds\tProtein Name\tProteinDescription\tOrganisms\tOrganismId\tGenes\tProtein Existence\tSequence Version\tFASTAName\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_100ug_test_S4-A1_1_25843\t2\t\t_ALVAT[+80]PGK_\t\tTrue\t_ALVAT[Phospho (STY)]PGK_\tALVATPGK\t-5.032703\t0.758\t-5.032703\tP19338\tFalse\t_ALVAT[+80]PGK_\t_ALVAT[Phospho (STY)]PGK_\t418.717511324722\t0\t10352\tnoloss\t3\ty\t1\t301.187031733932\t53.1991\tFalse\tsp\tP19338\tP19338\tNUCL_HUMAN\tNucleolin\tHomo sapiens\t\tNCL\t1\t3\tMCT_human_UP000005640_9606\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_100ug_test_S4-A1_1_25843\t2\t\t_ALVAT[+80]PGK_\t\tTrue\t_ALVAT[Phospho (STY)]PGK_\tALVATPGK\t-5.032703\t0.758\t-5.032703\tP19338\tFalse\t_ALVAT[+80]PGK_\t_ALVAT[Phospho (STY)]PGK_\t418.717511324722\t0\t10352\tH3PO4\t4\ty\t1\t384.224142529733\t26.31595\tFalse\tsp\tP19338\tP19338\tNUCL_HUMAN\tNucleolin\tHomo sapiens\t\tNCL\t1\t3\tMCT_human_UP000005640_9606\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_100ug_test_S4-A1_1_25843\t2\t\t_TLT[+80]PPLR_\t\tTrue\t_TLT[Phospho (STY)]PPLR_\tTLTPPLR\t27.71659\t0.818\t27.71659\tQ5T200\tFalse\t_TLT[+80]PPLR_\t_TLT[Phospho (STY)]PPLR_\t439.230785875227\t0.000138389150379226\t23117\tnoloss\t3\tb\t1\t396.153027901512\t6.3264\tFalse\tsp\tQ5T200\tQ5T200\tZC3HD_HUMAN\tZinc finger CCCH domain-containing protein 13\tHomo sapiens\t\tZC3H13\t1\t1\tMCT_human_UP000005640_9606\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_100ug_test_S4-A1_1_25843\t2\t\t_TLT[+80]PPLR_\t\tTrue\t_TLT[Phospho (STY)]PPLR_\tTLTPPLR\t27.71659\t0.818\t27.71659\tQ5T200\tFalse\t_TLT[+80]PPLR_\t_TLT[Phospho (STY)]PPLR_\t439.230785875227\t0.000138389150379226\t23117\tnoloss\t3\ty\t1\t385.255780000092\t29.70625\tFalse\tsp\tQ5T200\tQ5T200\tZC3HD_HUMAN\tZinc finger CCCH domain-containing protein 13\tHomo sapiens\t\tZC3H13\t1\t1\tMCT_human_UP000005640_9606\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_library25_S4-C1_1_25867\t2\t\t_LFVT[+80]PPEGSSR_\t\tTrue\t_[Acetyl (Protein N-term)]LFVS[Phospho (STY)]PPEGSSR_\tLFVSPPEGSSR\t38.05031\t0.917\t38.05031\tQ14244;Q14244-6;Q14244-7\tFalse\t_LFVT[+80]PPEGSSR_\t_LFVT[Phospho (STY)]PPEGSSR_\t635.297385373987\t0\t14164\tH3PO4\t4\tb\t1\t443.265279065723\t12.24525\tFalse\tsp\tQ14244;Q14244-6;Q14244-7\tQ14244;Q14244-6;Q14244-7\tMAP7_HUMAN\tEnsconsin;Isoform of Q14244, Isoform 6 of Ensconsin;Isoform of Q14244, Isoform 7 of Ensconsin\tHomo sapiens\t\tMAP7\t1;;\t1;;\tMCT_human_UP000005640_9606;MCT_human2_UP000005640_9606_additional;MCT_human2_UP000005640_9606_additional\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_library25_S4-C1_1_25867\t2\t\t_LFVT[+80]PPEGSSR_\t\tTrue\t_[Acetyl (Protein N-term)]LFVS[Phospho (STY)]PPEGSSR_\tLFVSPPEGSSR\t38.05031\t0.917\t38.05031\tQ14244;Q14244-6;Q14244-7\tFalse\t_LFVT[+80]PPEGSSR_\t_LFVT[Phospho (STY)]PPEGSSR_\t635.297385373987\t0\t14164\tnoloss\t6\ty\t1\t632.299829640042\t46.07855\tFalse\tsp\tQ14244;Q14244-6;Q14244-7\tQ14244;Q14244-6;Q14244-7\tMAP7_HUMAN\tEnsconsin;Isoform of Q14244, Isoform 6 of Ensconsin;Isoform of Q14244, Isoform 7 of Ensconsin\tHomo sapiens\t\tMAP7\t1;;\t1;;\tMCT_human_UP000005640_9606;MCT_human2_UP000005640_9606_additional;MCT_human2_UP000005640_9606_additional\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_library25_S4-C1_1_25867\t2\t\t_LFVT[+80]PPEGSSR_\t\tTrue\t_[Acetyl (Protein N-term)]LFVS[Phospho (STY)]PPEGSSR_\tLFVSPPEGSSR\t38.05031\t0.917\t38.05031\tQ14244;Q14244-6;Q14244-7\tFalse\t_LFVT[+80]PPEGSSR_\t_LFVT[Phospho (STY)]PPEGSSR_\t635.297385373987\t0\t14164\tnoloss\t7\ty\t1\t729.352593488892\t100\tFalse\tsp\tQ14244;Q14244-6;Q14244-7\tQ14244;Q14244-6;Q14244-7\tMAP7_HUMAN\tEnsconsin;Isoform of Q14244, Isoform 6 of Ensconsin;Isoform of Q14244, Isoform 7 of Ensconsin\tHomo sapiens\t\tMAP7\t1;;\t1;;\tMCT_human_UP000005640_9606;MCT_human2_UP000005640_9606_additional;MCT_human2_UP000005640_9606_additional\n",
    "''')\n",
    "\n",
    "\n",
    "reader = psm_reader_provider.get_reader('spectronaut')\n",
    "reader.load(tsv)\n",
    "reader.psm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "def test_reader():\n",
    "    assert len(reader.psm_df) == 3\n",
    "    assert reader.psm_df.mods.values[0] == 'Phospho@T'\n",
    "    assert reader.psm_df.mod_sites.values[0] == '5'\n",
    "    assert reader.psm_df.mods.values[1] == 'Phospho@T'\n",
    "    assert reader.psm_df.mod_sites.values[1] == '3'\n",
    "    assert reader.psm_df.mods.values[2] == 'Acetyl@Protein N-term;Phospho@S'\n",
    "    assert reader.psm_df.mod_sites.values[2] == '0;4'\n",
    "test_reader()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "8a3b27e141e49c996c9b863f8707e97aabd49c4a7e8445b9b783b34e4a21a9b2"
  },
  "kernelspec": {
   "display_name": "Python 3.8.3 64-bit ('base': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
