{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using peptdeep for MHC class I immunopeptidomics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook introduces how to generate spectral libraries for immunopeptidomics analysis from a list of protein sequences. This entails several steps:\n",
    "\n",
    "1. unspecific digestion of protein sequences\n",
    "2. selection of peptide sequences used for library prediction by peptdeep-hla predicition\n",
    "   2.1 using the pretrained model\n",
    "   2.2 using an improved model by including a transfer learning step\n",
    "3. spectral library prediction\n",
    "4. matching the peptides back to the proteins (this can be done before or after library prediction or seach)  \n",
    "\n",
    "\n",
    "\n",
    "Note that pydivsufsort package is not installed by peptdeep by default. Install by:\n",
    "```\n",
    "pip install \"peptdeep[development,hla]\"\n",
    "```\n",
    "\n",
    "Or install within jupyter notebook:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: Ignoring invalid distribution -lpharaw (c:\\users\\wahle\\.conda\\envs\\feng\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -lpharaw (c:\\users\\wahle\\.conda\\envs\\feng\\lib\\site-packages)\n"
     ]
    }
   ],
   "source": [
    "%pip install -q pydivsufsort"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Unspecific digestion in alphabase"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The unspecific digestion workflow uses the longest common prefix (LCP) algorithm, which is based on suffix array data structure, has been proven to be very efficient for unspecific digestion [https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-577]. Here we used `pydivsufsort`, a Python wrapper of a high-performance C library libdivsufsort [https://github.com/y-256/libdivsufsort], to facilitate LCP-based digestion.\n",
    "\n",
    "This means, the digestion is performed on a single sequence of strings and retrives both the peptide sequence as well as the start and stop indices of the peptide within the complete sequence. Therefore, unspecific digestion in alphabase involves two steps:\n",
    "\n",
    "1. concatenation of protein sequences into a single sequence\n",
    "2. unspecific digestion\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.1 Concatenate protein sequences into a single sequence\n",
    "\n",
    "The protein sequences are concatenated into a single sequence. The sequences are seperated by a sentinel character, in this case '$', so that no peptides across proteins are formed. Note that the first and last sentinel characters are crutial as well.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def concat_sequences_for_nonspecific_digestion(seq_list, sep=\"$\"):\n",
    "    return sep + sep.join(seq_list) + sep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'$MABCDEKFGHIJKLMNOPQRST$FGHIJKLMNOPQR$'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_seq_list = [\"MABCDEKFGHIJKLMNOPQRST\",\"FGHIJKLMNOPQR\"]\n",
    "cat_prot = concat_sequences_for_nonspecific_digestion(prot_seq_list, sep=\"$\")\n",
    "cat_prot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The same can be done directly from a fasta: \n",
    "@ Feng do you have an example fasta somwhere? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>full_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_org</th>\n",
       "      <th>description</th>\n",
       "      <th>sequence</th>\n",
       "      <th>nAA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>tr|A0A024R161|A0A024R161_HUMAN</th>\n",
       "      <td>A0A024R161</td>\n",
       "      <td>tr|A0A024R161|A0A024R161_HUMAN</td>\n",
       "      <td>DNAJC25-GNG10</td>\n",
       "      <td>A0A024R161_HUMAN</td>\n",
       "      <td>tr|A0A024R161|A0A024R161_HUMAN Guanine nucleot...</td>\n",
       "      <td>MGAPLLSPGWGAGAAGRRWWMLLAPLLPALLLVRPAGALVEGLYCG...</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tr|A0A024RAP8|A0A024RAP8_HUMAN</th>\n",
       "      <td>A0A024RAP8</td>\n",
       "      <td>tr|A0A024RAP8|A0A024RAP8_HUMAN</td>\n",
       "      <td>KLRC4-KLRK1</td>\n",
       "      <td>A0A024RAP8_HUMAN</td>\n",
       "      <td>tr|A0A024RAP8|A0A024RAP8_HUMAN HCG2009644, iso...</td>\n",
       "      <td>MGWIRGRRSRHSWEMSEFHNYNLDLKKSDFSTRWQKQRCPVVKSKC...</td>\n",
       "      <td>216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                protein_id                       full_name  \\\n",
       "tr|A0A024R161|A0A024R161_HUMAN  A0A024R161  tr|A0A024R161|A0A024R161_HUMAN   \n",
       "tr|A0A024RAP8|A0A024RAP8_HUMAN  A0A024RAP8  tr|A0A024RAP8|A0A024RAP8_HUMAN   \n",
       "\n",
       "                                    gene_name          gene_org  \\\n",
       "tr|A0A024R161|A0A024R161_HUMAN  DNAJC25-GNG10  A0A024R161_HUMAN   \n",
       "tr|A0A024RAP8|A0A024RAP8_HUMAN    KLRC4-KLRK1  A0A024RAP8_HUMAN   \n",
       "\n",
       "                                                                      description  \\\n",
       "tr|A0A024R161|A0A024R161_HUMAN  tr|A0A024R161|A0A024R161_HUMAN Guanine nucleot...   \n",
       "tr|A0A024RAP8|A0A024RAP8_HUMAN  tr|A0A024RAP8|A0A024RAP8_HUMAN HCG2009644, iso...   \n",
       "\n",
       "                                                                         sequence  \\\n",
       "tr|A0A024R161|A0A024R161_HUMAN  MGAPLLSPGWGAGAAGRRWWMLLAPLLPALLLVRPAGALVEGLYCG...   \n",
       "tr|A0A024RAP8|A0A024RAP8_HUMAN  MGWIRGRRSRHSWEMSEFHNYNLDLKKSDFSTRWQKQRCPVVKSKC...   \n",
       "\n",
       "                                nAA  \n",
       "tr|A0A024R161|A0A024R161_HUMAN  153  \n",
       "tr|A0A024RAP8|A0A024RAP8_HUMAN  216  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from peptdeep.hla.hla_utils import load_prot_df\n",
    "fasta_path = \"D:/Software/FASTA/Human/example.fasta\"\n",
    "fasta = load_prot_df(fasta_path)\n",
    "fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'$MGAPLLSPGWGAGAAGRRWWMLLAPLLPALLLVRPAGALVEGLYCGTRDCYEVLGVSRSAGKAEIARAYRQLARRYHPDRYRPQPGDEGPGRTPQSAEEAFLLVATAYETLKVSQAAAELQQYCMQNACKDALLVGVPAGSNPFREPRSCALL$MGWIRGRRSRHSWEMSEFHNYNLDLKKSDFSTRWQKQRCPVVKSKCRENASPFFFCCFIAVAMGIRFIIMVTIWSAVFLNSLFNQEVQIPLTESYCGPCPKNWICYKNNCYQFFDESKNWYESQASCMSQNASLLKVYSKEDQDLLKLVKSYHWMGLVHIPTNGSWQWEDGSILSPNLLTIIEMQKGDCALYASSFKGYIENCSTPNTYICMQRTV$'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from peptdeep.hla.hla_utils import cat_proteins\n",
    "cat_fasta = cat_proteins(fasta['sequence'])\n",
    "cat_fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.2 Unspecific digestion\n",
    "\n",
    "Use `alphabase.protein.lcp_digest.get_substring_indices` to get all non-redundant non-specific peptide sequences from the concatenated protein sequence. The digested peptide sequences are stored in a dataframe based on their start and stop indices in the concantenated protein sequence string. To save the RAM, the `peptdeep.hla` module works on start and stop indices instead of on peptide sequences directly. This will save about 8 times of the RAM for HLA-I peptides (length from 7 to 14, deomnstrated below). For a large protein sequence database, there will be millions of unspecific peptides, so working with strings is not feasible for a complete human fasta due to the requirements of extremely large RAM. (~ 70M unspecific sequences from the reviewed swissprot fasta require ~ 4-5 GB RAM already).\n",
    "\n",
    "Using the get_substring_indices function we extract the start and stop indices of all peptide sequences between 7 and 14 aa (min_len, max_len) from the concatenated protein sequences. All peptides sequences are unique, guranteed by the LCP algorithm."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_pos</th>\n",
       "      <th>stop_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2438</th>\n",
       "      <td>361</td>\n",
       "      <td>370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2439</th>\n",
       "      <td>361</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2440</th>\n",
       "      <td>362</td>\n",
       "      <td>370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2441</th>\n",
       "      <td>362</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2442</th>\n",
       "      <td>363</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2443 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      start_pos  stop_pos\n",
       "0             1         9\n",
       "1             1        10\n",
       "2             1        11\n",
       "3             1        12\n",
       "4             1        13\n",
       "...         ...       ...\n",
       "2438        361       370\n",
       "2439        361       371\n",
       "2440        362       370\n",
       "2441        362       371\n",
       "2442        363       371\n",
       "\n",
       "[2443 rows x 2 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from alphabase.protein.lcp_digest import get_substring_indices\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "start_idxes, stop_idxes = get_substring_indices(\n",
    "    cat_fasta, min_len=8, max_len=14, stop_char=\"$\"\n",
    ")\n",
    "digest_pos_df = pd.DataFrame({\n",
    "    \"start_pos\": start_idxes,\n",
    "    \"stop_pos\": stop_idxes,\n",
    "})\n",
    "digest_pos_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "RAM_use_idxes = sys.getsizeof(digest_pos_df)*1e-6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The unspecific peptide sequences can be localted by the `start_pos` and `stop_pos`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_pos</th>\n",
       "      <th>stop_pos</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>MGAPLLSP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>MGAPLLSPG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>MGAPLLSPGW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>MGAPLLSPGWG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>MGAPLLSPGWGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2438</th>\n",
       "      <td>361</td>\n",
       "      <td>370</td>\n",
       "      <td>NTYICMQRT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2439</th>\n",
       "      <td>361</td>\n",
       "      <td>371</td>\n",
       "      <td>NTYICMQRTV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2440</th>\n",
       "      <td>362</td>\n",
       "      <td>370</td>\n",
       "      <td>TYICMQRT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2441</th>\n",
       "      <td>362</td>\n",
       "      <td>371</td>\n",
       "      <td>TYICMQRTV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2442</th>\n",
       "      <td>363</td>\n",
       "      <td>371</td>\n",
       "      <td>YICMQRTV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2443 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      start_pos  stop_pos      sequence\n",
       "0             1         9      MGAPLLSP\n",
       "1             1        10     MGAPLLSPG\n",
       "2             1        11    MGAPLLSPGW\n",
       "3             1        12   MGAPLLSPGWG\n",
       "4             1        13  MGAPLLSPGWGA\n",
       "...         ...       ...           ...\n",
       "2438        361       370     NTYICMQRT\n",
       "2439        361       371    NTYICMQRTV\n",
       "2440        362       370      TYICMQRT\n",
       "2441        362       371     TYICMQRTV\n",
       "2442        363       371      YICMQRTV\n",
       "\n",
       "[2443 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "digest_pos_df[\"sequence\"] = digest_pos_df[\n",
    "    [\"start_pos\",\"stop_pos\"]\n",
    "].apply(lambda x: cat_fasta[slice(*x)], axis=1)\n",
    "digest_pos_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "RAM_use_seqs = sys.getsizeof(digest_pos_df[\"sequence\"])*1e-6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'seq RAM = 0.16621 Mb, idxes RAM = 0.01969, ratio = 8.44230'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f\"seq RAM = {RAM_use_seqs:.5f} Mb, idxes RAM = {RAM_use_idxes:.5f}, ratio = {RAM_use_seqs/RAM_use_idxes:.5f}\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Selection of peptide sequences used for library prediction\n",
    "The digest_prot_df contains all unspecifically digested peptide sequences between 7 and 14 aa generatable from the concatenated protein sequences. This list is reduced using a HLA1_Binding_Classifier from peptdeep.hla.hla_class1. Two different model architectures are available, an LSTM model (HLA_Class_I_LSTM) and a BERT model (HLA_Class_I_BERT). A pretrained model is only available for the LSTM model architecture.\n",
    "The HLA1_Binding_Classifer can be used with a pretrained model, tuned with existing peptide data or trained from scratch. Training of a new model should be considered carefully and will not be covered in this tutorial.\n",
    "   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Selection of peptide seqeuence candidates without transferlearning\n",
    "\n",
    "Selection of peptide sequences for library predicition using the pretrained model can be done in a few steps. First, the Classifier model needs to be initialized and the pretrained model is loaded. Next, we can use any kind of dataframe containing peptide sequences to predict how likely there are HLA peptides, the only requirement beeing that the column containing the peptides is called 'sequence'.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_pos</th>\n",
       "      <th>stop_pos</th>\n",
       "      <th>sequence</th>\n",
       "      <th>nAA</th>\n",
       "      <th>HLA_prob_pred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>MGAPLLSP</td>\n",
       "      <td>8</td>\n",
       "      <td>0.239477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>145</td>\n",
       "      <td>153</td>\n",
       "      <td>REPRSCAL</td>\n",
       "      <td>8</td>\n",
       "      <td>0.061692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>146</td>\n",
       "      <td>154</td>\n",
       "      <td>EPRSCALL</td>\n",
       "      <td>8</td>\n",
       "      <td>0.137313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>155</td>\n",
       "      <td>163</td>\n",
       "      <td>MGWIRGRR</td>\n",
       "      <td>8</td>\n",
       "      <td>0.056462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>156</td>\n",
       "      <td>164</td>\n",
       "      <td>GWIRGRRS</td>\n",
       "      <td>8</td>\n",
       "      <td>0.001298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2438</th>\n",
       "      <td>112</td>\n",
       "      <td>126</td>\n",
       "      <td>KVSQAAAELQQYCM</td>\n",
       "      <td>14</td>\n",
       "      <td>0.243115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2439</th>\n",
       "      <td>317</td>\n",
       "      <td>331</td>\n",
       "      <td>NGSWQWEDGSILSP</td>\n",
       "      <td>14</td>\n",
       "      <td>0.021114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2440</th>\n",
       "      <td>79</td>\n",
       "      <td>93</td>\n",
       "      <td>DRYRPQPGDEGPGR</td>\n",
       "      <td>14</td>\n",
       "      <td>0.060635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2441</th>\n",
       "      <td>113</td>\n",
       "      <td>127</td>\n",
       "      <td>VSQAAAELQQYCMQ</td>\n",
       "      <td>14</td>\n",
       "      <td>0.355900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2442</th>\n",
       "      <td>190</td>\n",
       "      <td>204</td>\n",
       "      <td>KQRCPVVKSKCREN</td>\n",
       "      <td>14</td>\n",
       "      <td>0.000362</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2443 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      start_pos  stop_pos        sequence  nAA  HLA_prob_pred\n",
       "0             1         9        MGAPLLSP    8       0.239477\n",
       "1           145       153        REPRSCAL    8       0.061692\n",
       "2           146       154        EPRSCALL    8       0.137313\n",
       "3           155       163        MGWIRGRR    8       0.056462\n",
       "4           156       164        GWIRGRRS    8       0.001298\n",
       "...         ...       ...             ...  ...            ...\n",
       "2438        112       126  KVSQAAAELQQYCM   14       0.243115\n",
       "2439        317       331  NGSWQWEDGSILSP   14       0.021114\n",
       "2440         79        93  DRYRPQPGDEGPGR   14       0.060635\n",
       "2441        113       127  VSQAAAELQQYCMQ   14       0.355900\n",
       "2442        190       204  KQRCPVVKSKCREN   14       0.000362\n",
       "\n",
       "[2443 rows x 5 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from peptdeep.hla.hla_class1 import HLA1_Binding_Classifier\n",
    "\n",
    "model = HLA1_Binding_Classifier()\n",
    "model.load_pretrained_hla_model()\n",
    "manual_prediction = model.predict(digest_pos_df)\n",
    "manual_prediction\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we can filter the list based on the HLA_prob_pred. The higher the probability, the more likely it is for the peptide sequence to be present in a immunopeptidomics sample. It is not recommended to use a cut-off below 0.7 as this inflates the spectral library. It is rather recommended to use more conservative cut-offs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_pos</th>\n",
       "      <th>stop_pos</th>\n",
       "      <th>sequence</th>\n",
       "      <th>nAA</th>\n",
       "      <th>HLA_prob_pred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>130</td>\n",
       "      <td>138</td>\n",
       "      <td>KDALLVGV</td>\n",
       "      <td>8</td>\n",
       "      <td>0.817415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>137</td>\n",
       "      <td>145</td>\n",
       "      <td>VPAGSNPF</td>\n",
       "      <td>8</td>\n",
       "      <td>0.751329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>170</td>\n",
       "      <td>178</td>\n",
       "      <td>SEFHNYNL</td>\n",
       "      <td>8</td>\n",
       "      <td>0.940019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>181</td>\n",
       "      <td>189</td>\n",
       "      <td>KSDFSTRW</td>\n",
       "      <td>8</td>\n",
       "      <td>0.895964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2318</th>\n",
       "      <td>95</td>\n",
       "      <td>109</td>\n",
       "      <td>QSAEEAFLLVATAY</td>\n",
       "      <td>14</td>\n",
       "      <td>0.969541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2378</th>\n",
       "      <td>329</td>\n",
       "      <td>343</td>\n",
       "      <td>SPNLLTIIEMQKGD</td>\n",
       "      <td>14</td>\n",
       "      <td>0.756001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2382</th>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>LLSPGWGAGAAGRR</td>\n",
       "      <td>14</td>\n",
       "      <td>0.733784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2408</th>\n",
       "      <td>110</td>\n",
       "      <td>124</td>\n",
       "      <td>TLKVSQAAAELQQY</td>\n",
       "      <td>14</td>\n",
       "      <td>0.891976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2419</th>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>LSPGWGAGAAGRRW</td>\n",
       "      <td>14</td>\n",
       "      <td>0.842583</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>148 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      start_pos  stop_pos        sequence  nAA  HLA_prob_pred\n",
       "17          168       176        EMSEFHNY    8       0.793702\n",
       "24          130       138        KDALLVGV    8       0.817415\n",
       "31          137       145        VPAGSNPF    8       0.751329\n",
       "37          170       178        SEFHNYNL    8       0.940019\n",
       "67          181       189        KSDFSTRW    8       0.895964\n",
       "...         ...       ...             ...  ...            ...\n",
       "2318         95       109  QSAEEAFLLVATAY   14       0.969541\n",
       "2378        329       343  SPNLLTIIEMQKGD   14       0.756001\n",
       "2382          5        19  LLSPGWGAGAAGRR   14       0.733784\n",
       "2408        110       124  TLKVSQAAAELQQY   14       0.891976\n",
       "2419          6        20  LSPGWGAGAAGRRW   14       0.842583\n",
       "\n",
       "[148 rows x 5 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "manual_prediction[manual_prediction['HLA_prob_pred'] > 0.7]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As described above, directly using the sequences for classification can be memory intense for large lists of sequences. Thereby, the manual concatenation, unspecific digestion, predicition and filtering is only suggested for small sets of proteins or integration of selected sequences (e.g mutations, nuORFs etc.). This can be circumvented by directly predicting and filtering from a fasta using model.predict_from_proteins(). This executes the concatenation, unspecific digestion, predicition and filtering automatically in batches. Thereby the whole process can be done more efficient and be performed without a specialized computation infrastructure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00,  1.20it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_pos</th>\n",
       "      <th>stop_pos</th>\n",
       "      <th>nAA</th>\n",
       "      <th>HLA_prob_pred</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>EMSEFHNY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>130</td>\n",
       "      <td>138</td>\n",
       "      <td>8</td>\n",
       "      <td>0.817415</td>\n",
       "      <td>KDALLVGV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>137</td>\n",
       "      <td>145</td>\n",
       "      <td>8</td>\n",
       "      <td>0.751329</td>\n",
       "      <td>VPAGSNPF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>170</td>\n",
       "      <td>178</td>\n",
       "      <td>8</td>\n",
       "      <td>0.940019</td>\n",
       "      <td>SEFHNYNL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>181</td>\n",
       "      <td>189</td>\n",
       "      <td>8</td>\n",
       "      <td>0.895964</td>\n",
       "      <td>KSDFSTRW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>95</td>\n",
       "      <td>109</td>\n",
       "      <td>14</td>\n",
       "      <td>0.969541</td>\n",
       "      <td>QSAEEAFLLVATAY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>329</td>\n",
       "      <td>343</td>\n",
       "      <td>14</td>\n",
       "      <td>0.756001</td>\n",
       "      <td>SPNLLTIIEMQKGD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>14</td>\n",
       "      <td>0.733784</td>\n",
       "      <td>LLSPGWGAGAAGRR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>110</td>\n",
       "      <td>124</td>\n",
       "      <td>14</td>\n",
       "      <td>0.891976</td>\n",
       "      <td>TLKVSQAAAELQQY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>14</td>\n",
       "      <td>0.842583</td>\n",
       "      <td>LSPGWGAGAAGRRW</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>148 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     start_pos  stop_pos  nAA  HLA_prob_pred        sequence\n",
       "0          168       176    8       0.793702        EMSEFHNY\n",
       "1          130       138    8       0.817415        KDALLVGV\n",
       "2          137       145    8       0.751329        VPAGSNPF\n",
       "3          170       178    8       0.940019        SEFHNYNL\n",
       "4          181       189    8       0.895964        KSDFSTRW\n",
       "..         ...       ...  ...            ...             ...\n",
       "143         95       109   14       0.969541  QSAEEAFLLVATAY\n",
       "144        329       343   14       0.756001  SPNLLTIIEMQKGD\n",
       "145          5        19   14       0.733784  LLSPGWGAGAAGRR\n",
       "146        110       124   14       0.891976  TLKVSQAAAELQQY\n",
       "147          6        20   14       0.842583  LSPGWGAGAAGRRW\n",
       "\n",
       "[148 rows x 5 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences = model.predict_from_proteins(fasta, prob_threshold=0.7)\n",
    "sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Selection of peptide seqeuence candidates with transferlearning\n",
    "\n",
    "To perform transferlearning we need a list of peptide sequences we expect to be present in our sample. These peptides can be retrived from several different sources like DDA or directDIA search results. It is recommended to use at the very least 1000 sequences for transferlearning. The more sequences available the better the transferlearning step works. The model performance can be assessed after transferlearning and should be assessed before predicition. \n",
    "\n",
    "First, the Classifier model needs to be initialized and the pretrained model is loaded. Next, a protein dataframe is added, in this example the previousely loaded fasta file. The protein dataframe is used by the Classifier internaly to draw negative training data during model training and testing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = HLA1_Binding_Classifier()\n",
    "model.load_pretrained_hla_model()\n",
    "model.load_proteins(fasta)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we load the peptide sequences wee use for transferlearning and split it into a training and testing dataset. This step is very important to assess the model performance after transferlearning. Here, we use the digest_pos_df generated above. As these are no immunopeptides, but a list of unspecifically digested proteins, the model performance will not improve, but the pronciples remain the same.  \n",
    "@ Feng should we include a example file so that the model is actually improved or just use this? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1954, 489)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_seq_df = digest_pos_df.sample(frac=0.2)\n",
    "train_seq_df = digest_pos_df.drop(index=test_seq_df.index)\n",
    "len(train_seq_df), len(test_seq_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we train the model using the training sequence dataframe. In this example we use 10 training epochs, in a real experiment more should be used. Good starting points are 40 epochs for a training dataset of around 10000 sequences or 100 epochs for a training dataset of around 1000 sequences. For a real experiment the warmup_epochs can be increased to 10.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-07-22 09:21:38> Training with fixed sequence length: 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\wahle\\.conda\\envs\\feng\\lib\\site-packages\\torch\\optim\\lr_scheduler.py:156: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.\n",
      "  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Training] Epoch=1, lr=2e-05, loss=1.415909733091082\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\wahle\\.conda\\envs\\feng\\lib\\site-packages\\torch\\optim\\lr_scheduler.py:156: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.\n",
      "  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Training] Epoch=2, lr=4e-05, loss=1.0947138496807642\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\wahle\\.conda\\envs\\feng\\lib\\site-packages\\torch\\optim\\lr_scheduler.py:156: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.\n",
      "  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Training] Epoch=3, lr=6e-05, loss=0.8823633790016174\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\wahle\\.conda\\envs\\feng\\lib\\site-packages\\torch\\optim\\lr_scheduler.py:156: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.\n",
      "  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Training] Epoch=4, lr=8e-05, loss=0.7819523641041347\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\wahle\\.conda\\envs\\feng\\lib\\site-packages\\torch\\optim\\lr_scheduler.py:156: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.\n",
      "  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Training] Epoch=5, lr=0.0001, loss=0.7255220583506993\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\wahle\\.conda\\envs\\feng\\lib\\site-packages\\torch\\optim\\lr_scheduler.py:156: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.\n",
      "  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Training] Epoch=6, lr=0.0001, loss=0.705090846334185\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\wahle\\.conda\\envs\\feng\\lib\\site-packages\\torch\\optim\\lr_scheduler.py:156: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.\n",
      "  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Training] Epoch=7, lr=9.045084971874738e-05, loss=0.7013667055538723\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\wahle\\.conda\\envs\\feng\\lib\\site-packages\\torch\\optim\\lr_scheduler.py:156: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.\n",
      "  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Training] Epoch=8, lr=6.545084971874738e-05, loss=0.6968921593257359\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\wahle\\.conda\\envs\\feng\\lib\\site-packages\\torch\\optim\\lr_scheduler.py:156: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.\n",
      "  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Training] Epoch=9, lr=3.4549150281252636e-05, loss=0.6968518495559692\n",
      "[Training] Epoch=10, lr=9.549150281252633e-06, loss=0.6932548114231655\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\wahle\\.conda\\envs\\feng\\lib\\site-packages\\torch\\optim\\lr_scheduler.py:156: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.\n",
      "  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "model.train(train_seq_df,\n",
    "            epoch=10, warmup_epoch=5, \n",
    "            verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can assess the model performance after transferlearning using the model.test() function on the training and testing data. This can also be done before transferlearning to assess how well the model fits the available data already. The test assesses the precision, recall and fals positive rate of the model at different probability cut offs. As a rule of thumb a false postitve rate above 7% (@FENG adjust in case lower/higher) is not recomendable because the peptide list gets disproportionally larger, leading to lower IDs during the search. In case of a high false postitive rate, the probability cut off at which the peptides are predicted should be increased.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HLA_prob_pred</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>false_positive</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.5</td>\n",
       "      <td>0.496400</td>\n",
       "      <td>0.599795</td>\n",
       "      <td>0.608495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.6</td>\n",
       "      <td>0.622951</td>\n",
       "      <td>0.019447</td>\n",
       "      <td>0.011771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   HLA_prob_pred  precision    recall  false_positive\n",
       "0            0.5   0.496400  0.599795        0.608495\n",
       "1            0.6   0.622951  0.019447        0.011771\n",
       "2            0.7        NaN  0.000000        0.000000\n",
       "3            0.8        NaN  0.000000        0.000000\n",
       "4            0.9        NaN  0.000000        0.000000"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.test(train_seq_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HLA_prob_pred</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>false_positive</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.5</td>\n",
       "      <td>0.480159</td>\n",
       "      <td>0.494888</td>\n",
       "      <td>0.535787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.6</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>0.012270</td>\n",
       "      <td>0.014315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   HLA_prob_pred  precision    recall  false_positive\n",
       "0            0.5   0.480159  0.494888        0.535787\n",
       "1            0.6   0.461538  0.012270        0.014315\n",
       "2            0.7        NaN  0.000000        0.000000\n",
       "3            0.8        NaN  0.000000        0.000000\n",
       "4            0.9        NaN  0.000000        0.000000"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.test(test_seq_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After transferlearning and testing the new model, peptides can be predicted as with the pretrained model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00,  1.20it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_pos</th>\n",
       "      <th>stop_pos</th>\n",
       "      <th>nAA</th>\n",
       "      <th>HLA_prob_pred</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>143</td>\n",
       "      <td>151</td>\n",
       "      <td>8</td>\n",
       "      <td>0.606630</td>\n",
       "      <td>PFREPRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>170</td>\n",
       "      <td>178</td>\n",
       "      <td>8</td>\n",
       "      <td>0.697908</td>\n",
       "      <td>SEFHNYNL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>62</td>\n",
       "      <td>70</td>\n",
       "      <td>8</td>\n",
       "      <td>0.602259</td>\n",
       "      <td>KAEIARAY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>87</td>\n",
       "      <td>95</td>\n",
       "      <td>8</td>\n",
       "      <td>0.611214</td>\n",
       "      <td>DEGPGRTP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>299</td>\n",
       "      <td>307</td>\n",
       "      <td>8</td>\n",
       "      <td>0.611188</td>\n",
       "      <td>LLKLVKSY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>346</td>\n",
       "      <td>354</td>\n",
       "      <td>8</td>\n",
       "      <td>0.620160</td>\n",
       "      <td>YASSFKGY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>344</td>\n",
       "      <td>352</td>\n",
       "      <td>8</td>\n",
       "      <td>0.601700</td>\n",
       "      <td>ALYASSFK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>223</td>\n",
       "      <td>231</td>\n",
       "      <td>8</td>\n",
       "      <td>0.605099</td>\n",
       "      <td>IMVTIWSA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>258</td>\n",
       "      <td>266</td>\n",
       "      <td>8</td>\n",
       "      <td>0.618778</td>\n",
       "      <td>ICYKNNCY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>363</td>\n",
       "      <td>371</td>\n",
       "      <td>8</td>\n",
       "      <td>0.602542</td>\n",
       "      <td>YICMQRTV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>17</td>\n",
       "      <td>25</td>\n",
       "      <td>8</td>\n",
       "      <td>0.605605</td>\n",
       "      <td>RRWWMLLA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>294</td>\n",
       "      <td>303</td>\n",
       "      <td>9</td>\n",
       "      <td>0.605901</td>\n",
       "      <td>KEDQDLLKL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>298</td>\n",
       "      <td>307</td>\n",
       "      <td>9</td>\n",
       "      <td>0.635218</td>\n",
       "      <td>DLLKLVKSY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>235</td>\n",
       "      <td>244</td>\n",
       "      <td>9</td>\n",
       "      <td>0.610781</td>\n",
       "      <td>SLFNQEVQI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>221</td>\n",
       "      <td>230</td>\n",
       "      <td>9</td>\n",
       "      <td>0.622273</td>\n",
       "      <td>FIIMVTIWS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>222</td>\n",
       "      <td>231</td>\n",
       "      <td>9</td>\n",
       "      <td>0.612193</td>\n",
       "      <td>IIMVTIWSA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>22</td>\n",
       "      <td>31</td>\n",
       "      <td>9</td>\n",
       "      <td>0.604106</td>\n",
       "      <td>LLAPLLPAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>257</td>\n",
       "      <td>266</td>\n",
       "      <td>9</td>\n",
       "      <td>0.620672</td>\n",
       "      <td>WICYKNNCY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>267</td>\n",
       "      <td>276</td>\n",
       "      <td>9</td>\n",
       "      <td>0.614074</td>\n",
       "      <td>FFDESKNWY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>17</td>\n",
       "      <td>26</td>\n",
       "      <td>9</td>\n",
       "      <td>0.604426</td>\n",
       "      <td>RRWWMLLAP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>327</td>\n",
       "      <td>336</td>\n",
       "      <td>9</td>\n",
       "      <td>0.614008</td>\n",
       "      <td>ILSPNLLTI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>255</td>\n",
       "      <td>265</td>\n",
       "      <td>10</td>\n",
       "      <td>0.601901</td>\n",
       "      <td>KNWICYKNNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>344</td>\n",
       "      <td>354</td>\n",
       "      <td>10</td>\n",
       "      <td>0.630664</td>\n",
       "      <td>ALYASSFKGY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>232</td>\n",
       "      <td>242</td>\n",
       "      <td>10</td>\n",
       "      <td>0.634032</td>\n",
       "      <td>FLNSLFNQEV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>221</td>\n",
       "      <td>231</td>\n",
       "      <td>10</td>\n",
       "      <td>0.632162</td>\n",
       "      <td>FIIMVTIWSA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>222</td>\n",
       "      <td>232</td>\n",
       "      <td>10</td>\n",
       "      <td>0.606705</td>\n",
       "      <td>IIMVTIWSAV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>353</td>\n",
       "      <td>363</td>\n",
       "      <td>10</td>\n",
       "      <td>0.611286</td>\n",
       "      <td>YIENCSTPNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>205</td>\n",
       "      <td>215</td>\n",
       "      <td>10</td>\n",
       "      <td>0.606018</td>\n",
       "      <td>SPFFFCCFIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>195</td>\n",
       "      <td>206</td>\n",
       "      <td>11</td>\n",
       "      <td>0.607188</td>\n",
       "      <td>VVKSKCRENAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>221</td>\n",
       "      <td>232</td>\n",
       "      <td>11</td>\n",
       "      <td>0.616940</td>\n",
       "      <td>FIIMVTIWSAV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>298</td>\n",
       "      <td>309</td>\n",
       "      <td>11</td>\n",
       "      <td>0.600725</td>\n",
       "      <td>DLLKLVKSYHW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>353</td>\n",
       "      <td>364</td>\n",
       "      <td>11</td>\n",
       "      <td>0.618278</td>\n",
       "      <td>YIENCSTPNTY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>42</td>\n",
       "      <td>54</td>\n",
       "      <td>12</td>\n",
       "      <td>0.606224</td>\n",
       "      <td>GLYCGTRDCYEV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>351</td>\n",
       "      <td>363</td>\n",
       "      <td>12</td>\n",
       "      <td>0.633097</td>\n",
       "      <td>KGYIENCSTPNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>200</td>\n",
       "      <td>212</td>\n",
       "      <td>12</td>\n",
       "      <td>0.608198</td>\n",
       "      <td>CRENASPFFFCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>224</td>\n",
       "      <td>236</td>\n",
       "      <td>12</td>\n",
       "      <td>0.606180</td>\n",
       "      <td>MVTIWSAVFLNS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>195</td>\n",
       "      <td>207</td>\n",
       "      <td>12</td>\n",
       "      <td>0.612207</td>\n",
       "      <td>VVKSKCRENASP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>166</td>\n",
       "      <td>179</td>\n",
       "      <td>13</td>\n",
       "      <td>0.628934</td>\n",
       "      <td>SWEMSEFHNYNLD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>351</td>\n",
       "      <td>364</td>\n",
       "      <td>13</td>\n",
       "      <td>0.604953</td>\n",
       "      <td>KGYIENCSTPNTY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>35</td>\n",
       "      <td>48</td>\n",
       "      <td>13</td>\n",
       "      <td>0.601324</td>\n",
       "      <td>PAGALVEGLYCGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>130</td>\n",
       "      <td>143</td>\n",
       "      <td>13</td>\n",
       "      <td>0.603384</td>\n",
       "      <td>KDALLVGVPAGSN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>333</td>\n",
       "      <td>347</td>\n",
       "      <td>14</td>\n",
       "      <td>0.601238</td>\n",
       "      <td>LTIIEMQKGDCALY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>185</td>\n",
       "      <td>199</td>\n",
       "      <td>14</td>\n",
       "      <td>0.610031</td>\n",
       "      <td>STRWQKQRCPVVKS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>117</td>\n",
       "      <td>131</td>\n",
       "      <td>14</td>\n",
       "      <td>0.600326</td>\n",
       "      <td>AAELQQYCMQNACK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    start_pos  stop_pos  nAA  HLA_prob_pred        sequence\n",
       "0         143       151    8       0.606630        PFREPRSC\n",
       "1         170       178    8       0.697908        SEFHNYNL\n",
       "2          62        70    8       0.602259        KAEIARAY\n",
       "3          87        95    8       0.611214        DEGPGRTP\n",
       "4         299       307    8       0.611188        LLKLVKSY\n",
       "5         346       354    8       0.620160        YASSFKGY\n",
       "6         344       352    8       0.601700        ALYASSFK\n",
       "7         223       231    8       0.605099        IMVTIWSA\n",
       "8         258       266    8       0.618778        ICYKNNCY\n",
       "9         363       371    8       0.602542        YICMQRTV\n",
       "10         17        25    8       0.605605        RRWWMLLA\n",
       "11        294       303    9       0.605901       KEDQDLLKL\n",
       "12        298       307    9       0.635218       DLLKLVKSY\n",
       "13        235       244    9       0.610781       SLFNQEVQI\n",
       "14        221       230    9       0.622273       FIIMVTIWS\n",
       "15        222       231    9       0.612193       IIMVTIWSA\n",
       "16         22        31    9       0.604106       LLAPLLPAL\n",
       "17        257       266    9       0.620672       WICYKNNCY\n",
       "18        267       276    9       0.614074       FFDESKNWY\n",
       "19         17        26    9       0.604426       RRWWMLLAP\n",
       "20        327       336    9       0.614008       ILSPNLLTI\n",
       "21        255       265   10       0.601901      KNWICYKNNC\n",
       "22        344       354   10       0.630664      ALYASSFKGY\n",
       "23        232       242   10       0.634032      FLNSLFNQEV\n",
       "24        221       231   10       0.632162      FIIMVTIWSA\n",
       "25        222       232   10       0.606705      IIMVTIWSAV\n",
       "26        353       363   10       0.611286      YIENCSTPNT\n",
       "27        205       215   10       0.606018      SPFFFCCFIA\n",
       "28        195       206   11       0.607188     VVKSKCRENAS\n",
       "29        221       232   11       0.616940     FIIMVTIWSAV\n",
       "30        298       309   11       0.600725     DLLKLVKSYHW\n",
       "31        353       364   11       0.618278     YIENCSTPNTY\n",
       "32         42        54   12       0.606224    GLYCGTRDCYEV\n",
       "33        351       363   12       0.633097    KGYIENCSTPNT\n",
       "34        200       212   12       0.608198    CRENASPFFFCC\n",
       "35        224       236   12       0.606180    MVTIWSAVFLNS\n",
       "36        195       207   12       0.612207    VVKSKCRENASP\n",
       "37        166       179   13       0.628934   SWEMSEFHNYNLD\n",
       "38        351       364   13       0.604953   KGYIENCSTPNTY\n",
       "39         35        48   13       0.601324   PAGALVEGLYCGT\n",
       "40        130       143   13       0.603384   KDALLVGVPAGSN\n",
       "41        333       347   14       0.601238  LTIIEMQKGDCALY\n",
       "42        185       199   14       0.610031  STRWQKQRCPVVKS\n",
       "43        117       131   14       0.600326  AAELQQYCMQNACK"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.predict_from_proteins(fasta, prob_threshold=0.6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spectral library prediciton"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now the spectral library for the filtered peptide list can be predicted using PredictSpecLibFasta. First, one needs to select the models for rt/ccs/ms2 prediction using the ModelManager. One can select from a set of pretrained models or load externally trained models. Here we load the 'HLA' model (at the moment this still loads the generic model, but in the futer this is supposed to be replaced by an HLA specfic internal model). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "from peptdeep.spec_lib.predict_lib import  ModelManager\n",
    "from peptdeep.protein.fasta import PredictSpecLibFasta\n",
    "\n",
    "model_mgr = ModelManager()\n",
    "model_mgr.load_installed_models(model_type='HLA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the next step, the PredictSpecLibFasta is initialized using the preloaded model. The presettings here are selected for the prediction of tryptic libraries so some parameters need to be adjusted, in particular precursor_charge_min, precursor_charge_max. By default Carbamidomethylation is set as a fixed modification (fix_mod) and Acetylation and Oxidation are set as variable modifications (var_mod). Those can be removed by adding an empty list as shown for the variable modifications. \n",
    "\n",
    "Of note, PredictSpecLibFasta can also be used to predict a library from a fasta file. Therfore one can also set the protease (default trypsin) and the minimum and maximum peptide length (7 to 35). Wee dont need to change those parameters here, as we wont make use of the digestion functions but rather provide a already digested sequence table. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "speclib = PredictSpecLibFasta(model_manager=model_mgr,\n",
    "                              precursor_charge_min=1,\n",
    "                              precursor_charge_max=3,\n",
    "                              fix_mods=[])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To reduce the size of the dataframe and predicted library we give each peptide sequence a unique protein identifier (number). This enables the use of search engines that rely on protein information (such as AlphaDIA) but one needs to keep in mind to remove filtering steps based on how many peptides per protein are identified during data analysis. Alternatively, proteins the peptide sequences could originate from can be infered using prot_infer (demonstrated below).   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_pos</th>\n",
       "      <th>stop_pos</th>\n",
       "      <th>nAA</th>\n",
       "      <th>HLA_prob_pred</th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_idxes</th>\n",
       "      <th>full_name</th>\n",
       "      <th>gene_org</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>is_prot_nterm</th>\n",
       "      <th>is_prot_cterm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>130</td>\n",
       "      <td>138</td>\n",
       "      <td>8</td>\n",
       "      <td>0.817415</td>\n",
       "      <td>KDALLVGV</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>137</td>\n",
       "      <td>145</td>\n",
       "      <td>8</td>\n",
       "      <td>0.751329</td>\n",
       "      <td>VPAGSNPF</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>170</td>\n",
       "      <td>178</td>\n",
       "      <td>8</td>\n",
       "      <td>0.940019</td>\n",
       "      <td>SEFHNYNL</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>181</td>\n",
       "      <td>189</td>\n",
       "      <td>8</td>\n",
       "      <td>0.895964</td>\n",
       "      <td>KSDFSTRW</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>95</td>\n",
       "      <td>109</td>\n",
       "      <td>14</td>\n",
       "      <td>0.969541</td>\n",
       "      <td>QSAEEAFLLVATAY</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>329</td>\n",
       "      <td>343</td>\n",
       "      <td>14</td>\n",
       "      <td>0.756001</td>\n",
       "      <td>SPNLLTIIEMQKGD</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>14</td>\n",
       "      <td>0.733784</td>\n",
       "      <td>LLSPGWGAGAAGRR</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>110</td>\n",
       "      <td>124</td>\n",
       "      <td>14</td>\n",
       "      <td>0.891976</td>\n",
       "      <td>TLKVSQAAAELQQY</td>\n",
       "      <td>146</td>\n",
       "      <td>146</td>\n",
       "      <td>146</td>\n",
       "      <td>146</td>\n",
       "      <td>146</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>14</td>\n",
       "      <td>0.842583</td>\n",
       "      <td>LSPGWGAGAAGRRW</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>148 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     start_pos  stop_pos  nAA  HLA_prob_pred        sequence protein_id  \\\n",
       "0          168       176    8       0.793702        EMSEFHNY          0   \n",
       "1          130       138    8       0.817415        KDALLVGV          1   \n",
       "2          137       145    8       0.751329        VPAGSNPF          2   \n",
       "3          170       178    8       0.940019        SEFHNYNL          3   \n",
       "4          181       189    8       0.895964        KSDFSTRW          4   \n",
       "..         ...       ...  ...            ...             ...        ...   \n",
       "143         95       109   14       0.969541  QSAEEAFLLVATAY        143   \n",
       "144        329       343   14       0.756001  SPNLLTIIEMQKGD        144   \n",
       "145          5        19   14       0.733784  LLSPGWGAGAAGRR        145   \n",
       "146        110       124   14       0.891976  TLKVSQAAAELQQY        146   \n",
       "147          6        20   14       0.842583  LSPGWGAGAAGRRW        147   \n",
       "\n",
       "    protein_idxes full_name gene_org gene_name  is_prot_nterm  is_prot_cterm  \n",
       "0               0         0        0         0          False          False  \n",
       "1               1         1        1         1          False          False  \n",
       "2               2         2        2         2          False          False  \n",
       "3               3         3        3         3          False          False  \n",
       "4               4         4        4         4          False          False  \n",
       "..            ...       ...      ...       ...            ...            ...  \n",
       "143           143       143      143       143          False          False  \n",
       "144           144       144      144       144          False          False  \n",
       "145           145       145      145       145          False          False  \n",
       "146           146       146      146       146          False          False  \n",
       "147           147       147      147       147          False          False  \n",
       "\n",
       "[148 rows x 12 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences['protein_id'] = [str(i) for i in range(len(sequences))]\n",
    "sequences['protein_idxes'] = sequences.protein_id.astype(\"U\")\n",
    "sequences['full_name'] = sequences['protein_id'] \n",
    "sequences['gene_org'] = sequences['protein_id'] \n",
    "sequences['gene_name'] = sequences['protein_id']\n",
    "sequences[\"is_prot_nterm\"] = False\n",
    "sequences[\"is_prot_cterm\"] = False\n",
    "sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The sequence dataframe contains all the relevant information to be passed to the protein_df and the precursor_df."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>nAA</th>\n",
       "      <th>full_name</th>\n",
       "      <th>gene_org</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KDALLVGV</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VPAGSNPF</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SEFHNYNL</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KSDFSTRW</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>QSAEEAFLLVATAY</td>\n",
       "      <td>143</td>\n",
       "      <td>14</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>SPNLLTIIEMQKGD</td>\n",
       "      <td>144</td>\n",
       "      <td>14</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>LLSPGWGAGAAGRR</td>\n",
       "      <td>145</td>\n",
       "      <td>14</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>TLKVSQAAAELQQY</td>\n",
       "      <td>146</td>\n",
       "      <td>14</td>\n",
       "      <td>146</td>\n",
       "      <td>146</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>LSPGWGAGAAGRRW</td>\n",
       "      <td>147</td>\n",
       "      <td>14</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>148 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sequence protein_id  nAA full_name gene_org gene_name\n",
       "0          EMSEFHNY          0    8         0        0         0\n",
       "1          KDALLVGV          1    8         1        1         1\n",
       "2          VPAGSNPF          2    8         2        2         2\n",
       "3          SEFHNYNL          3    8         3        3         3\n",
       "4          KSDFSTRW          4    8         4        4         4\n",
       "..              ...        ...  ...       ...      ...       ...\n",
       "143  QSAEEAFLLVATAY        143   14       143      143       143\n",
       "144  SPNLLTIIEMQKGD        144   14       144      144       144\n",
       "145  LLSPGWGAGAAGRR        145   14       145      145       145\n",
       "146  TLKVSQAAAELQQY        146   14       146      146       146\n",
       "147  LSPGWGAGAAGRRW        147   14       147      147       147\n",
       "\n",
       "[148 rows x 6 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "speclib.protein_df = sequences[[\"sequence\",\"protein_id\",\"nAA\", 'full_name', 'gene_org', 'gene_name']].copy()\n",
    "speclib.protein_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_idxes</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>stop_pos</th>\n",
       "      <th>nAA</th>\n",
       "      <th>HLA_prob_pred</th>\n",
       "      <th>is_prot_nterm</th>\n",
       "      <th>is_prot_cterm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KDALLVGV</td>\n",
       "      <td>1</td>\n",
       "      <td>130</td>\n",
       "      <td>138</td>\n",
       "      <td>8</td>\n",
       "      <td>0.817415</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VPAGSNPF</td>\n",
       "      <td>2</td>\n",
       "      <td>137</td>\n",
       "      <td>145</td>\n",
       "      <td>8</td>\n",
       "      <td>0.751329</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SEFHNYNL</td>\n",
       "      <td>3</td>\n",
       "      <td>170</td>\n",
       "      <td>178</td>\n",
       "      <td>8</td>\n",
       "      <td>0.940019</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KSDFSTRW</td>\n",
       "      <td>4</td>\n",
       "      <td>181</td>\n",
       "      <td>189</td>\n",
       "      <td>8</td>\n",
       "      <td>0.895964</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>QSAEEAFLLVATAY</td>\n",
       "      <td>143</td>\n",
       "      <td>95</td>\n",
       "      <td>109</td>\n",
       "      <td>14</td>\n",
       "      <td>0.969541</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>SPNLLTIIEMQKGD</td>\n",
       "      <td>144</td>\n",
       "      <td>329</td>\n",
       "      <td>343</td>\n",
       "      <td>14</td>\n",
       "      <td>0.756001</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>LLSPGWGAGAAGRR</td>\n",
       "      <td>145</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>14</td>\n",
       "      <td>0.733784</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>TLKVSQAAAELQQY</td>\n",
       "      <td>146</td>\n",
       "      <td>110</td>\n",
       "      <td>124</td>\n",
       "      <td>14</td>\n",
       "      <td>0.891976</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>LSPGWGAGAAGRRW</td>\n",
       "      <td>147</td>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>14</td>\n",
       "      <td>0.842583</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>148 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sequence protein_idxes  start_pos  stop_pos  nAA  HLA_prob_pred  \\\n",
       "0          EMSEFHNY             0        168       176    8       0.793702   \n",
       "1          KDALLVGV             1        130       138    8       0.817415   \n",
       "2          VPAGSNPF             2        137       145    8       0.751329   \n",
       "3          SEFHNYNL             3        170       178    8       0.940019   \n",
       "4          KSDFSTRW             4        181       189    8       0.895964   \n",
       "..              ...           ...        ...       ...  ...            ...   \n",
       "143  QSAEEAFLLVATAY           143         95       109   14       0.969541   \n",
       "144  SPNLLTIIEMQKGD           144        329       343   14       0.756001   \n",
       "145  LLSPGWGAGAAGRR           145          5        19   14       0.733784   \n",
       "146  TLKVSQAAAELQQY           146        110       124   14       0.891976   \n",
       "147  LSPGWGAGAAGRRW           147          6        20   14       0.842583   \n",
       "\n",
       "     is_prot_nterm  is_prot_cterm  \n",
       "0            False          False  \n",
       "1            False          False  \n",
       "2            False          False  \n",
       "3            False          False  \n",
       "4            False          False  \n",
       "..             ...            ...  \n",
       "143          False          False  \n",
       "144          False          False  \n",
       "145          False          False  \n",
       "146          False          False  \n",
       "147          False          False  \n",
       "\n",
       "[148 rows x 8 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "speclib.precursor_df = sequences[[\"sequence\",\"protein_idxes\",\"start_pos\",\"stop_pos\",\"nAA\",\"HLA_prob_pred\", 'is_prot_nterm', 'is_prot_cterm']].copy()\n",
    "speclib.precursor_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_idxes</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>stop_pos</th>\n",
       "      <th>nAA</th>\n",
       "      <th>HLA_prob_pred</th>\n",
       "      <th>is_prot_nterm</th>\n",
       "      <th>is_prot_cterm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KDALLVGV</td>\n",
       "      <td>1</td>\n",
       "      <td>130</td>\n",
       "      <td>138</td>\n",
       "      <td>8</td>\n",
       "      <td>0.817415</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VPAGSNPF</td>\n",
       "      <td>2</td>\n",
       "      <td>137</td>\n",
       "      <td>145</td>\n",
       "      <td>8</td>\n",
       "      <td>0.751329</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SEFHNYNL</td>\n",
       "      <td>3</td>\n",
       "      <td>170</td>\n",
       "      <td>178</td>\n",
       "      <td>8</td>\n",
       "      <td>0.940019</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KSDFSTRW</td>\n",
       "      <td>4</td>\n",
       "      <td>181</td>\n",
       "      <td>189</td>\n",
       "      <td>8</td>\n",
       "      <td>0.895964</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>QSAEEAFLLVATAY</td>\n",
       "      <td>143</td>\n",
       "      <td>95</td>\n",
       "      <td>109</td>\n",
       "      <td>14</td>\n",
       "      <td>0.969541</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>SPNLLTIIEMQKGD</td>\n",
       "      <td>144</td>\n",
       "      <td>329</td>\n",
       "      <td>343</td>\n",
       "      <td>14</td>\n",
       "      <td>0.756001</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>LLSPGWGAGAAGRR</td>\n",
       "      <td>145</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>14</td>\n",
       "      <td>0.733784</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>TLKVSQAAAELQQY</td>\n",
       "      <td>146</td>\n",
       "      <td>110</td>\n",
       "      <td>124</td>\n",
       "      <td>14</td>\n",
       "      <td>0.891976</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>LSPGWGAGAAGRRW</td>\n",
       "      <td>147</td>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>14</td>\n",
       "      <td>0.842583</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>148 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sequence protein_idxes  start_pos  stop_pos  nAA  HLA_prob_pred  \\\n",
       "0          EMSEFHNY             0        168       176    8       0.793702   \n",
       "1          KDALLVGV             1        130       138    8       0.817415   \n",
       "2          VPAGSNPF             2        137       145    8       0.751329   \n",
       "3          SEFHNYNL             3        170       178    8       0.940019   \n",
       "4          KSDFSTRW             4        181       189    8       0.895964   \n",
       "..              ...           ...        ...       ...  ...            ...   \n",
       "143  QSAEEAFLLVATAY           143         95       109   14       0.969541   \n",
       "144  SPNLLTIIEMQKGD           144        329       343   14       0.756001   \n",
       "145  LLSPGWGAGAAGRR           145          5        19   14       0.733784   \n",
       "146  TLKVSQAAAELQQY           146        110       124   14       0.891976   \n",
       "147  LSPGWGAGAAGRRW           147          6        20   14       0.842583   \n",
       "\n",
       "     is_prot_nterm  is_prot_cterm  \n",
       "0            False          False  \n",
       "1            False          False  \n",
       "2            False          False  \n",
       "3            False          False  \n",
       "4            False          False  \n",
       "..             ...            ...  \n",
       "143          False          False  \n",
       "144          False          False  \n",
       "145          False          False  \n",
       "146          False          False  \n",
       "147          False          False  \n",
       "\n",
       "[148 rows x 8 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "speclib.precursor_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, the modifications and charges can be added to the peptide dataframe using add_modifications and add_charge. This creates a unique entry for every combination of charge and modification for all the sequences in the precursor dataframe. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_idxes</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>stop_pos</th>\n",
       "      <th>nAA</th>\n",
       "      <th>HLA_prob_pred</th>\n",
       "      <th>is_prot_nterm</th>\n",
       "      <th>is_prot_cterm</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>charge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>TLKVSQAAAELQQY</td>\n",
       "      <td>146</td>\n",
       "      <td>110</td>\n",
       "      <td>124</td>\n",
       "      <td>14</td>\n",
       "      <td>0.891976</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>494</th>\n",
       "      <td>TLKVSQAAAELQQY</td>\n",
       "      <td>146</td>\n",
       "      <td>110</td>\n",
       "      <td>124</td>\n",
       "      <td>14</td>\n",
       "      <td>0.891976</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>LSPGWGAGAAGRRW</td>\n",
       "      <td>147</td>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>14</td>\n",
       "      <td>0.842583</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>LSPGWGAGAAGRRW</td>\n",
       "      <td>147</td>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>14</td>\n",
       "      <td>0.842583</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>LSPGWGAGAAGRRW</td>\n",
       "      <td>147</td>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>14</td>\n",
       "      <td>0.842583</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>498 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sequence protein_idxes  start_pos  stop_pos  nAA  HLA_prob_pred  \\\n",
       "0          EMSEFHNY             0        168       176    8       0.793702   \n",
       "1          EMSEFHNY             0        168       176    8       0.793702   \n",
       "2          EMSEFHNY             0        168       176    8       0.793702   \n",
       "3          EMSEFHNY             0        168       176    8       0.793702   \n",
       "4          EMSEFHNY             0        168       176    8       0.793702   \n",
       "..              ...           ...        ...       ...  ...            ...   \n",
       "493  TLKVSQAAAELQQY           146        110       124   14       0.891976   \n",
       "494  TLKVSQAAAELQQY           146        110       124   14       0.891976   \n",
       "495  LSPGWGAGAAGRRW           147          6        20   14       0.842583   \n",
       "496  LSPGWGAGAAGRRW           147          6        20   14       0.842583   \n",
       "497  LSPGWGAGAAGRRW           147          6        20   14       0.842583   \n",
       "\n",
       "     is_prot_nterm  is_prot_cterm         mods mod_sites  charge  \n",
       "0            False          False  Oxidation@M         2       1  \n",
       "1            False          False  Oxidation@M         2       2  \n",
       "2            False          False  Oxidation@M         2       3  \n",
       "3            False          False                              1  \n",
       "4            False          False                              2  \n",
       "..             ...            ...          ...       ...     ...  \n",
       "493          False          False                              2  \n",
       "494          False          False                              3  \n",
       "495          False          False                              1  \n",
       "496          False          False                              2  \n",
       "497          False          False                              3  \n",
       "\n",
       "[498 rows x 11 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "speclib.add_modifications()\n",
    "speclib.add_charge()\n",
    "speclib.precursor_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now ccs, rt and ms2 can be predicted for each entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-07-22 09:22:23> Predicting RT/IM/MS2 for 400 precursors ...\n",
      "2024-07-22 09:22:23> Predicting RT ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 7/7 [00:00<00:00, 69.31it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-07-22 09:22:23> Predicting mobility ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 7/7 [00:00<00:00, 72.89it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-07-22 09:22:23> Predicting MS2 ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 7/7 [00:00<00:00, 22.52it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-07-22 09:22:24> End predicting RT/IM/MS2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "speclib.predict_all()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "iRTs can be added using translate_rt_to_irt_pred. This is not neccessary for search engines like DIA-NN or AlphaDIA but required for Spectronaut."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predict RT for 11 iRT precursors.\n",
      "Linear regression of `rt_pred` to `irt`:\n",
      "   R_square         R       slope  intercept  test_num\n",
      "0   0.99007  0.995022  152.235621  -39.23216        11\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_idxes</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>stop_pos</th>\n",
       "      <th>nAA</th>\n",
       "      <th>HLA_prob_pred</th>\n",
       "      <th>is_prot_nterm</th>\n",
       "      <th>is_prot_cterm</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>...</th>\n",
       "      <th>precursor_mz</th>\n",
       "      <th>rt_pred</th>\n",
       "      <th>rt_norm_pred</th>\n",
       "      <th>ccs_pred</th>\n",
       "      <th>mobility_pred</th>\n",
       "      <th>nce</th>\n",
       "      <th>instrument</th>\n",
       "      <th>frag_start_idx</th>\n",
       "      <th>frag_stop_idx</th>\n",
       "      <th>irt_pred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1072.404037</td>\n",
       "      <td>0.189650</td>\n",
       "      <td>0.189650</td>\n",
       "      <td>254.195923</td>\n",
       "      <td>1.253140</td>\n",
       "      <td>30.0</td>\n",
       "      <td>Lumos</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>-10.360729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>536.705657</td>\n",
       "      <td>0.189650</td>\n",
       "      <td>0.189650</td>\n",
       "      <td>337.328583</td>\n",
       "      <td>0.831494</td>\n",
       "      <td>30.0</td>\n",
       "      <td>Lumos</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>-10.360729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>1056.409123</td>\n",
       "      <td>0.289261</td>\n",
       "      <td>0.289261</td>\n",
       "      <td>255.103760</td>\n",
       "      <td>1.257373</td>\n",
       "      <td>30.0</td>\n",
       "      <td>Lumos</td>\n",
       "      <td>14</td>\n",
       "      <td>21</td>\n",
       "      <td>4.803679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>528.708200</td>\n",
       "      <td>0.289261</td>\n",
       "      <td>0.289261</td>\n",
       "      <td>337.444641</td>\n",
       "      <td>0.831621</td>\n",
       "      <td>30.0</td>\n",
       "      <td>Lumos</td>\n",
       "      <td>21</td>\n",
       "      <td>28</td>\n",
       "      <td>4.803679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KDALLVGV</td>\n",
       "      <td>1</td>\n",
       "      <td>130</td>\n",
       "      <td>138</td>\n",
       "      <td>8</td>\n",
       "      <td>0.817415</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>814.503280</td>\n",
       "      <td>0.433791</td>\n",
       "      <td>0.433791</td>\n",
       "      <td>256.615234</td>\n",
       "      <td>1.260001</td>\n",
       "      <td>30.0</td>\n",
       "      <td>Lumos</td>\n",
       "      <td>28</td>\n",
       "      <td>35</td>\n",
       "      <td>26.806266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>TLKVSQAAAELQQY</td>\n",
       "      <td>146</td>\n",
       "      <td>110</td>\n",
       "      <td>124</td>\n",
       "      <td>14</td>\n",
       "      <td>0.891976</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>775.414662</td>\n",
       "      <td>0.489545</td>\n",
       "      <td>0.489545</td>\n",
       "      <td>429.360870</td>\n",
       "      <td>1.062514</td>\n",
       "      <td>30.0</td>\n",
       "      <td>Lumos</td>\n",
       "      <td>3810</td>\n",
       "      <td>3823</td>\n",
       "      <td>35.294021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>TLKVSQAAAELQQY</td>\n",
       "      <td>146</td>\n",
       "      <td>110</td>\n",
       "      <td>124</td>\n",
       "      <td>14</td>\n",
       "      <td>0.891976</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>517.278867</td>\n",
       "      <td>0.489545</td>\n",
       "      <td>0.489545</td>\n",
       "      <td>463.231110</td>\n",
       "      <td>0.764225</td>\n",
       "      <td>30.0</td>\n",
       "      <td>Lumos</td>\n",
       "      <td>3823</td>\n",
       "      <td>3836</td>\n",
       "      <td>35.294021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>LSPGWGAGAAGRRW</td>\n",
       "      <td>147</td>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>14</td>\n",
       "      <td>0.842583</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>1441.744742</td>\n",
       "      <td>0.377743</td>\n",
       "      <td>0.377743</td>\n",
       "      <td>289.200989</td>\n",
       "      <td>1.430378</td>\n",
       "      <td>30.0</td>\n",
       "      <td>Lumos</td>\n",
       "      <td>3836</td>\n",
       "      <td>3849</td>\n",
       "      <td>18.273781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>LSPGWGAGAAGRRW</td>\n",
       "      <td>147</td>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>14</td>\n",
       "      <td>0.842583</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>721.376009</td>\n",
       "      <td>0.377743</td>\n",
       "      <td>0.377743</td>\n",
       "      <td>404.633667</td>\n",
       "      <td>1.000659</td>\n",
       "      <td>30.0</td>\n",
       "      <td>Lumos</td>\n",
       "      <td>3849</td>\n",
       "      <td>3862</td>\n",
       "      <td>18.273781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>LSPGWGAGAAGRRW</td>\n",
       "      <td>147</td>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>14</td>\n",
       "      <td>0.842583</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>481.253098</td>\n",
       "      <td>0.377743</td>\n",
       "      <td>0.377743</td>\n",
       "      <td>476.655640</td>\n",
       "      <td>0.785851</td>\n",
       "      <td>30.0</td>\n",
       "      <td>Lumos</td>\n",
       "      <td>3862</td>\n",
       "      <td>3875</td>\n",
       "      <td>18.273781</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>400 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sequence protein_idxes  start_pos  stop_pos  nAA  HLA_prob_pred  \\\n",
       "0          EMSEFHNY             0        168       176    8       0.793702   \n",
       "1          EMSEFHNY             0        168       176    8       0.793702   \n",
       "2          EMSEFHNY             0        168       176    8       0.793702   \n",
       "3          EMSEFHNY             0        168       176    8       0.793702   \n",
       "4          KDALLVGV             1        130       138    8       0.817415   \n",
       "..              ...           ...        ...       ...  ...            ...   \n",
       "395  TLKVSQAAAELQQY           146        110       124   14       0.891976   \n",
       "396  TLKVSQAAAELQQY           146        110       124   14       0.891976   \n",
       "397  LSPGWGAGAAGRRW           147          6        20   14       0.842583   \n",
       "398  LSPGWGAGAAGRRW           147          6        20   14       0.842583   \n",
       "399  LSPGWGAGAAGRRW           147          6        20   14       0.842583   \n",
       "\n",
       "     is_prot_nterm  is_prot_cterm         mods mod_sites  ...  precursor_mz  \\\n",
       "0            False          False  Oxidation@M         2  ...   1072.404037   \n",
       "1            False          False  Oxidation@M         2  ...    536.705657   \n",
       "2            False          False                         ...   1056.409123   \n",
       "3            False          False                         ...    528.708200   \n",
       "4            False          False                         ...    814.503280   \n",
       "..             ...            ...          ...       ...  ...           ...   \n",
       "395          False          False                         ...    775.414662   \n",
       "396          False          False                         ...    517.278867   \n",
       "397          False          False                         ...   1441.744742   \n",
       "398          False          False                         ...    721.376009   \n",
       "399          False          False                         ...    481.253098   \n",
       "\n",
       "      rt_pred  rt_norm_pred    ccs_pred  mobility_pred   nce  instrument  \\\n",
       "0    0.189650      0.189650  254.195923       1.253140  30.0       Lumos   \n",
       "1    0.189650      0.189650  337.328583       0.831494  30.0       Lumos   \n",
       "2    0.289261      0.289261  255.103760       1.257373  30.0       Lumos   \n",
       "3    0.289261      0.289261  337.444641       0.831621  30.0       Lumos   \n",
       "4    0.433791      0.433791  256.615234       1.260001  30.0       Lumos   \n",
       "..        ...           ...         ...            ...   ...         ...   \n",
       "395  0.489545      0.489545  429.360870       1.062514  30.0       Lumos   \n",
       "396  0.489545      0.489545  463.231110       0.764225  30.0       Lumos   \n",
       "397  0.377743      0.377743  289.200989       1.430378  30.0       Lumos   \n",
       "398  0.377743      0.377743  404.633667       1.000659  30.0       Lumos   \n",
       "399  0.377743      0.377743  476.655640       0.785851  30.0       Lumos   \n",
       "\n",
       "    frag_start_idx  frag_stop_idx   irt_pred  \n",
       "0                0              7 -10.360729  \n",
       "1                7             14 -10.360729  \n",
       "2               14             21   4.803679  \n",
       "3               21             28   4.803679  \n",
       "4               28             35  26.806266  \n",
       "..             ...            ...        ...  \n",
       "395           3810           3823  35.294021  \n",
       "396           3823           3836  35.294021  \n",
       "397           3836           3849  18.273781  \n",
       "398           3849           3862  18.273781  \n",
       "399           3862           3875  18.273781  \n",
       "\n",
       "[400 rows x 21 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "speclib.translate_rt_to_irt_pred()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, the predicted library can be exported in an hdf format (AlphaDIA) or translated to a tsv. The tsv translation can be very time consuming. Before the spectral library can be translated, the gene and protein column need to be mapped from the protein_df into the precursor_df. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "hdf_path = \"D:\\Software\\FASTA\\Human\\speclib_example.hdf\"\n",
    "\n",
    "speclib.save_hdf(hdf_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:01<00:00,  1.51s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Translation finished, it will take several minutes to export the rest precursors to the tsv file...\n"
     ]
    }
   ],
   "source": [
    "from peptdeep.spec_lib.translate import translate_to_tsv\n",
    "speclib.append_protein_name()\n",
    "translate_to_tsv(speclib=speclib, \n",
    "                tsv =  \"D:\\Software\\FASTA\\Human\\speclib_example.tsv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4. matching peptides back to proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The peptide sequnces can be matched back to proteins using annotate_precursor_df, requiring a 'sequence' column and a protein_df like the previously loaded fasta file. This can be done with the sequence output of any search engine or before the library is generated. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 2/2 [00:00<?, ?it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_pos</th>\n",
       "      <th>stop_pos</th>\n",
       "      <th>nAA</th>\n",
       "      <th>HLA_prob_pred</th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_idxes</th>\n",
       "      <th>full_name</th>\n",
       "      <th>gene_org</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>is_prot_nterm</th>\n",
       "      <th>is_prot_cterm</th>\n",
       "      <th>genes</th>\n",
       "      <th>proteins</th>\n",
       "      <th>cardinality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>168</td>\n",
       "      <td>176</td>\n",
       "      <td>8</td>\n",
       "      <td>0.793702</td>\n",
       "      <td>EMSEFHNY</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>A0A024RAP8_HUMAN</td>\n",
       "      <td>A0A024RAP8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>130</td>\n",
       "      <td>138</td>\n",
       "      <td>8</td>\n",
       "      <td>0.817415</td>\n",
       "      <td>KDALLVGV</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>A0A024R161_HUMAN</td>\n",
       "      <td>A0A024R161</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>137</td>\n",
       "      <td>145</td>\n",
       "      <td>8</td>\n",
       "      <td>0.751329</td>\n",
       "      <td>VPAGSNPF</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>A0A024R161_HUMAN</td>\n",
       "      <td>A0A024R161</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>170</td>\n",
       "      <td>178</td>\n",
       "      <td>8</td>\n",
       "      <td>0.940019</td>\n",
       "      <td>SEFHNYNL</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>A0A024RAP8_HUMAN</td>\n",
       "      <td>A0A024RAP8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>181</td>\n",
       "      <td>189</td>\n",
       "      <td>8</td>\n",
       "      <td>0.895964</td>\n",
       "      <td>KSDFSTRW</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>A0A024RAP8_HUMAN</td>\n",
       "      <td>A0A024RAP8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>95</td>\n",
       "      <td>109</td>\n",
       "      <td>14</td>\n",
       "      <td>0.969541</td>\n",
       "      <td>QSAEEAFLLVATAY</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>A0A024R161_HUMAN</td>\n",
       "      <td>A0A024R161</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>329</td>\n",
       "      <td>343</td>\n",
       "      <td>14</td>\n",
       "      <td>0.756001</td>\n",
       "      <td>SPNLLTIIEMQKGD</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>A0A024RAP8_HUMAN</td>\n",
       "      <td>A0A024RAP8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>14</td>\n",
       "      <td>0.733784</td>\n",
       "      <td>LLSPGWGAGAAGRR</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>A0A024R161_HUMAN</td>\n",
       "      <td>A0A024R161</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>110</td>\n",
       "      <td>124</td>\n",
       "      <td>14</td>\n",
       "      <td>0.891976</td>\n",
       "      <td>TLKVSQAAAELQQY</td>\n",
       "      <td>146</td>\n",
       "      <td>146</td>\n",
       "      <td>146</td>\n",
       "      <td>146</td>\n",
       "      <td>146</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>A0A024R161_HUMAN</td>\n",
       "      <td>A0A024R161</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>14</td>\n",
       "      <td>0.842583</td>\n",
       "      <td>LSPGWGAGAAGRRW</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>147</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>A0A024R161_HUMAN</td>\n",
       "      <td>A0A024R161</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>148 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     start_pos  stop_pos  nAA  HLA_prob_pred        sequence protein_id  \\\n",
       "0          168       176    8       0.793702        EMSEFHNY          0   \n",
       "1          130       138    8       0.817415        KDALLVGV          1   \n",
       "2          137       145    8       0.751329        VPAGSNPF          2   \n",
       "3          170       178    8       0.940019        SEFHNYNL          3   \n",
       "4          181       189    8       0.895964        KSDFSTRW          4   \n",
       "..         ...       ...  ...            ...             ...        ...   \n",
       "143         95       109   14       0.969541  QSAEEAFLLVATAY        143   \n",
       "144        329       343   14       0.756001  SPNLLTIIEMQKGD        144   \n",
       "145          5        19   14       0.733784  LLSPGWGAGAAGRR        145   \n",
       "146        110       124   14       0.891976  TLKVSQAAAELQQY        146   \n",
       "147          6        20   14       0.842583  LSPGWGAGAAGRRW        147   \n",
       "\n",
       "    protein_idxes full_name gene_org gene_name  is_prot_nterm  is_prot_cterm  \\\n",
       "0               0         0        0         0          False          False   \n",
       "1               1         1        1         1          False          False   \n",
       "2               2         2        2         2          False          False   \n",
       "3               3         3        3         3          False          False   \n",
       "4               4         4        4         4          False          False   \n",
       "..            ...       ...      ...       ...            ...            ...   \n",
       "143           143       143      143       143          False          False   \n",
       "144           144       144      144       144          False          False   \n",
       "145           145       145      145       145          False          False   \n",
       "146           146       146      146       146          False          False   \n",
       "147           147       147      147       147          False          False   \n",
       "\n",
       "                genes    proteins  cardinality  \n",
       "0    A0A024RAP8_HUMAN  A0A024RAP8            1  \n",
       "1    A0A024R161_HUMAN  A0A024R161            1  \n",
       "2    A0A024R161_HUMAN  A0A024R161            1  \n",
       "3    A0A024RAP8_HUMAN  A0A024RAP8            1  \n",
       "4    A0A024RAP8_HUMAN  A0A024RAP8            1  \n",
       "..                ...         ...          ...  \n",
       "143  A0A024R161_HUMAN  A0A024R161            1  \n",
       "144  A0A024RAP8_HUMAN  A0A024RAP8            1  \n",
       "145  A0A024R161_HUMAN  A0A024R161            1  \n",
       "146  A0A024R161_HUMAN  A0A024R161            1  \n",
       "147  A0A024R161_HUMAN  A0A024R161            1  \n",
       "\n",
       "[148 rows x 15 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from alphabase.protein.fasta import annotate_precursor_df\n",
    "inferred_sequences = annotate_precursor_df(sequences, fasta)\n",
    "inferred_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
