{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HLA peptide prediction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Building a DL model (binary classifier) that distinguish whether a peptide is a presented HLA peptide.\n",
    "\n",
    "1. Use experimental HLA peptides to train a DL model to distinguish HLA peptides from non-HLA peptides.\n",
    "2. Apply the model to classify all peptides from a fasta file with non-specific digestion.\n",
    "3. Keep peptides with high predicted probabilities as HLA peptides."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read experimental HLA peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>nAA</th>\n",
       "      <th>HLA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>KQYDYDSSTIRKK</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>HVLEEVESLNRKY</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>KVEEAEPEEFVVEK</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>KTLPAMLGTGKL</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>GVEEEEEDGEMRE</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167737</th>\n",
       "      <td>KLKKQLKIY</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167738</th>\n",
       "      <td>KLIDPQTQV</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167741</th>\n",
       "      <td>MATSAPLRSL</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167742</th>\n",
       "      <td>GTIDEIQK</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167743</th>\n",
       "      <td>DIITKVANV</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>158923 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              sequence  nAA  HLA\n",
       "32       KQYDYDSSTIRKK   13    1\n",
       "54       HVLEEVESLNRKY   13    1\n",
       "74      KVEEAEPEEFVVEK   14    1\n",
       "76        KTLPAMLGTGKL   12    1\n",
       "86       GVEEEEEDGEMRE   13    1\n",
       "...                ...  ...  ...\n",
       "167737       KLKKQLKIY    9    1\n",
       "167738       KLIDPQTQV    9    1\n",
       "167741      MATSAPLRSL   10    1\n",
       "167742        GTIDEIQK    8    1\n",
       "167743       DIITKVANV    9    1\n",
       "\n",
       "[158923 rows x 3 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "pep_df = pd.read_table('/User/Feng/HLA/datasets/all_HLA_seqs.txt')\n",
    "save_as = '/User/Feng/HLA/model/pan_HLA_model.pth'\n",
    "pep_df['nAA'] = pep_df.sequence.str.len()\n",
    "pep_df['HLA'] = 1\n",
    "pep_df = pep_df[(pep_df.nAA>=8)&(pep_df.nAA<=14)]\n",
    "pep_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split data into training, validing, testing\n",
    "\n",
    "ratio = 7:1:2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_to_two_dfs(df, ratio=0.7):\n",
    "    train_df = df.sample(frac=ratio, replace=False)\n",
    "    test_df = df.drop(train_df.index)\n",
    "    return train_df.reset_index(drop=True), test_df.reset_index(drop=True)\n",
    "\n",
    "pos_train_df, pos_test_df = split_to_two_dfs(pep_df, 0.7)\n",
    "pos_valid_df, pos_test_df = split_to_two_dfs(pos_test_df, 0.33)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from alphabase.protein.fasta import load_all_proteins\n",
    "\n",
    "protein_dict = load_all_proteins(['/User/Feng/fasta/uniprot_human_reviewed_20210309.fasta'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>full_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "      <th>sequence</th>\n",
       "      <th>nAA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q9H9K5</td>\n",
       "      <td>sp|Q9H9K5|MER34_HUMAN</td>\n",
       "      <td>ERVMER34-1</td>\n",
       "      <td>sp|Q9H9K5|MER34_HUMAN Endogenous retroviral en...</td>\n",
       "      <td>MGSLSNYALLQLTLTAFLTILVQPQHLLAPVFRTLSILTNQSNCWL...</td>\n",
       "      <td>563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P04439</td>\n",
       "      <td>sp|P04439|HLAA_HUMAN</td>\n",
       "      <td>HLA-A</td>\n",
       "      <td>sp|P04439|HLAA_HUMAN HLA class I histocompatib...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P01911</td>\n",
       "      <td>sp|P01911|DRB1_HUMAN</td>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>sp|P01911|DRB1_HUMAN HLA class II histocompati...</td>\n",
       "      <td>MVCLKLPGGSCMTALTVTLMVLSSPLALSGDTRPRFLWQPKRECHF...</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P01889</td>\n",
       "      <td>sp|P01889|HLAB_HUMAN</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>sp|P01889|HLAB_HUMAN HLA class I histocompatib...</td>\n",
       "      <td>MLVMAPRTVLLLLSAALALTETWAGSHSMRYFYTSVSRPGRGEPRF...</td>\n",
       "      <td>362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P31689</td>\n",
       "      <td>sp|P31689|DNJA1_HUMAN</td>\n",
       "      <td>DNAJA1</td>\n",
       "      <td>sp|P31689|DNJA1_HUMAN DnaJ homolog subfamily A...</td>\n",
       "      <td>MVKETTYYDVLGVKPNATQEELKKAYRKLALKYHPDKNPNEGEKFK...</td>\n",
       "      <td>397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20381</th>\n",
       "      <td>Q8WVZ7</td>\n",
       "      <td>sp|Q8WVZ7|RN133_HUMAN</td>\n",
       "      <td>RNF133</td>\n",
       "      <td>sp|Q8WVZ7|RN133_HUMAN E3 ubiquitin-protein lig...</td>\n",
       "      <td>MHLLKVGTWRNNTASSWLMKFSVLWLVSQNCCRASVVWMAYMNISF...</td>\n",
       "      <td>376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20382</th>\n",
       "      <td>P05387</td>\n",
       "      <td>sp|P05387|RLA2_HUMAN</td>\n",
       "      <td>RPLP2</td>\n",
       "      <td>sp|P05387|RLA2_HUMAN 60S acidic ribosomal prot...</td>\n",
       "      <td>MRYVASYLLAALGGNSSPSAKDIKKILDSVGIEADDDRLNKVISEL...</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20383</th>\n",
       "      <td>P51991</td>\n",
       "      <td>sp|P51991|ROA3_HUMAN</td>\n",
       "      <td>HNRNPA3</td>\n",
       "      <td>sp|P51991|ROA3_HUMAN Heterogeneous nuclear rib...</td>\n",
       "      <td>MEVKPPPGRPQPDSGRRRRRRGEEGHDPKEPEQLRKLFIGGLSFET...</td>\n",
       "      <td>378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20384</th>\n",
       "      <td>Q9BZX4</td>\n",
       "      <td>sp|Q9BZX4|ROP1B_HUMAN</td>\n",
       "      <td>ROPN1B</td>\n",
       "      <td>sp|Q9BZX4|ROP1B_HUMAN Ropporin-1B OS=Homo sapi...</td>\n",
       "      <td>MAQTDKPTCIPPELPKMLKEFAKAAIRAQPQDLIQWGADYFEALSR...</td>\n",
       "      <td>212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20385</th>\n",
       "      <td>P34096</td>\n",
       "      <td>sp|P34096|RNAS4_HUMAN</td>\n",
       "      <td>RNASE4</td>\n",
       "      <td>sp|P34096|RNAS4_HUMAN Ribonuclease 4 OS=Homo s...</td>\n",
       "      <td>MALQRTHSLLLLLLLTLLGLGLVQPSYGQDGMYQRFLRQHVHPEET...</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20386 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      protein_id              full_name   gene_name  \\\n",
       "0         Q9H9K5  sp|Q9H9K5|MER34_HUMAN  ERVMER34-1   \n",
       "1         P04439   sp|P04439|HLAA_HUMAN       HLA-A   \n",
       "2         P01911   sp|P01911|DRB1_HUMAN    HLA-DRB1   \n",
       "3         P01889   sp|P01889|HLAB_HUMAN       HLA-B   \n",
       "4         P31689  sp|P31689|DNJA1_HUMAN      DNAJA1   \n",
       "...          ...                    ...         ...   \n",
       "20381     Q8WVZ7  sp|Q8WVZ7|RN133_HUMAN      RNF133   \n",
       "20382     P05387   sp|P05387|RLA2_HUMAN       RPLP2   \n",
       "20383     P51991   sp|P51991|ROA3_HUMAN     HNRNPA3   \n",
       "20384     Q9BZX4  sp|Q9BZX4|ROP1B_HUMAN      ROPN1B   \n",
       "20385     P34096  sp|P34096|RNAS4_HUMAN      RNASE4   \n",
       "\n",
       "                                             description  \\\n",
       "0      sp|Q9H9K5|MER34_HUMAN Endogenous retroviral en...   \n",
       "1      sp|P04439|HLAA_HUMAN HLA class I histocompatib...   \n",
       "2      sp|P01911|DRB1_HUMAN HLA class II histocompati...   \n",
       "3      sp|P01889|HLAB_HUMAN HLA class I histocompatib...   \n",
       "4      sp|P31689|DNJA1_HUMAN DnaJ homolog subfamily A...   \n",
       "...                                                  ...   \n",
       "20381  sp|Q8WVZ7|RN133_HUMAN E3 ubiquitin-protein lig...   \n",
       "20382  sp|P05387|RLA2_HUMAN 60S acidic ribosomal prot...   \n",
       "20383  sp|P51991|ROA3_HUMAN Heterogeneous nuclear rib...   \n",
       "20384  sp|Q9BZX4|ROP1B_HUMAN Ropporin-1B OS=Homo sapi...   \n",
       "20385  sp|P34096|RNAS4_HUMAN Ribonuclease 4 OS=Homo s...   \n",
       "\n",
       "                                                sequence  nAA  \n",
       "0      MGSLSNYALLQLTLTAFLTILVQPQHLLAPVFRTLSILTNQSNCWL...  563  \n",
       "1      MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...  365  \n",
       "2      MVCLKLPGGSCMTALTVTLMVLSSPLALSGDTRPRFLWQPKRECHF...  266  \n",
       "3      MLVMAPRTVLLLLSAALALTETWAGSHSMRYFYTSVSRPGRGEPRF...  362  \n",
       "4      MVKETTYYDVLGVKPNATQEELKKAYRKLALKYHPDKNPNEGEKFK...  397  \n",
       "...                                                  ...  ...  \n",
       "20381  MHLLKVGTWRNNTASSWLMKFSVLWLVSQNCCRASVVWMAYMNISF...  376  \n",
       "20382  MRYVASYLLAALGGNSSPSAKDIKKILDSVGIEADDDRLNKVISEL...  115  \n",
       "20383  MEVKPPPGRPQPDSGRRRRRRGEEGHDPKEPEQLRKLFIGGLSFET...  378  \n",
       "20384  MAQTDKPTCIPPELPKMLKEFAKAAIRAQPQDLIQWGADYFEALSR...  212  \n",
       "20385  MALQRTHSLLLLLLLTLLGLGLVQPSYGQDGMYQRFLRQHVHPEET...  147  \n",
       "\n",
       "[20386 rows x 6 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_df = pd.DataFrame().from_dict(protein_dict, orient='index')\n",
    "prot_df['nAA'] = prot_df.sequence.str.len()\n",
    "prot_df = prot_df[prot_df.nAA>=pep_df.nAA.max()]\n",
    "prot_df.reset_index(drop=True, inplace=True)\n",
    "prot_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate non-HLA peptides\n",
    "\n",
    "Generate random peptides from protein sequences as non-HLA peptides when training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['VLPDLLEEE', 'GDTSGYYQR', 'MSARVDAVA', 'WYSLVIRED', 'ALLLLAVAK',\n",
       "       'TWNHNNMFW', 'ALAAGMQVV', 'CRGHSTKAD', 'KDAPQGCKE', 'TILALARQG'],\n",
       "      dtype='<U9')"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "def get_subseq(x, pep_len):\n",
    "    sequence, prot_len = x\n",
    "    start = np.random.randint(0,prot_len-pep_len)\n",
    "    return sequence[start:start+pep_len]\n",
    "\n",
    "def get_random_sequences(prot_df, n, pep_len):\n",
    "    return prot_df.sample(n, replace=True)[['sequence','nAA']].apply(\n",
    "        get_subseq, pep_len=pep_len, axis=1\n",
    "    ).values.astype('U')\n",
    "get_random_sequences(prot_df, 10, 9)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Concatenate HLA and non-HLA peptide DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>nAA</th>\n",
       "      <th>HLA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AQSYAKRI</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EYRGEWKEGR</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DVAEGDLIEHF</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EVLISRVY</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RPTGGVGAVAL</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31461</th>\n",
       "      <td>SQEVCGLRWAPDGR</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31462</th>\n",
       "      <td>APSTPGRPERAARP</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31463</th>\n",
       "      <td>SLEDLRTKLTVIPQ</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31464</th>\n",
       "      <td>LMELGLSRMSNLSV</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31465</th>\n",
       "      <td>GATAVEDRLQDKVR</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31466 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             sequence  nAA  HLA\n",
       "0            AQSYAKRI    8    1\n",
       "1          EYRGEWKEGR   10    1\n",
       "2         DVAEGDLIEHF   11    1\n",
       "3            EVLISRVY    8    1\n",
       "4         RPTGGVGAVAL   11    1\n",
       "...               ...  ...  ...\n",
       "31461  SQEVCGLRWAPDGR   14    0\n",
       "31462  APSTPGRPERAARP   14    0\n",
       "31463  SLEDLRTKLTVIPQ   14    0\n",
       "31464  LMELGLSRMSNLSV   14    0\n",
       "31465  GATAVEDRLQDKVR   14    0\n",
       "\n",
       "[31466 rows x 3 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def concat_neg_df(pos_df, prot_df):\n",
    "    df_list = [pos_df]\n",
    "    for nAA, group_df in pos_df.groupby('nAA'):\n",
    "        rnd_seqs = get_random_sequences(\n",
    "            prot_df, \n",
    "            n=len(group_df),\n",
    "            pep_len = nAA\n",
    "        )\n",
    "        df_list.append(pd.DataFrame(\n",
    "            dict(sequence=rnd_seqs,nAA=nAA,HLA=0)\n",
    "        ))\n",
    "    return pd.concat(df_list).reset_index(drop=True)\n",
    "        \n",
    "valid_df = concat_neg_df(pos_valid_df, prot_df)\n",
    "valid_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Design the HLA classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import tqdm\n",
    "import numba\n",
    "\n",
    "import peptdeep.model.building_block as building_block\n",
    "from peptdeep.model.model_interface import ModelInterface\n",
    "from peptdeep.model.featurize import get_ascii_indices\n",
    "\n",
    "from alphabase.protein.lcp_digest import get_substring_indices\n",
    "from alphabase.protein.fasta import load_all_proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The HLA classifier based on LSTM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class HLA_LSTM(torch.nn.Module):\n",
    "    def __init__(self, *,\n",
    "        hidden_dim=256,\n",
    "        input_dim=128,\n",
    "        n_lstm_layers=4,\n",
    "        dropout=0.1,\n",
    "    ):\n",
    "        super().__init__()\n",
    "        self.dropout = torch.nn.Dropout(dropout)\n",
    "        \n",
    "        self.nn = torch.nn.Sequential(\n",
    "            torch.nn.Embedding(input_dim, hidden_dim//4),\n",
    "            building_block.SeqCNN(hidden_dim//4),\n",
    "            self.dropout,\n",
    "            building_block.SeqLSTM(hidden_dim, hidden_dim, rnn_layer=n_lstm_layers),\n",
    "            building_block.SeqAttentionSum(hidden_dim),\n",
    "            self.dropout,\n",
    "            torch.nn.Linear(hidden_dim,64),\n",
    "            torch.nn.GELU(),\n",
    "            torch.nn.Linear(64, 1),\n",
    "            torch.nn.Sigmoid()\n",
    "        )\n",
    "    def forward(self, x):\n",
    "        return self.nn(x).squeeze(-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The ModelInterface the interact with the HLA_LSTM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class HLA_ModelInterface(ModelInterface):\n",
    "    def __init__(self, \n",
    "        dropout=0.1,\n",
    "        model_class:torch.nn.Module=HLA_LSTM, #model defined above\n",
    "        device:str='gpu',\n",
    "        **kwargs,\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Class to predict retention times from precursor dataframes.\n",
    "        \"\"\"\n",
    "        super().__init__(device=device)\n",
    "        self.build(\n",
    "            model_class,\n",
    "            dropout=dropout,\n",
    "            **kwargs\n",
    "        )\n",
    "        self.loss_func = torch.nn.BCELoss()\n",
    "        self.target_column_to_predict = 'HLA_prob_pred'\n",
    "\n",
    "    def _prepare_predict_data_df(self,\n",
    "        precursor_df:pd.DataFrame,\n",
    "    ):\n",
    "        self.__training = False\n",
    "        precursor_df[self.target_column_to_predict] = 0.\n",
    "        self.predict_df = precursor_df\n",
    "        \n",
    "    def _prepare_train_data_df(self, precursor_df, prot_df, **kwargs):\n",
    "        self.__training = True\n",
    "        self.__n_neg_per_pos = 1\n",
    "        self.prot_df = prot_df\n",
    "\n",
    "    def _get_features_from_batch_df(self, \n",
    "        batch_df: pd.DataFrame,\n",
    "        **kwargs,\n",
    "    ):\n",
    "        aa_indices = self._as_tensor(\n",
    "            get_ascii_indices(\n",
    "                batch_df['sequence'].values.astype('U')\n",
    "            ), dtype=torch.long\n",
    "        )\n",
    "        \n",
    "        if self.__training:\n",
    "            rnd_seqs = get_random_sequences(\n",
    "                self.prot_df, \n",
    "                n=int(len(batch_df)*self.__n_neg_per_pos),\n",
    "                pep_len = batch_df.nAA.values[0]\n",
    "            )\n",
    "            aa_indices = torch.cat(\n",
    "                [aa_indices, \n",
    "                 self._as_tensor(\n",
    "                     get_ascii_indices(rnd_seqs), dtype=torch.long\n",
    "                 )\n",
    "                ], axis=0\n",
    "            )\n",
    "\n",
    "        return aa_indices\n",
    "\n",
    "    def _get_targets_from_batch_df(self, \n",
    "        batch_df: pd.DataFrame,\n",
    "        **kwargs\n",
    "    ) -> torch.Tensor:\n",
    "        x = torch.zeros(\n",
    "            len(batch_df)+(int(len(batch_df)*self.__n_neg_per_pos) if self.__training else 0),\n",
    "            device=self.device\n",
    "        )\n",
    "        x[:len(batch_df)] = 1\n",
    "        return x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build, train and predict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1669697"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = HLA_ModelInterface(model_class=HLA_LSTM)\n",
    "model.get_parameter_num()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Training] Epoch=1, lr=5e-06, loss=0.6932349891534874\n",
      "[Training] Epoch=2, lr=1e-05, loss=0.6930731327405998\n",
      "[Training] Epoch=3, lr=1.5e-05, loss=0.6907350452882903\n",
      "[Training] Epoch=4, lr=2e-05, loss=0.6277477278241089\n",
      "[Training] Epoch=5, lr=2.5e-05, loss=0.5420684284929719\n",
      "[Training] Epoch=6, lr=3e-05, loss=0.4841006744120802\n",
      "[Training] Epoch=7, lr=3.5e-05, loss=0.4482630943613393\n",
      "[Training] Epoch=8, lr=4e-05, loss=0.4203483752374138\n",
      "[Training] Epoch=9, lr=4.5e-05, loss=0.4024100367512022\n",
      "[Training] Epoch=10, lr=5e-05, loss=0.39221225013690336\n",
      "[Training] Epoch=11, lr=5.500000000000001e-05, loss=0.3849370711083923\n",
      "[Training] Epoch=12, lr=6e-05, loss=0.3784518970974854\n",
      "[Training] Epoch=13, lr=6.500000000000001e-05, loss=0.37486615936670986\n",
      "[Training] Epoch=14, lr=7e-05, loss=0.36704409840915886\n",
      "[Training] Epoch=15, lr=7.500000000000001e-05, loss=0.3703517203352281\n",
      "[Training] Epoch=16, lr=8e-05, loss=0.36510397733322214\n",
      "[Training] Epoch=17, lr=8.5e-05, loss=0.36357397745762554\n",
      "[Training] Epoch=18, lr=9e-05, loss=0.3598494923540524\n",
      "[Training] Epoch=19, lr=9.5e-05, loss=0.3568886015564203\n",
      "[Training] Epoch=20, lr=0.0001, loss=0.35551583607281956\n",
      "[Training] Epoch=21, lr=9.996145181203615e-05, loss=0.351618001769696\n",
      "[Training] Epoch=22, lr=9.98458666866564e-05, loss=0.34656563454440664\n",
      "[Training] Epoch=23, lr=9.965342284774632e-05, loss=0.3421939039336784\n",
      "[Training] Epoch=24, lr=9.938441702975689e-05, loss=0.34410261682101656\n",
      "[Training] Epoch=25, lr=9.903926402016153e-05, loss=0.33958916632192476\n",
      "[Training] Epoch=26, lr=9.861849601988383e-05, loss=0.3389470726251602\n",
      "[Training] Epoch=27, lr=9.812276182268236e-05, loss=0.33507983865482466\n",
      "[Training] Epoch=28, lr=9.755282581475769e-05, loss=0.332846123991268\n",
      "[Training] Epoch=29, lr=9.690956679612421e-05, loss=0.33244901177074226\n",
      "[Training] Epoch=30, lr=9.619397662556435e-05, loss=0.3289957943239382\n",
      "[Training] Epoch=31, lr=9.540715869125407e-05, loss=0.32688310981861185\n",
      "[Training] Epoch=32, lr=9.45503262094184e-05, loss=0.3219278638384172\n",
      "[Training] Epoch=33, lr=9.362480035363986e-05, loss=0.3218766458864723\n",
      "[Training] Epoch=34, lr=9.263200821770461e-05, loss=0.31904784137649195\n",
      "[Training] Epoch=35, lr=9.157348061512727e-05, loss=0.3186663123113768\n",
      "[Training] Epoch=36, lr=9.045084971874738e-05, loss=0.31249263430280344\n",
      "[Training] Epoch=37, lr=8.926584654403724e-05, loss=0.31230173259973526\n",
      "[Training] Epoch=38, lr=8.802029828000156e-05, loss=0.31291880112673554\n",
      "[Training] Epoch=39, lr=8.671612547178428e-05, loss=0.30887279619595837\n",
      "[Training] Epoch=40, lr=8.535533905932738e-05, loss=0.3110689180237906\n",
      "[Training] Epoch=41, lr=8.39400372766471e-05, loss=0.30739381403795313\n",
      "[Training] Epoch=42, lr=8.247240241650918e-05, loss=0.3040785589920623\n",
      "[Training] Epoch=43, lr=8.095469746549172e-05, loss=0.30485487835747854\n",
      "[Training] Epoch=44, lr=7.938926261462366e-05, loss=0.3053571893168347\n",
      "[Training] Epoch=45, lr=7.777851165098012e-05, loss=0.30356999513294014\n",
      "[Training] Epoch=46, lr=7.612492823579745e-05, loss=0.30445077110614094\n",
      "[Training] Epoch=47, lr=7.443106207484776e-05, loss=0.2998486476551209\n",
      "[Training] Epoch=48, lr=7.269952498697734e-05, loss=0.30041886600000517\n",
      "[Training] Epoch=49, lr=7.09329868768714e-05, loss=0.29990482543195995\n",
      "[Training] Epoch=50, lr=6.91341716182545e-05, loss=0.2962615843862295\n",
      "[Training] Epoch=51, lr=6.730585285387465e-05, loss=0.2953455512012754\n",
      "[Training] Epoch=52, lr=6.545084971874738e-05, loss=0.2983233828895858\n",
      "[Training] Epoch=53, lr=6.357202249325371e-05, loss=0.29659172386995386\n",
      "[Training] Epoch=54, lr=6.167226819279528e-05, loss=0.2934789817248072\n",
      "[Training] Epoch=55, lr=5.9754516100806423e-05, loss=0.29255389847925733\n",
      "[Training] Epoch=56, lr=5.782172325201155e-05, loss=0.29031418782791923\n",
      "[Training] Epoch=57, lr=5.587686987289189e-05, loss=0.29398182925901245\n",
      "[Training] Epoch=58, lr=5.392295478639225e-05, loss=0.2915504536192332\n",
      "[Training] Epoch=59, lr=5.196299078795344e-05, loss=0.29035683535039425\n",
      "[Training] Epoch=60, lr=5e-05, loss=0.2890172975936106\n",
      "[Training] Epoch=61, lr=4.8037009212046586e-05, loss=0.28899696349565474\n",
      "[Training] Epoch=62, lr=4.607704521360776e-05, loss=0.29178350432110683\n",
      "[Training] Epoch=63, lr=4.412313012710813e-05, loss=0.2871770095080137\n",
      "[Training] Epoch=64, lr=4.2178276747988446e-05, loss=0.2871080610368933\n",
      "[Training] Epoch=65, lr=4.0245483899193595e-05, loss=0.2866311443171331\n",
      "[Training] Epoch=66, lr=3.832773180720475e-05, loss=0.28931954729237724\n",
      "[Training] Epoch=67, lr=3.642797750674629e-05, loss=0.28504496267331497\n",
      "[Training] Epoch=68, lr=3.4549150281252636e-05, loss=0.2868912652400987\n",
      "[Training] Epoch=69, lr=3.2694147146125345e-05, loss=0.2862606699179326\n",
      "[Training] Epoch=70, lr=3.086582838174551e-05, loss=0.2859947222418019\n",
      "[Training] Epoch=71, lr=2.9067013123128613e-05, loss=0.2840251099052174\n",
      "[Training] Epoch=72, lr=2.7300475013022663e-05, loss=0.28321384771593977\n",
      "[Training] Epoch=73, lr=2.556893792515227e-05, loss=0.2838441148134215\n",
      "[Training] Epoch=74, lr=2.3875071764202563e-05, loss=0.2838261653003948\n",
      "[Training] Epoch=75, lr=2.2221488349019903e-05, loss=0.2828914923593402\n",
      "[Training] Epoch=76, lr=2.061073738537635e-05, loss=0.28325096557715107\n",
      "[Training] Epoch=77, lr=1.9045302534508297e-05, loss=0.2845324899203011\n",
      "[Training] Epoch=78, lr=1.7527597583490822e-05, loss=0.2816537321944322\n",
      "[Training] Epoch=79, lr=1.605996272335291e-05, loss=0.2819592153121318\n",
      "[Training] Epoch=80, lr=1.4644660940672627e-05, loss=0.2798138106507914\n",
      "[Training] Epoch=81, lr=1.3283874528215733e-05, loss=0.2817491751962474\n",
      "[Training] Epoch=82, lr=1.1979701719998453e-05, loss=0.2798189613968134\n",
      "[Training] Epoch=83, lr=1.0734153455962765e-05, loss=0.28218166237430914\n",
      "[Training] Epoch=84, lr=9.549150281252633e-06, loss=0.2786750893241593\n",
      "[Training] Epoch=85, lr=8.426519384872733e-06, loss=0.279946924187243\n",
      "[Training] Epoch=86, lr=7.367991782295391e-06, loss=0.2787400723568031\n",
      "[Training] Epoch=87, lr=6.375199646360142e-06, loss=0.27927093487232924\n",
      "[Training] Epoch=88, lr=5.449673790581611e-06, loss=0.279260845189648\n",
      "[Training] Epoch=89, lr=4.592841308745932e-06, loss=0.28106644683118376\n",
      "[Training] Epoch=90, lr=3.8060233744356633e-06, loss=0.2817429155111313\n",
      "[Training] Epoch=91, lr=3.0904332038757977e-06, loss=0.27846601472369265\n",
      "[Training] Epoch=92, lr=2.4471741852423237e-06, loss=0.2785775065422058\n",
      "[Training] Epoch=93, lr=1.8772381773176417e-06, loss=0.27893196724887404\n",
      "[Training] Epoch=94, lr=1.3815039801161721e-06, loss=0.27897446136921644\n",
      "[Training] Epoch=95, lr=9.607359798384785e-07, loss=0.2776993236371449\n",
      "[Training] Epoch=96, lr=6.15582970243117e-07, loss=0.27884254830756355\n",
      "[Training] Epoch=97, lr=3.465771522536854e-07, loss=0.27829928629632505\n",
      "[Training] Epoch=98, lr=1.5413331334360182e-07, loss=0.27700028847903013\n",
      "[Training] Epoch=99, lr=3.8548187963854956e-08, loss=0.2769414125276463\n",
      "[Training] Epoch=100, lr=0.0, loss=0.2781089919486216\n"
     ]
    }
   ],
   "source": [
    "model.train(pos_train_df, prot_df=prot_df, epoch=100, warmup_epoch=20, verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.save(save_as)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>nAA</th>\n",
       "      <th>HLA</th>\n",
       "      <th>HLA_prob_pred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KTDAEFVCERTLKY</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>0.964588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KSAIPHPLIM</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>0.915043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KSLVIELDKELY</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.974512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EVKRLKVTELR</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>0.968476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAEDDEDDDVDTKK</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>0.594708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63883</th>\n",
       "      <td>QKFHINIHILKDCP</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63884</th>\n",
       "      <td>HTACMMGVTFCANN</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63885</th>\n",
       "      <td>LEEAINVATAAMQQ</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0.052008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63886</th>\n",
       "      <td>ASPSVQGAPREVVD</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0.119104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63887</th>\n",
       "      <td>KNVEPEKNESWEWV</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0.010799</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63888 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             sequence  nAA  HLA  HLA_prob_pred\n",
       "0      KTDAEFVCERTLKY   14    1       0.964588\n",
       "1          KSAIPHPLIM   10    1       0.915043\n",
       "2        KSLVIELDKELY   12    1       0.974512\n",
       "3         EVKRLKVTELR   11    1       0.968476\n",
       "4      AAEDDEDDDVDTKK   14    1       0.594708\n",
       "...               ...  ...  ...            ...\n",
       "63883  QKFHINIHILKDCP   14    0       0.001380\n",
       "63884  HTACMMGVTFCANN   14    0       0.000554\n",
       "63885  LEEAINVATAAMQQ   14    0       0.052008\n",
       "63886  ASPSVQGAPREVVD   14    0       0.119104\n",
       "63887  KNVEPEKNESWEWV   14    0       0.010799\n",
       "\n",
       "[63888 rows x 4 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df = concat_neg_df(pos_test_df, prot_df)\n",
    "model.predict(test_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Precision = 0.893620613961028\n",
      "Recall = 0.8283558727773603\n",
      "False Positive Rate = 0.09861006761833213\n"
     ]
    }
   ],
   "source": [
    "prob=0.7\n",
    "print(\"Precision =\", test_df[test_df.HLA_prob_pred>prob]['HLA'].mean())\n",
    "print(\"Recall =\", test_df[test_df.HLA_prob_pred>prob]['HLA'].sum()/len(test_df)*2)\n",
    "print(\"False Positive Rate =\", 1-(1-test_df[test_df.HLA_prob_pred<prob]['HLA']).sum()/len(test_df)*2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Apply the model for all protein sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11384928"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cat_prot = '$'+'$'.join(prot_df.sequence.values)+'$'\n",
    "len(cat_prot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "pos_starts, pos_ends = get_substring_indices(cat_prot, 8, 14)\n",
    "digest_df = pd.DataFrame(dict(start_pos=pos_starts, end_pos=pos_ends))\n",
    "digest_df[\"nAA\"] = digest_df.end_pos-digest_df.start_pos\n",
    "digest_df.sort_values('nAA', inplace=True)\n",
    "digest_df.reset_index(inplace=True, drop=True)\n",
    "digest_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tqdm\n",
    "\n",
    "def get_seq(x, cat_prot=cat_prot):\n",
    "    return cat_prot[slice(x[0],x[1])]\n",
    "\n",
    "def get_seq_series(df, cat_prot):\n",
    "    return df[[\"start_pos\",\"end_pos\"]].apply(\n",
    "        get_seq, axis=1, cat_prot=cat_prot\n",
    "    )\n",
    "\n",
    "def predict_digest_df(model, substr_index_df, cat_prot, batch_size=1024000):\n",
    "    substr_index_df['HLA_prob_pred'] = 0.0\n",
    "    for i in tqdm.tqdm(range(0, len(substr_index_df), batch_size)):\n",
    "        _df = substr_index_df.iloc[i:i+batch_size]\n",
    "        seq_df = get_seq_series(_df, cat_prot).to_frame('sequence')\n",
    "        seq_df['nAA'] = _df.nAA\n",
    "        model.predict(seq_df, batch_size=4096)\n",
    "        substr_index_df.HLA_prob_pred.values[i:i+batch_size] = seq_df.HLA_prob_pred\n",
    "    return substr_index_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test\n",
    "slice_df = digest_df.iloc[100:10000].copy()\n",
    "predict_digest_df(model, slice_df, cat_prot, batch_size=1024)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "digest_df = predict_digest_df(model, digest_df, cat_prot, batch_size=1024000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get all sequences with >prob (0.7) predicted probabilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hla_df = digest_df[digest_df.HLA_prob_pred>prob].copy()\n",
    "hla_df['sequence'] = get_seq_series(hla_df, cat_prot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hla_df.to_csv('Predicted_HLA.tsv',index=False, sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.3 ('base')",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
