<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>alphapeptdeep - AlphaPeptDeep</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./peptdeep.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="alphapeptdeep - AlphaPeptDeep">
<meta property="og:description" content="">
<meta property="og:site-name" content="alphapeptdeep">
<meta name="twitter:title" content="alphapeptdeep - AlphaPeptDeep">
<meta name="twitter:description" content="">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">alphapeptdeep</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/MannLabs/alphapeptdeep"><i class="bi bi-github" role="img">
</i> 
 </a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">AlphaPeptDeep</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link active">AlphaPeptDeep</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pipeline_api.html" class="sidebar-item-text sidebar-link">pipeline API</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pretrained_models.html" class="sidebar-item-text sidebar-link">Pretrained Models</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./settings.html" class="sidebar-item-text sidebar-link">Settings</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./test_ui.html" class="sidebar-item-text sidebar-link">Testing GUI</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./utils.html" class="sidebar-item-text sidebar-link">Utils</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">mass_spec</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mass_spec/mass_calibration.html" class="sidebar-item-text sidebar-link">Mass calibration</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mass_spec/match.html" class="sidebar-item-text sidebar-link">Match</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mass_spec/ms_reader.html" class="sidebar-item-text sidebar-link">MS Reader</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">model</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./model/building_block.html" class="sidebar-item-text sidebar-link">NN Building Block</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./model/ccs.html" class="sidebar-item-text sidebar-link">CCS model and interface</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./model/featurize.html" class="sidebar-item-text sidebar-link">Featurize</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./model/generic_property_prediction.html" class="sidebar-item-text sidebar-link">Generic Property Prediction</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./model/model_interface.html" class="sidebar-item-text sidebar-link">Model Interface</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./model/model_shop.html" class="sidebar-item-text sidebar-link">Model shop</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./model/ms2.html" class="sidebar-item-text sidebar-link">MS2 prediction</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./model/rt.html" class="sidebar-item-text sidebar-link">Retention Time Prediction</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">protein</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./protein/fasta.html" class="sidebar-item-text sidebar-link">Spectral Library from fasta</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">psm_frag_reader</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./psm_frag_reader/library_frag_reader.html" class="sidebar-item-text sidebar-link">TSV Fragment Reader</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./psm_frag_reader/maxquant_frag_reader.html" class="sidebar-item-text sidebar-link">MaxQuant Fragment Reader</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./psm_frag_reader/psm_frag_reader.html" class="sidebar-item-text sidebar-link">Base PSM Fragment Reader</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./psm_frag_reader/psmlabel_reader.html" class="sidebar-item-text sidebar-link">.psmlabel Fragment Reader</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">rescore</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./rescore/fdr.html" class="sidebar-item-text sidebar-link">FDR</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./rescore/feature_extractor.html" class="sidebar-item-text sidebar-link">Feature Extractor</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./rescore/percolator.html" class="sidebar-item-text sidebar-link">Percolator</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">spec_lib</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spec_lib/library_factory.html" class="sidebar-item-text sidebar-link">Library Factory</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spec_lib/predict_lib.html" class="sidebar-item-text sidebar-link">Predict Spectral Library</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spec_lib/test_translate_tsv.html" class="sidebar-item-text sidebar-link">test_translate_tsv.html</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spec_lib/translate.html" class="sidebar-item-text sidebar-link">Translate Spectral Library</a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#about" id="toc-about" class="nav-link active" data-scroll-target="#about">About</a>
  <ul class="collapse">
  <li><a href="#alpha-ms-repositories" id="toc-alpha-ms-repositories" class="nav-link" data-scroll-target="#alpha-ms-repositories">Alpha-MS repositories:</a></li>
  </ul></li>
  <li><a href="#license" id="toc-license" class="nav-link" data-scroll-target="#license">License</a></li>
  <li><a href="#installation" id="toc-installation" class="nav-link" data-scroll-target="#installation">Installation</a>
  <ul class="collapse">
  <li><a href="#one-click-gui" id="toc-one-click-gui" class="nav-link" data-scroll-target="#one-click-gui">One-click GUI</a></li>
  <li><a href="#pip" id="toc-pip" class="nav-link" data-scroll-target="#pip">Pip</a></li>
  <li><a href="#use-gpu" id="toc-use-gpu" class="nav-link" data-scroll-target="#use-gpu">Use GPU</a></li>
  <li><a href="#developer" id="toc-developer" class="nav-link" data-scroll-target="#developer">Developer</a></li>
  </ul></li>
  <li><a href="#usage" id="toc-usage" class="nav-link" data-scroll-target="#usage">Usage</a>
  <ul class="collapse">
  <li><a href="#gui" id="toc-gui" class="nav-link" data-scroll-target="#gui">GUI</a></li>
  <li><a href="#cli" id="toc-cli" class="nav-link" data-scroll-target="#cli">CLI</a></li>
  <li><a href="#python-and-jupyter-notebooks" id="toc-python-and-jupyter-notebooks" class="nav-link" data-scroll-target="#python-and-jupyter-notebooks">Python and Jupyter notebooks</a></li>
  </ul></li>
  <li><a href="#troubleshooting" id="toc-troubleshooting" class="nav-link" data-scroll-target="#troubleshooting">Troubleshooting</a></li>
  <li><a href="#citations" id="toc-citations" class="nav-link" data-scroll-target="#citations">Citations</a></li>
  <li><a href="#how-to-contribute" id="toc-how-to-contribute" class="nav-link" data-scroll-target="#how-to-contribute">How to contribute</a></li>
  <li><a href="#changelog" id="toc-changelog" class="nav-link" data-scroll-target="#changelog">Changelog</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/MannLabs/alphapeptdeep/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">AlphaPeptDeep</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p><a href="https://github.com/MannLabs/alphapeptdeep/workflows/Default%20installation%20and%20tests/badge.svg" alt="Pip Installation"> <img src="https://github.com/MannLabs/alphapeptdeep/workflows/Default%20installation%20and%20tests/badge.svg"></a> <a href="https://github.com/MannLabs/alphapeptdeep/workflows/Publish%20on%20PyPi%20and%20release%20on%20GitHub/badge.svg" alt="GUI and PyPi releases"> <img src="https://github.com/MannLabs/alphapeptdeep/workflows/Publish%20on%20PyPi%20and%20release%20on%20GitHub/badge.svg"></a> <a href="https://pypi.org/project/peptdeep" alt="pypi"> <img src="https://img.shields.io/pypi/v/peptdeep"></a> <a href="https://github.com/MannLabs/alphapeptdeep/releases" alt="release"> <img src="https://img.shields.io/github/v/release/mannlabs/alphapeptdeep?display_name=tag"></a> <a href="https://github.com/MannLabs/alphapeptdeep/releases" alt="release"> <img src="https://img.shields.io/github/downloads/mannlabs/alphapeptdeep/total?label=github%20downloads"></a> <a href="https://github.com/MannLabs/alphapeptdeep/releases/tag/pre-trained-models" alt="release"> <img src="https://img.shields.io/github/downloads/mannlabs/alphapeptdeep/pre-trained-models/total"></a> <a href="https://pypi.org/project/peptdeep" alt="release"> <img src="https://img.shields.io/pypi/dm/peptdeep?color=blue&amp;label=pip%20downloads"></a> <img src="https://img.shields.io/pypi/pyversions/peptdeep"></p>
<ul>
<li><a href="#about"><strong>About</strong></a></li>
<li><a href="#license"><strong>License</strong></a></li>
<li><a href="#installation"><strong>Installation</strong></a>
<ul>
<li><a href="#one-click-gui"><strong>One-click GUI</strong></a></li>
<li><a href="#pip"><strong>Pip installer</strong></a></li>
<li><a href="#use-gpu"><strong>Use GPU</strong></a></li>
<li><a href="#developer"><strong>Developer installer</strong></a></li>
</ul></li>
<li><a href="#usage"><strong>Usage</strong></a>
<ul>
<li><a href="#gui"><strong>GUI</strong></a></li>
<li><a href="#cli"><strong>CLI</strong></a></li>
<li><a href="#python-and-jupyter-notebooks"><strong>Python and jupyter notebooks</strong></a></li>
</ul></li>
<li><a href="#troubleshooting"><strong>Troubleshooting</strong></a></li>
<li><a href="#citations"><strong>Citations</strong></a></li>
<li><a href="#how-to-contribute"><strong>How to contribute</strong></a></li>
<li><a href="#changelog"><strong>Changelog</strong></a></li>
</ul>
<hr>
<section id="about" class="level2">
<h2 class="anchored" data-anchor-id="about">About</h2>
<p>AlphaPeptDeep (<code>peptdeep</code> for short) aims to easily build new deep learning models for shotgun proteomics studies. Transfer learning is also easy to apply using AlphaPeptDeep.</p>
<p>It contains some built-in models such as retention time (RT), collision cross section (CCS), and tandem mass spectrum (MS2) prediction for given peptides. With these models, one can easily generate a predicted library from fasta files.</p>
<p>For details, check out our <a href="#citations">publications</a>.</p>
<p>For documentation, see <a href="https://mannlabs.github.io/alphapeptdeep/">GitHub pages</a>.</p>
<section id="alpha-ms-repositories" class="level3">
<h3 class="anchored" data-anchor-id="alpha-ms-repositories">Alpha-MS repositories:</h3>
<ul>
<li><a href="https://github.com/MannLabs/alphabase"><strong>alphabase</strong></a>: Base data structures for Alpha-MS-Ecosystem</li>
<li><a href="https://github.com/MannLabs/alphapept"><strong>alphapept</strong></a>: DDA search engine</li>
<li><a href="https://github.com/MannLabs/alphapeptdeep"><strong>alphapeptdeep</strong></a>: Deep learning for proteomics</li>
<li><a href="https://github.com/MannLabs/alpharaw"><strong>alpharaw</strong></a>: Raw data accessing</li>
<li><a href="https://github.com/MannLabs/alphaviz"><strong>alphaviz</strong></a>: MS data and results visualization</li>
<li><a href="https://github.com/MannLabs/alphatims"><strong>alphatims</strong></a>: timsTOF data accessing</li>
</ul>
<hr>
</section>
</section>
<section id="license" class="level2">
<h2 class="anchored" data-anchor-id="license">License</h2>
<p>AlphaPeptDeep was developed by the <a href="https://www.biochem.mpg.de/mann">Mann Labs at the Max Planck Institute of Biochemistry</a> and the <a href="https://www.cpr.ku.dk/research/proteomics/mann/">University of Copenhagen</a> and is freely available with an <a href="LICENSE.txt">Apache License</a>. External Python packages (available in the <a href="requirements">requirements</a> folder) have their own licenses, which can be consulted on their respective websites.</p>
<hr>
</section>
<section id="installation" class="level2">
<h2 class="anchored" data-anchor-id="installation">Installation</h2>
<p>AlphaPeptDeep can be installed and used on all major operating systems (Windows, macOS and Linux).</p>
<p>There are three different types of installation possible:</p>
<ul>
<li><a href="#one-click-gui"><strong>One-click GUI installer:</strong></a> Choose this installation if you only want the GUI and/or keep things as simple as possible.</li>
<li><a href="#pip"><strong>Pip installer:</strong></a> Choose this installation if you want to use peptdeep as a Python package in an existing Python (recommended Python 3.8) environment (e.g.&nbsp;a Jupyter notebook). If needed, the GUI and CLI can be installed with pip as well.</li>
<li><a href="#developer"><strong>Developer installer:</strong></a> Choose this installation if you are familiar with CLI tools, <a href="https://docs.conda.io/en/latest/">conda</a> and Python. This installation allows access to all available features of peptdeep and even allows to modify its source code directly. Generally, the developer version of peptdeep outperforms the precompiled versions which makes this the installation of choice for high-throughput experiments.</li>
</ul>
<section id="one-click-gui" class="level3">
<h3 class="anchored" data-anchor-id="one-click-gui">One-click GUI</h3>
<p>The GUI of peptdeep is a completely stand-alone tool that requires no knowledge of Python or CLI tools. Click on one of the links below to download the latest release for:</p>
<ul>
<li><a href="https://github.com/MannLabs/alphapeptdeep/releases/latest/download/peptdeep_gui_installer_windows.exe"><strong>Windows</strong></a></li>
<li><a href="https://github.com/MannLabs/alphapeptdeep/releases/latest/download/peptdeep_gui_installer_macos.pkg"><strong>macOS</strong></a></li>
<li><a href="https://github.com/MannLabs/alphapeptdeep/releases/latest/download/peptdeep_gui_installer_linux.deb"><strong>Linux</strong></a></li>
</ul>
<p>Older releases remain available on the <a href="https://github.com/MannLabs/alphapeptdeep/releases">release page</a>, but no backwards compatibility is guaranteed.</p>
<p>Note that these installers do not have GPU supported.</p>
</section>
<section id="pip" class="level3">
<h3 class="anchored" data-anchor-id="pip">Pip</h3>
<blockquote class="blockquote">
<p>PythonNET must be installed to access Thermo or Sciex raw data.</p>
<p><em>Legacy, should be replaced by AlphaRaw in the near future.</em></p>
<h4 id="pythonnet-in-windows" class="anchored">PythonNET in Windows</h4>
<p>Automatically installed for Windows.</p>
<h4 id="pythonnet-in-linux" class="anchored">PythonNET in Linux</h4>
<ol type="1">
<li>Install Mono from mono-project website <a href="https://www.mono-project.com/download/stable/#download-lin">Mono Linux</a>. NOTE, the installed mono version should be at least 6.10, which requires you to add the ppa to your trusted sources!</li>
<li>Install PythonNET with <code>pip install pythonnet</code>.</li>
</ol>
<h4 id="pythonnet-in-macos" class="anchored">PythonNET in MacOS</h4>
<ol type="1">
<li>Install <a href="https://brew.sh">brew</a> and pkg-config: <code>brew install pkg-config</code> 3. Install Mono from mono-project website <a href="https://www.mono-project.com/download/stable/">Mono Mac</a></li>
<li>Register the Mono-Path to your system: For macOS Catalina, open the configuration of zsh via the terminal:</li>
</ol>
<ul>
<li>Type <code>nano ~/.zshrc</code> to open the configuration of the terminal</li>
<li>Append the mono path to your <code>PKG_CONFIG_PATH</code>: <code>export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig:/usr/lib/pkgconfig:/Library/Frameworks/Mono.framework/Versions/Current/lib/pkgconfig:$PKG_CONFIG_PATH</code>.</li>
<li>Save everything and execute <code>. ~/.zshrc</code></li>
</ul>
<ol start="3" type="1">
<li>Install PythonNET with <code>pip install pythonnet</code>.</li>
</ol>
</blockquote>
<p>peptdeep can be installed in an existing Python 3.8 environment with a single <code>bash</code> command. <em>This <code>bash</code> command can also be run directly from within a Jupyter notebook by prepending it with a <code>!</code></em>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install peptdeep</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Installing peptdeep like this avoids conflicts when integrating it in other tools, as this does not enforce strict versioning of dependancies. However, if new versions of dependancies are released, they are not guaranteed to be fully compatible with peptdeep. This should only occur in rare cases where dependencies are not backwards compatible.</p>
<blockquote class="blockquote">
<p><strong>TODO</strong> You can always force peptdeep to use dependancy versions which are known to be compatible with:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install <span class="st">"peptdeep[stable]"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>NOTE: You might need to run <code>pip install pip</code> before installing peptdeep like this. Also note the double quotes <code>"</code>.</p>
</blockquote>
<p>For those who are really adventurous, it is also possible to directly install any branch (e.g.&nbsp;<code>@development</code>) with any extras (e.g.&nbsp;<code>#egg=peptdeep[stable,development-stable]</code>) from GitHub with e.g.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install <span class="st">"git+https://github.com/MannLabs/alphapeptdeep.git@development#egg=peptdeep[stable,development-stable]"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="use-gpu" class="level3">
<h3 class="anchored" data-anchor-id="use-gpu">Use GPU</h3>
<p>To enable GPU, GPU version of PyTorch is required, it can be installed with:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install torch <span class="at">--extra-index-url</span> https://download.pytorch.org/whl/cu116 <span class="at">--upgrade</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Note that this may depend on your NVIDIA driver version. Run the command to check your NVIDIA driver:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">nvidia-smi</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>See <a href="https://pytorch.org/get-started/locally/">pytorch.org</a> for details.</p>
</section>
<section id="developer" class="level3">
<h3 class="anchored" data-anchor-id="developer">Developer</h3>
<p>peptdeep can also be installed in editable (i.e.&nbsp;developer) mode with a few <code>bash</code> commands. This allows to fully customize the software and even modify the source code to your specific needs. When an editable Python package is installed, its source code is stored in a transparent location of your choice. While optional, it is advised to first (create and) navigate to e.g.&nbsp;a general software folder:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> ~/alphapeptdeep/project/folder</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/alphapeptdeep/project/folder</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong><em>The following commands assume you do not perform any additional <code>cd</code> commands anymore</em></strong>.</p>
<p>Next, download the peptdeep repository from GitHub either directly or with a <code>git</code> command. This creates a new peptdeep subfolder in your current directory.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/MannLabs/alphapeptdeep.git</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>For any Python package, it is highly recommended to use a separate <a href="https://docs.conda.io/en/latest/">conda virtual environment</a>, as otherwise <em>dependancy conflicts can occur with already existing packages</em>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">--name</span> peptdeep python=3.8 <span class="at">-y</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate peptdeep</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Finally, peptdeep and all its <a href="requirements">dependancies</a> need to be installed. To take advantage of all features and allow development (with the <code>-e</code> flag), this is best done by also installing the <a href="requirements/requirements_development.txt">development dependencies</a> instead of only the <a href="requirements/requirements.txt">core dependencies</a>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install <span class="at">-e</span> <span class="st">".[development]"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>By default this installs loose dependancies (no explicit versioning), although it is also possible to use stable dependencies (e.g.&nbsp;<code>pip install -e ".[stable,development-stable]"</code>).</p>
<p><strong><em>By using the editable flag <code>-e</code>, all modifications to the <a href="peptdeep">peptdeep source code folder</a> are directly reflected when running peptdeep. Note that the peptdeep folder cannot be moved and/or renamed if an editable version is installed. In case of confusion, you can always retrieve the location of any Python module with e.g.&nbsp;the command <code>import module</code> followed by <code>module.__file__</code>.</em></strong></p>
<p>We used <a href="https://nbdev.fast.ai/">nbdev v2</a> for developers to build Python source code and docs smoothly from Python notebooks, so please do not edit .py files directly, edit .ipynb in <code>nbdev_nbs</code> folder instead. After installing nbdev, cd to alphapeptdeep project folder and run:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">nbdev_install_hooks</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>to init gitconfig for nbdev. After editing the source code in .ipynb files, using <code>nbdev_export</code> to build python source code and <code>nbdev_test</code> to run all .ipynb files in <code>nbdev_nbs</code> for testing. Check <a href="https://nbdev.fast.ai/">nbdev docs</a> for more information.</p>
<hr>
</section>
</section>
<section id="usage" class="level2">
<h2 class="anchored" data-anchor-id="usage">Usage</h2>
<p>There are three ways to use peptdeep:</p>
<ul>
<li><a href="#gui"><strong>GUI</strong></a></li>
<li><a href="#cli"><strong>CLI</strong></a></li>
<li><a href="#python-and-jupyter-notebooks"><strong>Python</strong></a></li>
</ul>
<p>NOTE: The first time you use a fresh installation of peptdeep, it is often quite slow because some functions might still need compilation on your local operating system and architecture. Subsequent use should be a lot faster.</p>
<section id="gui" class="level3">
<h3 class="anchored" data-anchor-id="gui">GUI</h3>
<p>If the GUI was not installed through a one-click GUI installer, it can be launched with the following <code>bash</code> command:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">peptdeep</span> gui</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This command will start a web server and automatically open the default browser: <img src="https://user-images.githubusercontent.com/4646029/189301730-ac1f92cc-0e9d-4ba3-be1d-07c4d66032cd.jpg" class="img-fluid"></p>
<p>There are several options in the GUI (left panel):</p>
<ul>
<li>Server: Start/stop the task server, check tasks in the task queue</li>
<li>Settings: Configure common settings, load/save current settings</li>
<li>Model: Configure DL models for prediction or transfer learning</li>
<li>Transfer: Refine the models</li>
<li>Library: Predict a library</li>
<li>Rescore: Perform ML feature extraction and Percolator</li>
</ul>
<hr>
</section>
<section id="cli" class="level3">
<h3 class="anchored" data-anchor-id="cli">CLI</h3>
<p>The CLI can be run with the following command (after activating the <code>conda</code> environment with <code>conda activate peptdeep</code> or if an alias was set to the peptdeep executable):</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">peptdeep</span> <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>It is possible to get help about each function and their (required) parameters by using the <code>-h</code> flag. AlphaPeptDeep provides several commands for CLI uses:</p>
<ul>
<li><a href="#export-settings"><strong>export-settings</strong></a></li>
<li><a href="#library"><strong>library</strong></a></li>
<li><a href="#transfer"><strong>transfer</strong></a></li>
<li><a href="#rescore"><strong>rescore</strong></a></li>
<li><a href="#install-models"><strong>install-models</strong></a></li>
<li><a href="#gui"><strong>gui</strong></a></li>
</ul>
<p>Run a command to check usages:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">peptdeep</span> <span class="va">$command</span> <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>For example:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">peptdeep</span> library <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="export-settings" class="level4">
<h4 class="anchored" data-anchor-id="export-settings">export-settings</h4>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">peptdeep</span> export-settings yaml_file</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This command will export the default settings into the <code>yaml_file</code> as a template, users can edit the yaml file to run other commands.</p>
<p>Here is a section of the yaml file which controls global parameters for different tasks:</p>
<pre><code>model_url: "https://github.com/MannLabs/alphapeptdeep/releases/download/pre-trained-models/pretrained_models.zip"

thread_num: 8
torch_device:
  name: gpu
  ids: []
  name_choices:
    - gpu
    - cpu
    - mps

log_level: info
log_level_choices:
  - debug
  - info
  - warning
  - error
  - critical

common:
  modloss_importance_level: 1.0

peak_matching:
  ms2_ppm: True
  ms2_tol_value: 20.0
  ms1_ppm: True
  ms1_tol_value: 20.0

model_mgr:
  default_nce: 30.0
  default_instrument: Lumos
  mask_modloss: True
  model_type: generic
  model_choices:
  - generic
  - phos
  - hla # same as generic
  - digly
  external_ms2_model: ''
  external_rt_model: ''
  external_ccs_model: ''
  instrument_group:
    Lumos: Lumos
    QE: QE
    timsTOF: timsTOF
    SciexTOF: SciexTOF
    Fusion: Lumos
    Eclipse: Lumos
    Velos: Lumos # not important
    Elite: Lumos # not important
    OrbitrapTribrid: Lumos
    ThermoTribrid: Lumos
    QE+: QE
    QEHF: QE
    QEHFX: QE
    Exploris: QE
    Exploris480: QE
  predict:
    batch_size_ms2: 512
    batch_size_rt_ccs: 1024
    verbose: True
    multiprocessing: True</code></pre>
<p>The <code>model_mgr</code> section in the yaml defines the common settings for MS2/RT/CCS prediction.</p>
<hr>
</section>
<section id="library" class="level4">
<h4 class="anchored" data-anchor-id="library">library</h4>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">peptdeep</span> library settings_yaml</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This command will predict a spectral library for given settings_yaml file (exported by <a href="#export-settings">export-settings</a>). All the essential settings are in the <code>library</code> section in the settings_yaml file:</p>
<pre><code>library:
  input:
    infile_type: fasta
    infile_type_choices:
    - fasta
    - sequence_table
    - peptide_table
    - precursor_table
    infiles: 
    - xxx.fasta
    fasta:
      protease: '([KR])'
      protease_choices:
      - 'trypsin/P'
      - '([KR])'
      - 'trypsin'
      - '([KR](?=[^P]))'
      - 'lys-c'
      - 'K'
      - 'lys-n'
      - '\w(?=K)'
      - 'chymotrypsin'
      max_miss_cleave: 2
    fix_mods: 
    - Carbamidomethyl@C
    var_mods:
    - Acetyl@Protein N-term
    - Oxidation@M
    min_var_mod_num: 0
    max_var_mod_num: 2
    min_precursor_charge: 2
    max_precursor_charge: 4
    min_peptide_len: 7
    max_peptide_len: 35
    min_precursor_mz: 200.0
    max_precursor_mz: 2000.0
    decoy: pseudo_reverse
    decoy_choices:
    - pseudo_reverse
    - diann
    - None
    max_frag_charge: 2
    frag_types:
    - b
    - y
  output_folder: "{PEPTDEEP_HOME}/spec_libs"
  output_tsv:
    enabled: False
    min_fragment_mz: 200
    max_fragment_mz: 2000
    min_relative_intensity: 0.01
    keep_higest_k_peaks: 12
    translate_batch_size: 1000000
    translate_mod_to_unimod_id: False</code></pre>
<p>peptdeep will load sequence data based on <code>library:input:infile_type</code> and <code>library:input:infiles</code> for library prediction. <code>library:input:infiles</code> contains the list of files with <code>library:input:infile_type</code> defined in <code>library:input:infile_type_choices</code>:</p>
<ul>
<li>fasta: Protein fasta files, peptdeep will digest the protein sequences into peptide sequences.</li>
<li><a href="#sequence_table">sequence_table</a>: Tab/comma-delimited txt/tsv/csv (text) files which contain the column <code>sequence</code> for peptide sequences.</li>
<li><a href="#peptide_table">peptide_table</a>: Tab/comma-delimited txt/tsv/csv (text) files which contain the columns <code>sequence</code>, <code>mods</code>, and <code>mod_sites</code>. peptdeep will not add modifications for peptides of this file type.</li>
<li><a href="#precursor_table">precursor_table</a>: Tab/comma-delimited txt/tsv/csv (text) files which contain the columns <code>sequence</code>, <code>mods</code>, <code>mod_sites</code>, and <code>charge</code>. peptdeep will not add modifications and charge states for peptides of this file type.</li>
</ul>
<p>See examples:</p>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;python&quot;}">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame({</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">'sequence'</span>: [<span class="st">'ACDEFGHIK'</span>,<span class="st">'LMNPQRSTVK'</span>,<span class="st">'WYVSTR'</span>],</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">'mods'</span>: [<span class="st">'Carbamidomethyl@C'</span>,<span class="st">'Acetyl@Protein N-term;Phospho@S'</span>,<span class="st">''</span>],</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">'mod_sites'</span>: [<span class="st">'2'</span>,<span class="st">'0;7'</span>,<span class="st">''</span>],</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">'charge'</span>: [<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">1</span>],</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="sequence_table" class="level5">
<h5 class="anchored" data-anchor-id="sequence_table">sequence_table</h5>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;python&quot;}">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>df[[<span class="st">'sequence'</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ACDEFGHIK</td>
    </tr>
    <tr>
      <th>1</th>
      <td>LMNPQRSTVK</td>
    </tr>
    <tr>
      <th>2</th>
      <td>WYVSTR</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
<section id="peptide_table" class="level5">
<h5 class="anchored" data-anchor-id="peptide_table">peptide_table</h5>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;python&quot;}">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>df[[<span class="st">'sequence'</span>,<span class="st">'mods'</span>,<span class="st">'mod_sites'</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>mods</th>
      <th>mod_sites</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ACDEFGHIK</td>
      <td>Carbamidomethyl@C</td>
      <td>2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>LMNPQRSTVK</td>
      <td>Acetyl@Protein N-term;Phospho@S</td>
      <td>0;7</td>
    </tr>
    <tr>
      <th>2</th>
      <td>WYVSTR</td>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
<section id="precursor_table" class="level5">
<h5 class="anchored" data-anchor-id="precursor_table">precursor_table</h5>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;python&quot;}">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>charge</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ACDEFGHIK</td>
      <td>Carbamidomethyl@C</td>
      <td>2</td>
      <td>2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>LMNPQRSTVK</td>
      <td>Acetyl@Protein N-term;Phospho@S</td>
      <td>0;7</td>
      <td>3</td>
    </tr>
    <tr>
      <th>2</th>
      <td>WYVSTR</td>
      <td></td>
      <td></td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>Columns of <code>proteins</code> and <code>genes</code> are optional for these txt/tsv/csv files.</p>
</blockquote>
<p>peptdeep supports multiple files for library prediction, for example (in the yaml file):</p>
<pre><code>library:
  input:
    ...
    infile_type: fasta
    iifiles:
    - /path/to/fasta/human.fasta
    - /path/to/fasta/yeast.fasta
    ...</code></pre>
<p>The library in HDF5 (.hdf) format will be saved into <code>library:output_folder</code>. If <code>library:output_tsv:enabled</code> is True, a TSV spectral library that can be processed by DIA-NN and Spectronaut will also be saved into <code>library:output_folder</code>.</p>
<hr>
</section>
</section>
<section id="transfer" class="level4">
<h4 class="anchored" data-anchor-id="transfer">transfer</h4>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">peptdeep</span> transfer settings_yaml</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This command will apply transfer learning to refine RT/CCS/MS2 models based on <code>model_mgr:transfer:psm_files</code> and <code>model_mgr:transfer:psm_type</code>. All yaml settings (exported by <a href="#export-settings">export-settings</a>) related to this command are:</p>
<pre><code>model_mgr:
  transfer:
    model_output_folder: "{PEPTDEEP_HOME}/refined_models"
    epoch_ms2: 20
    warmup_epoch_ms2: 10
    batch_size_ms2: 512
    lr_ms2: 0.0001
    epoch_rt_ccs: 40
    warmup_epoch_rt_ccs: 10
    batch_size_rt_ccs: 1024
    lr_rt_ccs: 0.0001
    verbose: False
    grid_nce_search: True
    grid_nce_first: 15.0
    grid_nce_last: 45.0
    grid_nce_step: 3.0
    grid_instrument: ['Lumos']
    psm_type: alphapept
    psm_type_choices:
      - alphapept
      - pfind
      - maxquant
      - diann
      - speclib_tsv
    psm_files: []
    ms_file_type: alphapept_hdf
    ms_file_type_choices:
      - alphapept_hdf
      - thermo_raw
      - mgf
      - mzml
    ms_files: []
    psm_num_to_train_ms2: 100000000
    psm_num_per_mod_to_train_ms2: 50
    psm_num_to_train_rt_ccs: 100000000
    psm_num_per_mod_to_train_rt_ccs: 50
    top_n_mods_to_train: 10</code></pre>
<p>For DDA data, peptdeep can also extract MS2 intensities from the spectrum files from <code>model_mgr:transfer:ms_files</code> and <code>model_mgr:transfer:ms_file_type</code> for all PSMs. This will enable the transfer learning of the MS2 model.</p>
<p>For DIA data, only RT and CCS (if timsTOF) models will be refined.</p>
<p>For example of the settings yaml:</p>
<pre><code>model_mgr:
  transfer:
    ...
    psm_type: pfind
    psm_files:
    - /path/to/pFind.spectra
    - /path/to/other/pFind.spectra

    ms_file_type: thermo_raw
    ms_files:
    - /path/to/raw1.raw
    - /path/to/raw2.raw
    ...</code></pre>
<p>The refined models will be saved in <code>model_mgr:transfer:model_output_folder</code>. After transfer learning, users can apply the new models by replacing <code>model_mgr:external_ms2_model</code>, <code>model_mgr:external_rt_model</code> and <code>model_mgr:external_ccs_model</code> with the saved <code>ms2.pth</code>, <code>rt.pth</code> and <code>ccs.pth</code> in <code>model_mgr:transfer:model_output_folder</code>. This is useful to perform sample-specific library prediction.</p>
<hr>
</section>
<section id="rescore" class="level4">
<h4 class="anchored" data-anchor-id="rescore">rescore</h4>
<p>This command will apply Percolator to rescore DDA PSMs in <code>percolator:input_files:psm_files</code> and <code>percolator:input_files:psm_type</code>. All yaml settings (exported by <a href="#export-settings">export-settings</a>) related to this command are:</p>
<pre><code>percolator:
  require_model_tuning: True
  raw_num_to_tune: 8

  require_raw_specific_tuning: True
  raw_specific_ms2_tuning: False
  psm_num_per_raw_to_tune: 200
  epoch_per_raw_to_tune: 5

  multiprocessing: True

  top_k_frags_to_calc_spc: 10
  calibrate_frag_mass_error: False
  max_perc_train_sample: 1000000
  min_perc_train_sample: 100

  percolator_backend: sklearn
  percolator_backend_choices:
    - sklearn
    - pytorch
  percolator_model: linear
  percolator_model_choices:
    pytorch_as_backend:
      - linear # not fully tested, performance may be unstable
      - mlp # not implemented yet
    sklearn_as_backend:
      - linear # logistic regression
      - random_forest
  lr_percolator_torch_model: 0.1 # learning rate, only used when percolator_backend==pytorch 
  percolator_iter_num: 5 # percolator iteration number
  cv_fold: 1
  fdr: 0.01
  fdr_level: psm
  fdr_level_choices:
    - psm
    - precursor
    - peptide
    - sequence
  use_fdr_for_each_raw: False
  frag_types: ['b_z1','b_z2','y_z1','y_z2']
  input_files:
    psm_type: alphapept
    psm_type_choices:
      - alphapept
      - pfind
      - maxquant
    psm_files: []
    ms_file_type: alphapept_hdf
    ms_file_type_choices:
      - hdf
      - thermo_raw
      - mgf
      - mzml
    ms_files: []
    other_score_column_mapping:
      alphapept: {}
      pfind: 
        raw_score: Raw_Score
      msfragger:
        hyperscore: hyperscore
        nextscore: nextscore
      maxquant: {}
  output_folder: "{PEPTDEEP_HOME}/rescore"</code></pre>
<p>Transfer learning will be applied if <code>percolator:require_model_tuning</code> is True.</p>
<p>The corresponding MS files (<code>percolator:input_files:ms_files</code> and <code>percolator:input_files:ms_file_type</code>) must be provided to extract experimental fragment intensities.</p>
<hr>
</section>
<section id="install-models" class="level4">
<h4 class="anchored" data-anchor-id="install-models">install-models</h4>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="ex">peptdeep</span> install-models <span class="at">--model-file</span> url_or_local_model_zip <span class="at">--overwrite</span> True</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Running peptdeep for the first time, it will download and install models from <a href="https://github.com/MannLabs/alphapeptdeep/releases/download/pre-trained-models/pretrained_models.zip">models on github</a> defined in ‘model_url’ in the default yaml settings. This command will update <code>pretrained_models.zip</code> from <code>--model-file url_or_local_model_zip</code>.</p>
<p>It is also possible to use other models instead of the pretrained_models by providing <code>model_mgr:external_ms2_model</code>, <code>model_mgr:external_rt_model</code> and <code>model_mgr:external_ccs_model</code>.</p>
<hr>
</section>
</section>
<section id="python-and-jupyter-notebooks" class="level3">
<h3 class="anchored" data-anchor-id="python-and-jupyter-notebooks">Python and Jupyter notebooks</h3>
<p>Using peptdeep from Python script or notebook provides the most flexible way to access all features in peptdeep.</p>
<p>We will introduce several usages of peptdeep via Python notebook:</p>
<ul>
<li><a href="#global_settings"><strong>global_settings</strong></a></li>
<li><a href="#pipeline-apis"><strong>Pipeline APIs</strong></a></li>
<li><a href="#modelmanager"><strong>ModelManager</strong></a></li>
<li><a href="#library-prediction"><strong>Library Prediction</strong></a></li>
<li><a href="#dda-rescoring"><strong>DDA Rescoring</strong></a></li>
<li><a href="#hla-peptide-prediction"><strong>HLA Peptide Prediction</strong></a></li>
</ul>
<hr>
<section id="global_settings" class="level4">
<h4 class="anchored" data-anchor-id="global_settings">global_settings</h4>
<p>Most of the default parameters and attributes peptdeep functions and classes are controlled by <code>peptdeep.settings.global_settings</code> which is a <code>dict</code>.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> peptdeep.settings <span class="im">import</span> global_settings</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The default values of <code>global_settings</code> is defined in <a href="https://github.com/MannLabs/alphapeptdeep/blob/main/peptdeep/constants/default_settings.yaml">default_settings.yaml</a>.</p>
</section>
<section id="pipeline-apis" class="level4">
<h4 class="anchored" data-anchor-id="pipeline-apis">Pipeline APIs</h4>
<p>Pipeline APIs provides the same functionalities with <a href="#cli">CLI</a>, including <a href="#library">library prediction</a>, <a href="#transfer">transfer learning</a>, and <a href="#rescore">rescoring</a>.</p>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;python&quot;}">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> peptdeep.pipeline_api <span class="im">import</span> (</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>    generate_library,</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>    transfer_learn, </span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    rescore,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>All these functionalities take a <code>settings_dict</code> as the inputs, the dict structure is the same as the settings yaml file. See the documatation of <a href="https://MannLabs.github.io/alphapeptdeep/pipeline_api.html#generate_library"><code>generate_library</code></a>, <a href="https://MannLabs.github.io/alphapeptdeep/pipeline_api.html#transfer_learn"><code>transfer_learn</code></a>, and <a href="https://MannLabs.github.io/alphapeptdeep/pipeline_api.html#rescore"><code>rescore</code></a>.</p>
</section>
<section id="modelmanager" class="level4">
<h4 class="anchored" data-anchor-id="modelmanager">ModelManager</h4>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;python&quot;}">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> peptdeep.pretrained_models <span class="im">import</span> ModelManager</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><a href="https://MannLabs.github.io/alphapeptdeep/pretrained_models.html#modelmanager"><code>ModelManager</code></a> class is the main entry to access MS2/RT/CCS models. It provides functionalities to train/refine the models and then use the new models to predict the data.</p>
<p>Check out <a href="https://github.com/MannLabs/alphapeptdeep/blob/main/nbs/tutorial_model_manager.ipynb">tutorial_model_manager.ipynb</a> for details.</p>
</section>
<section id="library-prediction" class="level4">
<h4 class="anchored" data-anchor-id="library-prediction">Library Prediction</h4>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;python&quot;}">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> peptdeep.protein.fasta <span class="im">import</span> PredictFastaSpecLib</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><a href="https://MannLabs.github.io/alphapeptdeep/protein/fasta.html#predictfastaspeclib"><code>PredictFastaSpecLib</code></a> class provides functionalities to deal with fasta files or protein sequences and spectral libraries.</p>
<p>Check out <a href="https://github.com/MannLabs/alphapeptdeep/blob/main/nbs/tutorial_speclib_from_fasta.ipynb">tutorial_speclib_from_fasta.ipynb</a> for details.</p>
</section>
<section id="dda-rescoring" class="level4">
<h4 class="anchored" data-anchor-id="dda-rescoring">DDA Rescoring</h4>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;python&quot;}">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> peptdeep.rescore.percolator <span class="im">import</span> Percolator</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><a href="https://MannLabs.github.io/alphapeptdeep/rescore/percolator.html#percolator"><code>Percolator</code></a> class provides functionalities to rescore DDA PSMs search by <code>pFind</code> and <code>AlphaPept</code>, (and <code>MaxQuant</code> if output FDR=100%), …</p>
<p>Check out <a href="https://github.com/MannLabs/alphapeptdeep/blob/main/nbs_tests/test_percolator.ipynb">test_percolator.ipynb</a> for details.</p>
</section>
<section id="hla-peptide-prediction" class="level4">
<h4 class="anchored" data-anchor-id="hla-peptide-prediction">HLA Peptide Prediction</h4>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;python&quot;}">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> peptdeep.model.model_interface <span class="im">import</span> ModelInterface</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> peptdeep.model.generic_property_prediction <span class="co"># model shop</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Building new DL models for peptide property prediction is one of the key features of AlphaPeptDeep. The key functionalities are <a href="https://MannLabs.github.io/alphapeptdeep/model/model_interface.html#modelinterface"><code>ModelInterface</code></a> and the pre-designed models and model interfaces in the model shop (module <code>peptdeep.model.generic_property_prediction</code>).</p>
<p>For example, we can built a HLA classifier that distinguishes HLA peptides from non-HLA peptides, see <a href="https://github.com/MannLabs/alphapeptdeep/blob/main/nbs/HLA_peptide_prediction.ipynb">HLA_peptide_prediction.ipynb</a> for details.</p>
<hr>
</section>
</section>
</section>
<section id="troubleshooting" class="level2">
<h2 class="anchored" data-anchor-id="troubleshooting">Troubleshooting</h2>
<p>In case of issues, check out the following:</p>
<ul>
<li><p><a href="https://github.com/MannLabs/alphapeptdeep/issues">Issues</a>. Try a few different search terms to find out if a similar problem has been encountered before.</p></li>
<li><p><a href="https://github.com/MannLabs/alphapeptdeep/discussions">Discussions</a>. Check if your problem or feature requests has been discussed before.</p></li>
</ul>
<hr>
</section>
<section id="citations" class="level2">
<h2 class="anchored" data-anchor-id="citations">Citations</h2>
<p>Wen-Feng Zeng, Xie-Xuan Zhou, Sander Willems, Constantin Ammar, Maria Wahle, Isabell Bludau, Eugenia Voytik, Maximillian T. Strauss, Matthias Mann. <a href="https://www.biorxiv.org/content/10.1101/2022.07.14.499992v1">BioRxiv</a>.</p>
<hr>
</section>
<section id="how-to-contribute" class="level2">
<h2 class="anchored" data-anchor-id="how-to-contribute">How to contribute</h2>
<p>If you like this software, you can give us a <a href="https://github.com/MannLabs/alphapeptdeep/stargazers">star</a> to boost our visibility! All direct contributions are also welcome. Feel free to post a new <a href="https://github.com/MannLabs/alphapeptdeep/issues">issue</a> or clone the repository and create a <a href="https://github.com/MannLabs/alphapeptdeep/pulls">pull request</a> with a new branch. For an even more interactive participation, check out the <a href="https://github.com/MannLabs/alphapeptdeep/discussions">discussions</a> and the <a href="misc/CLA.md">the Contributors License Agreement</a>.</p>
<hr>
</section>
<section id="changelog" class="level2">
<h2 class="anchored" data-anchor-id="changelog">Changelog</h2>
<p>See the <a href="HISTORY.md">HISTORY.md</a> for a full overview of the changes made in each version.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>