<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.149">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>alphapeptdeep - Generic Property Prediction</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="alphapeptdeep - Generic Property Prediction">
<meta property="og:site-name" content="alphapeptdeep">
<meta name="twitter:title" content="alphapeptdeep - Generic Property Prediction">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">alphapeptdeep</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/MannLabs/alphapeptdeep/tree/main/"><i class="bi bi-github" role="img">
</i> 
 </a>
  </li>  
</ul>
              <div class="quarto-toggle-container">
                  <a href="" class="quarto-reader-toggle nav-link" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
              </div>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">Generic Property Prediction</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">Index</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pipeline_api.html" class="sidebar-item-text sidebar-link">The high-level pipeline APIs</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pretrained_models.html" class="sidebar-item-text sidebar-link">Integrated functionalities for MS2/RT/CCS models</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../settings.html" class="sidebar-item-text sidebar-link">All global settings for peptdeep</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../test_ui.html" class="sidebar-item-text sidebar-link">Testing GUI</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../utils.html" class="sidebar-item-text sidebar-link">Utils</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">mass_spec</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../mass_spec/mass_spec.mass_calibration.html" class="sidebar-item-text sidebar-link">Mass calibration</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../mass_spec/mass_spec.match.html" class="sidebar-item-text sidebar-link">Peak matching functionalities</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../mass_spec/mass_spec.ms_reader.html" class="sidebar-item-text sidebar-link">Basic spectrum file readers</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">model</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../model/model.building_block.html" class="sidebar-item-text sidebar-link">NN Building Block</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../model/model.ccs.html" class="sidebar-item-text sidebar-link">CCS model and interface</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../model/model.featurize.html" class="sidebar-item-text sidebar-link">Functionalities for featurization/tensorization</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../model/model.generic_property_prediction.html" class="sidebar-item-text sidebar-link active">Generic Property Prediction</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../model/model.model_interface.html" class="sidebar-item-text sidebar-link">Model Interface</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../model/model.model_shop.html" class="sidebar-item-text sidebar-link">legacy, see <code>generic_property_prediction.ipynb</code></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../model/model.ms2.html" class="sidebar-item-text sidebar-link">pDeepModel for MS/MS prediction</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../model/model.rt.html" class="sidebar-item-text sidebar-link">Retention Time Prediction Models</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../model/test_predict_mp.html" class="sidebar-item-text sidebar-link">Testing predict_mp()</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">protein</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../protein/protein.fasta.html" class="sidebar-item-text sidebar-link">Predicting libraries from fasta or sequences</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">psm_frag_reader</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_frag_reader/psm_frag_reader.library_frag_reader.html" class="sidebar-item-text sidebar-link">PSM reader with fragments: TSV lib</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_frag_reader/psm_frag_reader.maxquant_frag_reader.html" class="sidebar-item-text sidebar-link">PSM Reader with fragments: MaxQuant msms.txt</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_frag_reader/psm_frag_reader.psm_frag_reader.html" class="sidebar-item-text sidebar-link">PSM reader with fragments</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_frag_reader/psm_frag_reader.psmlabel_reader.html" class="sidebar-item-text sidebar-link">Legacy: psmlabel (pDeep) reader with fragments</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">rescore</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../rescore/rescore.fdr.html" class="sidebar-item-text sidebar-link">Functionalities to calculate FDRs</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../rescore/rescore.feature_extractor.html" class="sidebar-item-text sidebar-link">Functionalities to extract features from peptdeep prediction</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../rescore/rescore.percolator.html" class="sidebar-item-text sidebar-link">Percolator functionalities</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">spec_lib</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../spec_lib/spec_lib.library_factory.html" class="sidebar-item-text sidebar-link">Factory classes to predict libraries from different sources</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../spec_lib/spec_lib.predict_lib.html" class="sidebar-item-text sidebar-link">Base functionalities to predict libraries</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../spec_lib/test_translate_tsv.html" class="sidebar-item-text sidebar-link">Testing translate_tsv()</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../spec_lib/spec_lib.translate.html" class="sidebar-item-text sidebar-link">Translate peptdeep spectral libraries into other formats (e.g.&nbsp;TSV)</a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#description" id="toc-description" class="nav-link active" data-scroll-target="#description">Description</a>
  <ul>
  <li><a href="#scope-of-the-generic-property-prediction" id="toc-scope-of-the-generic-property-prediction" class="nav-link" data-scroll-target="#scope-of-the-generic-property-prediction">Scope of the generic property prediction</a></li>
  </ul></li>
  <li><a href="#imports" id="toc-imports" class="nav-link" data-scroll-target="#imports">Imports</a></li>
  <li><a href="#regression-models-for-predicting-a-scalar-value-for-a-given-amino-acid-sequence" id="toc-regression-models-for-predicting-a-scalar-value-for-a-given-amino-acid-sequence" class="nav-link" data-scroll-target="#regression-models-for-predicting-a-scalar-value-for-a-given-amino-acid-sequence">Regression models for predicting a scalar value for a given amino acid sequence</a>
  <ul>
  <li><a href="#modelinterface_for_generic_aaseq_regression" id="toc-modelinterface_for_generic_aaseq_regression" class="nav-link" data-scroll-target="#modelinterface_for_generic_aaseq_regression">ModelInterface_for_Generic_AASeq_Regression</a></li>
  <li><a href="#model_for_generic_aaseq_regression_transformer" id="toc-model_for_generic_aaseq_regression_transformer" class="nav-link" data-scroll-target="#model_for_generic_aaseq_regression_transformer">Model_for_Generic_AASeq_Regression_Transformer</a></li>
  <li><a href="#model_for_generic_aaseq_regression_lstm" id="toc-model_for_generic_aaseq_regression_lstm" class="nav-link" data-scroll-target="#model_for_generic_aaseq_regression_lstm">Model_for_Generic_AASeq_Regression_LSTM</a></li>
  <li><a href="#examples" id="toc-examples" class="nav-link" data-scroll-target="#examples">Examples</a>
  <ul class="collapse">
  <li><a href="#define-example-table" id="toc-define-example-table" class="nav-link" data-scroll-target="#define-example-table">Define example Table</a></li>
  <li><a href="#building-an-rt-model-based-on-model_for_generic_aaseq_regression_lstm" id="toc-building-an-rt-model-based-on-model_for_generic_aaseq_regression_lstm" class="nav-link" data-scroll-target="#building-an-rt-model-based-on-model_for_generic_aaseq_regression_lstm">Building an RT model based on <code>Model_for_Generic_AASeq_Regression_LSTM</code></a></li>
  <li><a href="#building-an-rt-model-for-only-sequences-based-on-model_for_generic_aaseq_regression_transformer" id="toc-building-an-rt-model-for-only-sequences-based-on-model_for_generic_aaseq_regression_transformer" class="nav-link" data-scroll-target="#building-an-rt-model-for-only-sequences-based-on-model_for_generic_aaseq_regression_transformer">Building an RT model for only sequences based on <code>Model_for_Generic_AASeq_Regression_Transformer</code></a></li>
  </ul></li>
  </ul></li>
  <li><a href="#regression-models-for-predicting-a-scalar-value-for-a-given-amino-acid-sequence-and-site-specific-ptms" id="toc-regression-models-for-predicting-a-scalar-value-for-a-given-amino-acid-sequence-and-site-specific-ptms" class="nav-link" data-scroll-target="#regression-models-for-predicting-a-scalar-value-for-a-given-amino-acid-sequence-and-site-specific-ptms">Regression models for predicting a scalar value for a given amino acid sequence and site-specific PTMs</a>
  <ul>
  <li><a href="#modelinterface_for_generic_modaaseq_regression" id="toc-modelinterface_for_generic_modaaseq_regression" class="nav-link" data-scroll-target="#modelinterface_for_generic_modaaseq_regression">ModelInterface_for_Generic_ModAASeq_Regression</a></li>
  <li><a href="#model_for_generic_modaaseq_regression_transformer" id="toc-model_for_generic_modaaseq_regression_transformer" class="nav-link" data-scroll-target="#model_for_generic_modaaseq_regression_transformer">Model_for_Generic_ModAASeq_Regression_Transformer</a></li>
  <li><a href="#model_for_generic_modaaseq_regression_lstm" id="toc-model_for_generic_modaaseq_regression_lstm" class="nav-link" data-scroll-target="#model_for_generic_modaaseq_regression_lstm">Model_for_Generic_ModAASeq_Regression_LSTM</a>
  <ul class="collapse">
  <li><a href="#scalar-regression-model-rt-with-modified-aa-sequences-using-model_for_generic_modaaseq_regression_lstm" id="toc-scalar-regression-model-rt-with-modified-aa-sequences-using-model_for_generic_modaaseq_regression_lstm" class="nav-link" data-scroll-target="#scalar-regression-model-rt-with-modified-aa-sequences-using-model_for_generic_modaaseq_regression_lstm">Scalar regression model (RT) with modified AA sequences using <code>Model_for_Generic_ModAASeq_Regression_LSTM</code></a></li>
  <li><a href="#scalar-regression-model-rt-with-modified-aa-sequences-using-model_for_generic_modaaseq_regression_transformer" id="toc-scalar-regression-model-rt-with-modified-aa-sequences-using-model_for_generic_modaaseq_regression_transformer" class="nav-link" data-scroll-target="#scalar-regression-model-rt-with-modified-aa-sequences-using-model_for_generic_modaaseq_regression_transformer">Scalar regression model (RT) with modified AA sequences using <code>Model_for_Generic_ModAASeq_Regression_Transformer</code></a></li>
  </ul></li>
  </ul></li>
  <li><a href="#binary-classification-models-for-a-given-amino-acid-sequence" id="toc-binary-classification-models-for-a-given-amino-acid-sequence" class="nav-link" data-scroll-target="#binary-classification-models-for-a-given-amino-acid-sequence">Binary classification models for a given amino acid sequence</a>
  <ul>
  <li><a href="#modelinterface_for_generic_aaseq_binaryclassification" id="toc-modelinterface_for_generic_aaseq_binaryclassification" class="nav-link" data-scroll-target="#modelinterface_for_generic_aaseq_binaryclassification">ModelInterface_for_Generic_AASeq_BinaryClassification</a></li>
  <li><a href="#model_for_generic_aaseq_binaryclassification_transformer" id="toc-model_for_generic_aaseq_binaryclassification_transformer" class="nav-link" data-scroll-target="#model_for_generic_aaseq_binaryclassification_transformer">Model_for_Generic_AASeq_BinaryClassification_Transformer</a></li>
  <li><a href="#model_for_generic_aaseq_binaryclassification_lstm" id="toc-model_for_generic_aaseq_binaryclassification_lstm" class="nav-link" data-scroll-target="#model_for_generic_aaseq_binaryclassification_lstm">Model_for_Generic_AASeq_BinaryClassification_LSTM</a></li>
  <li><a href="#examples-1" id="toc-examples-1" class="nav-link" data-scroll-target="#examples-1">Examples</a>
  <ul class="collapse">
  <li><a href="#a-sequence-classification-model-using-model_for_generic_aaseq_binaryclassification_lstm" id="toc-a-sequence-classification-model-using-model_for_generic_aaseq_binaryclassification_lstm" class="nav-link" data-scroll-target="#a-sequence-classification-model-using-model_for_generic_aaseq_binaryclassification_lstm">A sequence classification model using <code>Model_for_Generic_AASeq_BinaryClassification_LSTM</code></a></li>
  <li><a href="#a-sequence-classification-model-using-model_for_generic_aaseq_binaryclassification_transformer" id="toc-a-sequence-classification-model-using-model_for_generic_aaseq_binaryclassification_transformer" class="nav-link" data-scroll-target="#a-sequence-classification-model-using-model_for_generic_aaseq_binaryclassification_transformer">A sequence classification model using <code>Model_for_Generic_AASeq_BinaryClassification_Transformer</code></a></li>
  </ul></li>
  </ul></li>
  <li><a href="#binary-classification-models-for-given-amino-acid-sequence-and-site-specific-ptms" id="toc-binary-classification-models-for-given-amino-acid-sequence-and-site-specific-ptms" class="nav-link" data-scroll-target="#binary-classification-models-for-given-amino-acid-sequence-and-site-specific-ptms">Binary classification models for given amino acid sequence and site-specific PTMs</a>
  <ul>
  <li><a href="#modelinterface_for_generic_modaaseq_binaryclassification" id="toc-modelinterface_for_generic_modaaseq_binaryclassification" class="nav-link" data-scroll-target="#modelinterface_for_generic_modaaseq_binaryclassification">ModelInterface_for_Generic_ModAASeq_BinaryClassification</a></li>
  <li><a href="#model_for_generic_modaaseq_binaryclassification_transformer" id="toc-model_for_generic_modaaseq_binaryclassification_transformer" class="nav-link" data-scroll-target="#model_for_generic_modaaseq_binaryclassification_transformer">Model_for_Generic_ModAASeq_BinaryClassification_Transformer</a></li>
  <li><a href="#model_for_generic_modaaseq_binaryclassification_lstm" id="toc-model_for_generic_modaaseq_binaryclassification_lstm" class="nav-link" data-scroll-target="#model_for_generic_modaaseq_binaryclassification_lstm">Model_for_Generic_ModAASeq_BinaryClassification_LSTM</a></li>
  <li><a href="#examples-2" id="toc-examples-2" class="nav-link" data-scroll-target="#examples-2">Examples</a>
  <ul class="collapse">
  <li><a href="#a-sequence-classification-model-using-model_for_generic_modaaseq_binaryclassification_lstm" id="toc-a-sequence-classification-model-using-model_for_generic_modaaseq_binaryclassification_lstm" class="nav-link" data-scroll-target="#a-sequence-classification-model-using-model_for_generic_modaaseq_binaryclassification_lstm">A sequence classification model using <code>Model_for_Generic_ModAASeq_BinaryClassification_LSTM</code></a></li>
  <li><a href="#a-sequence-classification-model-using-model_for_generic_modaaseq_binaryclassification_transformer" id="toc-a-sequence-classification-model-using-model_for_generic_modaaseq_binaryclassification_transformer" class="nav-link" data-scroll-target="#a-sequence-classification-model-using-model_for_generic_modaaseq_binaryclassification_transformer">A sequence classification model using <code>Model_for_Generic_ModAASeq_BinaryClassification_Transformer</code></a></li>
  </ul></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/MannLabs/alphapeptdeep/tree/main/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">Generic Property Prediction</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="description" class="level2">
<h2 class="anchored" data-anchor-id="description">Description</h2>
<section id="scope-of-the-generic-property-prediction" class="level3">
<h3 class="anchored" data-anchor-id="scope-of-the-generic-property-prediction">Scope of the generic property prediction</h3>
<p>One focus of the AlpahPeptDeep framework is on the prediction of retention times, MS2 Spectra and Collisional Cross sections of peptides. The Models to predict these properties are defined in the notebooks <code>rt.ipynb</code>, <code>ms2.ipynb</code> and <code>ccs.ipynb</code> respectively. In the following notebook, we define a more generalized model, which in principle allows to predict or classify arbitrary properties of peptides. The classes allow users to create new deep learning predictions of peptide properties, even without prior expertise in deep learning. ### How to build new models, even without experience in deep learning In order to predict or classify novel properties of peptides, the user simply needs to provide a list of peptides with a corresponding property (e.g.&nbsp;‘binding affinity’) for each peptide. The user only needs to initialize one the <code>ModelInterface</code> classes below and specify the name of the <code>target_column_to_train</code> and optionally the <code>target_column_to_predict</code>. The generic model classes will try to derive predictions or classifications of the property from the peptide sequence. Additional classes are also given, which complement the peptide sequence info with site-specific PTM infos.</p>
</section>
</section>
<section id="imports" class="level2">
<h2 class="anchored" data-anchor-id="imports">Imports</h2>
</section>
<section id="regression-models-for-predicting-a-scalar-value-for-a-given-amino-acid-sequence" class="level2">
<h2 class="anchored" data-anchor-id="regression-models-for-predicting-a-scalar-value-for-a-given-amino-acid-sequence">Regression models for predicting a scalar value for a given amino acid sequence</h2>
<hr>
<section id="modelinterface_for_generic_aaseq_regression" class="level3">
<h3 class="anchored" data-anchor-id="modelinterface_for_generic_aaseq_regression">ModelInterface_for_Generic_AASeq_Regression</h3>
<blockquote class="blockquote">
<pre><code> ModelInterface_for_Generic_AASeq_Regression (dropout=0.1,
                                              model_class:torch.nn.modules
                                              .module.Module=&lt;class '__mai
                                              n__.Model_for_Generic_AASeq_
                                              Regression_LSTM'&gt;,
                                              device:str='gpu', **kwargs)</code></pre>
</blockquote>
<p>Provides standardized methods to interact with ml models. Inherit into new class and override the abstract (i.e.&nbsp;not implemented) methods.</p>
<hr>
</section>
<section id="model_for_generic_aaseq_regression_transformer" class="level3">
<h3 class="anchored" data-anchor-id="model_for_generic_aaseq_regression_transformer">Model_for_Generic_AASeq_Regression_Transformer</h3>
<blockquote class="blockquote">
<pre><code> Model_for_Generic_AASeq_Regression_Transformer (hidden_dim=256,
                                                 nlayers=4,
                                                 output_attentions=False,
                                                 dropout=0.1, **kwargs)</code></pre>
</blockquote>
<p>Base class for all neural network modules.</p>
<p>Your models should also subclass this class.</p>
<p>Modules can also contain other Modules, allowing to nest them in a tree structure. You can assign the submodules as regular attributes::</p>
<pre><code>import torch.nn as nn
import torch.nn.functional as F

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.conv1 = nn.Conv2d(1, 20, 5)
        self.conv2 = nn.Conv2d(20, 20, 5)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        return F.relu(self.conv2(x))</code></pre>
<p>Submodules assigned in this way will be registered, and will have their parameters converted too when you call :meth:<code>to</code>, etc.</p>
<p>:ivar training: Boolean represents whether this module is in training or evaluation mode. :vartype training: bool</p>
<hr>
</section>
<section id="model_for_generic_aaseq_regression_lstm" class="level3">
<h3 class="anchored" data-anchor-id="model_for_generic_aaseq_regression_lstm">Model_for_Generic_AASeq_Regression_LSTM</h3>
<blockquote class="blockquote">
<pre><code> Model_for_Generic_AASeq_Regression_LSTM (hidden_dim=256, n_lstm_layers=4,
                                          dropout=0.1, **kwargs)</code></pre>
</blockquote>
<p>Base class for all neural network modules.</p>
<p>Your models should also subclass this class.</p>
<p>Modules can also contain other Modules, allowing to nest them in a tree structure. You can assign the submodules as regular attributes::</p>
<pre><code>import torch.nn as nn
import torch.nn.functional as F

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.conv1 = nn.Conv2d(1, 20, 5)
        self.conv2 = nn.Conv2d(20, 20, 5)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        return F.relu(self.conv2(x))</code></pre>
<p>Submodules assigned in this way will be registered, and will have their parameters converted too when you call :meth:<code>to</code>, etc.</p>
<p>:ivar training: Boolean represents whether this module is in training or evaluation mode. :vartype training: bool</p>
</section>
<section id="examples" class="level3">
<h3 class="anchored" data-anchor-id="examples">Examples</h3>
<section id="define-example-table" class="level4">
<h4 class="anchored" data-anchor-id="define-example-table">Define example Table</h4>
<div class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> create_example_input_dataframe_normalized_irt():</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    <span class="im">from</span> peptdeep.model.rt <span class="im">import</span> IRT_PEPTIDE_DF</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    IRT_PEPTIDE_DF<span class="op">=</span>IRT_PEPTIDE_DF.copy()</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    IRT_PEPTIDE_DF[<span class="st">'normalized_irt'</span>] <span class="op">=</span> (</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>        IRT_PEPTIDE_DF.irt<span class="op">-</span>IRT_PEPTIDE_DF.irt.<span class="bu">min</span>()</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    )<span class="op">/</span>(IRT_PEPTIDE_DF.irt.<span class="bu">max</span>()<span class="op">-</span>IRT_PEPTIDE_DF.irt.<span class="bu">min</span>())</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> IRT_PEPTIDE_DF</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>create_example_input_dataframe_normalized_irt()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>pep_name</th>
      <th>irt</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>normalized_irt</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LGGNEQVTR</td>
      <td>RT-pep a</td>
      <td>-24.92</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GAGSSEPVTGLDAK</td>
      <td>RT-pep b</td>
      <td>0.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>0.199488</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VEATFGVDESNAK</td>
      <td>RT-pep c</td>
      <td>12.39</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.298671</td>
    </tr>
    <tr>
      <th>3</th>
      <td>YILAGVENSK</td>
      <td>RT-pep d</td>
      <td>19.79</td>
      <td></td>
      <td></td>
      <td>10</td>
      <td>0.357909</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TPVISGGPYEYR</td>
      <td>RT-pep e</td>
      <td>28.71</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.429315</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TPVITGAPYEYR</td>
      <td>RT-pep f</td>
      <td>33.38</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.466699</td>
    </tr>
    <tr>
      <th>6</th>
      <td>DGLDAASYYAPVR</td>
      <td>RT-pep g</td>
      <td>42.26</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.537784</td>
    </tr>
    <tr>
      <th>7</th>
      <td>ADVTPADFSEWSK</td>
      <td>RT-pep h</td>
      <td>54.62</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.636728</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GTFIIDPGGVIR</td>
      <td>RT-pep i</td>
      <td>70.52</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.764009</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GTFIIDPAAVIR</td>
      <td>RT-pep k</td>
      <td>87.23</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.897775</td>
    </tr>
    <tr>
      <th>10</th>
      <td>LFLQFGAQGSPFLK</td>
      <td>RT-pep l</td>
      <td>100.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>1.000000</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
<section id="building-an-rt-model-based-on-model_for_generic_aaseq_regression_lstm" class="level4">
<h4 class="anchored" data-anchor-id="building-an-rt-model-based-on-model_for_generic_aaseq_regression_lstm">Building an RT model based on <code>Model_for_Generic_AASeq_Regression_LSTM</code></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>example_df <span class="op">=</span> create_example_input_dataframe_normalized_irt()</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>display(example_df)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">#initialize the modelinterface, specify which of the models to use</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> ModelInterface_for_Generic_AASeq_Regression(</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    model_class<span class="op">=</span>Model_for_Generic_AASeq_Regression_LSTM</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co"># specify the name of the column you want to use for traning</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>model.target_column_to_train <span class="op">=</span> <span class="st">'normalized_irt'</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>model.target_column_to_predict <span class="op">=</span> <span class="st">'predicted_normalized_irt'</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>model.train(example_df, epoch<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>model.predict(example_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>pep_name</th>
      <th>irt</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>normalized_irt</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LGGNEQVTR</td>
      <td>RT-pep a</td>
      <td>-24.92</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GAGSSEPVTGLDAK</td>
      <td>RT-pep b</td>
      <td>0.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>0.199488</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VEATFGVDESNAK</td>
      <td>RT-pep c</td>
      <td>12.39</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.298671</td>
    </tr>
    <tr>
      <th>3</th>
      <td>YILAGVENSK</td>
      <td>RT-pep d</td>
      <td>19.79</td>
      <td></td>
      <td></td>
      <td>10</td>
      <td>0.357909</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TPVISGGPYEYR</td>
      <td>RT-pep e</td>
      <td>28.71</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.429315</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TPVITGAPYEYR</td>
      <td>RT-pep f</td>
      <td>33.38</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.466699</td>
    </tr>
    <tr>
      <th>6</th>
      <td>DGLDAASYYAPVR</td>
      <td>RT-pep g</td>
      <td>42.26</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.537784</td>
    </tr>
    <tr>
      <th>7</th>
      <td>ADVTPADFSEWSK</td>
      <td>RT-pep h</td>
      <td>54.62</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.636728</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GTFIIDPGGVIR</td>
      <td>RT-pep i</td>
      <td>70.52</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.764009</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GTFIIDPAAVIR</td>
      <td>RT-pep k</td>
      <td>87.23</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.897775</td>
    </tr>
    <tr>
      <th>10</th>
      <td>LFLQFGAQGSPFLK</td>
      <td>RT-pep l</td>
      <td>100.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>1.000000</td>
    </tr>
  </tbody>
</table>
</div>
</div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>pep_name</th>
      <th>irt</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>normalized_irt</th>
      <th>predicted_normalized_irt</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LGGNEQVTR</td>
      <td>RT-pep a</td>
      <td>-24.92</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>0.000000</td>
      <td>0.007052</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GAGSSEPVTGLDAK</td>
      <td>RT-pep b</td>
      <td>0.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>0.199488</td>
      <td>0.237059</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VEATFGVDESNAK</td>
      <td>RT-pep c</td>
      <td>12.39</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.298671</td>
      <td>0.321053</td>
    </tr>
    <tr>
      <th>3</th>
      <td>YILAGVENSK</td>
      <td>RT-pep d</td>
      <td>19.79</td>
      <td></td>
      <td></td>
      <td>10</td>
      <td>0.357909</td>
      <td>0.371977</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TPVISGGPYEYR</td>
      <td>RT-pep e</td>
      <td>28.71</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.429315</td>
      <td>0.423640</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TPVITGAPYEYR</td>
      <td>RT-pep f</td>
      <td>33.38</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.466699</td>
      <td>0.479110</td>
    </tr>
    <tr>
      <th>6</th>
      <td>DGLDAASYYAPVR</td>
      <td>RT-pep g</td>
      <td>42.26</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.537784</td>
      <td>0.711614</td>
    </tr>
    <tr>
      <th>7</th>
      <td>ADVTPADFSEWSK</td>
      <td>RT-pep h</td>
      <td>54.62</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.636728</td>
      <td>0.705609</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GTFIIDPGGVIR</td>
      <td>RT-pep i</td>
      <td>70.52</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.764009</td>
      <td>0.648197</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GTFIIDPAAVIR</td>
      <td>RT-pep k</td>
      <td>87.23</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.897775</td>
      <td>0.833484</td>
    </tr>
    <tr>
      <th>10</th>
      <td>LFLQFGAQGSPFLK</td>
      <td>RT-pep l</td>
      <td>100.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>1.000000</td>
      <td>0.978105</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
<section id="building-an-rt-model-for-only-sequences-based-on-model_for_generic_aaseq_regression_transformer" class="level4">
<h4 class="anchored" data-anchor-id="building-an-rt-model-for-only-sequences-based-on-model_for_generic_aaseq_regression_transformer">Building an RT model for only sequences based on <code>Model_for_Generic_AASeq_Regression_Transformer</code></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>example_df <span class="op">=</span> create_example_input_dataframe_normalized_irt()</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>display(example_df)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co">#initialize the modelinterface, specify which of the models to use</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> ModelInterface_for_Generic_AASeq_Regression(</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    model_class<span class="op">=</span>Model_for_Generic_AASeq_Regression_Transformer</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># specify the name of the column you want to use for traning</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>model.target_column_to_train <span class="op">=</span> <span class="st">'normalized_irt'</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>model.target_column_to_predict <span class="op">=</span> <span class="st">'predicted_normalized_irt'</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>model.train(example_df, epoch<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>model.predict(example_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>pep_name</th>
      <th>irt</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>normalized_irt</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LGGNEQVTR</td>
      <td>RT-pep a</td>
      <td>-24.92</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GAGSSEPVTGLDAK</td>
      <td>RT-pep b</td>
      <td>0.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>0.199488</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VEATFGVDESNAK</td>
      <td>RT-pep c</td>
      <td>12.39</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.298671</td>
    </tr>
    <tr>
      <th>3</th>
      <td>YILAGVENSK</td>
      <td>RT-pep d</td>
      <td>19.79</td>
      <td></td>
      <td></td>
      <td>10</td>
      <td>0.357909</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TPVISGGPYEYR</td>
      <td>RT-pep e</td>
      <td>28.71</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.429315</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TPVITGAPYEYR</td>
      <td>RT-pep f</td>
      <td>33.38</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.466699</td>
    </tr>
    <tr>
      <th>6</th>
      <td>DGLDAASYYAPVR</td>
      <td>RT-pep g</td>
      <td>42.26</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.537784</td>
    </tr>
    <tr>
      <th>7</th>
      <td>ADVTPADFSEWSK</td>
      <td>RT-pep h</td>
      <td>54.62</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.636728</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GTFIIDPGGVIR</td>
      <td>RT-pep i</td>
      <td>70.52</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.764009</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GTFIIDPAAVIR</td>
      <td>RT-pep k</td>
      <td>87.23</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.897775</td>
    </tr>
    <tr>
      <th>10</th>
      <td>LFLQFGAQGSPFLK</td>
      <td>RT-pep l</td>
      <td>100.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>1.000000</td>
    </tr>
  </tbody>
</table>
</div>
</div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>pep_name</th>
      <th>irt</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>normalized_irt</th>
      <th>predicted_normalized_irt</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LGGNEQVTR</td>
      <td>RT-pep a</td>
      <td>-24.92</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>0.000000</td>
      <td>0.098438</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GAGSSEPVTGLDAK</td>
      <td>RT-pep b</td>
      <td>0.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>0.199488</td>
      <td>0.310149</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VEATFGVDESNAK</td>
      <td>RT-pep c</td>
      <td>12.39</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.298671</td>
      <td>0.330474</td>
    </tr>
    <tr>
      <th>3</th>
      <td>YILAGVENSK</td>
      <td>RT-pep d</td>
      <td>19.79</td>
      <td></td>
      <td></td>
      <td>10</td>
      <td>0.357909</td>
      <td>0.464114</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TPVISGGPYEYR</td>
      <td>RT-pep e</td>
      <td>28.71</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.429315</td>
      <td>0.579034</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TPVITGAPYEYR</td>
      <td>RT-pep f</td>
      <td>33.38</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.466699</td>
      <td>0.531123</td>
    </tr>
    <tr>
      <th>6</th>
      <td>DGLDAASYYAPVR</td>
      <td>RT-pep g</td>
      <td>42.26</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.537784</td>
      <td>0.774658</td>
    </tr>
    <tr>
      <th>7</th>
      <td>ADVTPADFSEWSK</td>
      <td>RT-pep h</td>
      <td>54.62</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.636728</td>
      <td>0.752060</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GTFIIDPGGVIR</td>
      <td>RT-pep i</td>
      <td>70.52</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.764009</td>
      <td>0.837158</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GTFIIDPAAVIR</td>
      <td>RT-pep k</td>
      <td>87.23</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.897775</td>
      <td>0.967516</td>
    </tr>
    <tr>
      <th>10</th>
      <td>LFLQFGAQGSPFLK</td>
      <td>RT-pep l</td>
      <td>100.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>1.000000</td>
      <td>1.165970</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="regression-models-for-predicting-a-scalar-value-for-a-given-amino-acid-sequence-and-site-specific-ptms" class="level2">
<h2 class="anchored" data-anchor-id="regression-models-for-predicting-a-scalar-value-for-a-given-amino-acid-sequence-and-site-specific-ptms">Regression models for predicting a scalar value for a given amino acid sequence and site-specific PTMs</h2>
<hr>
<section id="modelinterface_for_generic_modaaseq_regression" class="level3">
<h3 class="anchored" data-anchor-id="modelinterface_for_generic_modaaseq_regression">ModelInterface_for_Generic_ModAASeq_Regression</h3>
<blockquote class="blockquote">
<pre><code> ModelInterface_for_Generic_ModAASeq_Regression (dropout=0.1,
                                                 model_class:torch.nn.modu
                                                 les.module.Module=&lt;class 
                                                 '__main__.Model_for_Gener
                                                 ic_ModAASeq_Regression_LS
                                                 TM'&gt;, device:str='gpu',
                                                 **kwargs)</code></pre>
</blockquote>
<p>Provides standardized methods to interact with ml models. Inherit into new class and override the abstract (i.e.&nbsp;not implemented) methods.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>dropout</td>
<td>float</td>
<td>0.1</td>
<td></td>
</tr>
<tr class="even">
<td>model_class</td>
<td>Module</td>
<td>Model_for_Generic_ModAASeq_Regression_LSTM</td>
<td>model defined above</td>
</tr>
<tr class="odd">
<td>device</td>
<td>str</td>
<td>gpu</td>
<td></td>
</tr>
<tr class="even">
<td>kwargs</td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="model_for_generic_modaaseq_regression_transformer" class="level3">
<h3 class="anchored" data-anchor-id="model_for_generic_modaaseq_regression_transformer">Model_for_Generic_ModAASeq_Regression_Transformer</h3>
<blockquote class="blockquote">
<pre><code> Model_for_Generic_ModAASeq_Regression_Transformer (hidden_dim=256,
                                                    nlayers=4, output_atte
                                                    ntions=False,
                                                    dropout=0.1, **kwargs)</code></pre>
</blockquote>
<p>Base class for all neural network modules.</p>
<p>Your models should also subclass this class.</p>
<p>Modules can also contain other Modules, allowing to nest them in a tree structure. You can assign the submodules as regular attributes::</p>
<pre><code>import torch.nn as nn
import torch.nn.functional as F

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.conv1 = nn.Conv2d(1, 20, 5)
        self.conv2 = nn.Conv2d(20, 20, 5)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        return F.relu(self.conv2(x))</code></pre>
<p>Submodules assigned in this way will be registered, and will have their parameters converted too when you call :meth:<code>to</code>, etc.</p>
<p>:ivar training: Boolean represents whether this module is in training or evaluation mode. :vartype training: bool</p>
<hr>
</section>
<section id="model_for_generic_modaaseq_regression_lstm" class="level3">
<h3 class="anchored" data-anchor-id="model_for_generic_modaaseq_regression_lstm">Model_for_Generic_ModAASeq_Regression_LSTM</h3>
<blockquote class="blockquote">
<pre><code> Model_for_Generic_ModAASeq_Regression_LSTM (hidden_dim=256,
                                             n_lstm_layers=4, dropout=0.1,
                                             **kwargs)</code></pre>
</blockquote>
<p>Base class for all neural network modules.</p>
<p>Your models should also subclass this class.</p>
<p>Modules can also contain other Modules, allowing to nest them in a tree structure. You can assign the submodules as regular attributes::</p>
<pre><code>import torch.nn as nn
import torch.nn.functional as F

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.conv1 = nn.Conv2d(1, 20, 5)
        self.conv2 = nn.Conv2d(20, 20, 5)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        return F.relu(self.conv2(x))</code></pre>
<p>Submodules assigned in this way will be registered, and will have their parameters converted too when you call :meth:<code>to</code>, etc.</p>
<p>:ivar training: Boolean represents whether this module is in training or evaluation mode. :vartype training: bool</p>
<section id="scalar-regression-model-rt-with-modified-aa-sequences-using-model_for_generic_modaaseq_regression_lstm" class="level4">
<h4 class="anchored" data-anchor-id="scalar-regression-model-rt-with-modified-aa-sequences-using-model_for_generic_modaaseq_regression_lstm">Scalar regression model (RT) with modified AA sequences using <code>Model_for_Generic_ModAASeq_Regression_LSTM</code></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>example_df <span class="op">=</span> create_example_input_dataframe_normalized_irt()</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="co">#initialize the modelinterface, specify which of the models to use</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> ModelInterface_for_Generic_ModAASeq_Regression(</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    model_class<span class="op">=</span>Model_for_Generic_ModAASeq_Regression_LSTM</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="co"># specify the name of the column you want to use for traning</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>model.target_column_to_train <span class="op">=</span> <span class="st">'normalized_irt'</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>model.target_column_to_predict <span class="op">=</span> <span class="st">'predicted_normalized_irt'</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>model.train(example_df, epoch<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>model.predict(example_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>pep_name</th>
      <th>irt</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>normalized_irt</th>
      <th>predicted_normalized_irt</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LGGNEQVTR</td>
      <td>RT-pep a</td>
      <td>-24.92</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>0.000000</td>
      <td>0.008056</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GAGSSEPVTGLDAK</td>
      <td>RT-pep b</td>
      <td>0.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>0.199488</td>
      <td>0.198335</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VEATFGVDESNAK</td>
      <td>RT-pep c</td>
      <td>12.39</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.298671</td>
      <td>0.304394</td>
    </tr>
    <tr>
      <th>3</th>
      <td>YILAGVENSK</td>
      <td>RT-pep d</td>
      <td>19.79</td>
      <td></td>
      <td></td>
      <td>10</td>
      <td>0.357909</td>
      <td>0.358489</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TPVISGGPYEYR</td>
      <td>RT-pep e</td>
      <td>28.71</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.429315</td>
      <td>0.427770</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TPVITGAPYEYR</td>
      <td>RT-pep f</td>
      <td>33.38</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.466699</td>
      <td>0.461542</td>
    </tr>
    <tr>
      <th>6</th>
      <td>DGLDAASYYAPVR</td>
      <td>RT-pep g</td>
      <td>42.26</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.537784</td>
      <td>0.518003</td>
    </tr>
    <tr>
      <th>7</th>
      <td>ADVTPADFSEWSK</td>
      <td>RT-pep h</td>
      <td>54.62</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.636728</td>
      <td>0.608539</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GTFIIDPGGVIR</td>
      <td>RT-pep i</td>
      <td>70.52</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.764009</td>
      <td>0.739531</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GTFIIDPAAVIR</td>
      <td>RT-pep k</td>
      <td>87.23</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.897775</td>
      <td>0.836452</td>
    </tr>
    <tr>
      <th>10</th>
      <td>LFLQFGAQGSPFLK</td>
      <td>RT-pep l</td>
      <td>100.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>1.000000</td>
      <td>0.946470</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
<section id="scalar-regression-model-rt-with-modified-aa-sequences-using-model_for_generic_modaaseq_regression_transformer" class="level4">
<h4 class="anchored" data-anchor-id="scalar-regression-model-rt-with-modified-aa-sequences-using-model_for_generic_modaaseq_regression_transformer">Scalar regression model (RT) with modified AA sequences using <code>Model_for_Generic_ModAASeq_Regression_Transformer</code></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>example_df <span class="op">=</span> create_example_input_dataframe_normalized_irt()</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co">#initialize the modelinterface, specify which of the models to use</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> ModelInterface_for_Generic_ModAASeq_Regression(</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>    model_class<span class="op">=</span>Model_for_Generic_ModAASeq_Regression_Transformer</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co"># specify the name of the column you want to use for traning</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>model.target_column_to_train <span class="op">=</span> <span class="st">'normalized_irt'</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>model.target_column_to_predict <span class="op">=</span> <span class="st">'predicted_normalized_irt'</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>model.train(example_df, epoch<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>model.predict(example_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>pep_name</th>
      <th>irt</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>normalized_irt</th>
      <th>predicted_normalized_irt</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LGGNEQVTR</td>
      <td>RT-pep a</td>
      <td>-24.92</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GAGSSEPVTGLDAK</td>
      <td>RT-pep b</td>
      <td>0.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>0.199488</td>
      <td>0.099054</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VEATFGVDESNAK</td>
      <td>RT-pep c</td>
      <td>12.39</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.298671</td>
      <td>0.231878</td>
    </tr>
    <tr>
      <th>3</th>
      <td>YILAGVENSK</td>
      <td>RT-pep d</td>
      <td>19.79</td>
      <td></td>
      <td></td>
      <td>10</td>
      <td>0.357909</td>
      <td>0.407409</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TPVISGGPYEYR</td>
      <td>RT-pep e</td>
      <td>28.71</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.429315</td>
      <td>0.294037</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TPVITGAPYEYR</td>
      <td>RT-pep f</td>
      <td>33.38</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.466699</td>
      <td>0.343179</td>
    </tr>
    <tr>
      <th>6</th>
      <td>DGLDAASYYAPVR</td>
      <td>RT-pep g</td>
      <td>42.26</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.537784</td>
      <td>0.548849</td>
    </tr>
    <tr>
      <th>7</th>
      <td>ADVTPADFSEWSK</td>
      <td>RT-pep h</td>
      <td>54.62</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.636728</td>
      <td>0.543506</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GTFIIDPGGVIR</td>
      <td>RT-pep i</td>
      <td>70.52</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.764009</td>
      <td>0.700900</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GTFIIDPAAVIR</td>
      <td>RT-pep k</td>
      <td>87.23</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.897775</td>
      <td>0.875348</td>
    </tr>
    <tr>
      <th>10</th>
      <td>LFLQFGAQGSPFLK</td>
      <td>RT-pep l</td>
      <td>100.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>1.000000</td>
      <td>1.009214</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="binary-classification-models-for-a-given-amino-acid-sequence" class="level2">
<h2 class="anchored" data-anchor-id="binary-classification-models-for-a-given-amino-acid-sequence">Binary classification models for a given amino acid sequence</h2>
<hr>
<section id="modelinterface_for_generic_aaseq_binaryclassification" class="level3">
<h3 class="anchored" data-anchor-id="modelinterface_for_generic_aaseq_binaryclassification">ModelInterface_for_Generic_AASeq_BinaryClassification</h3>
<blockquote class="blockquote">
<pre><code> ModelInterface_for_Generic_AASeq_BinaryClassification (dropout=0.1,
                                                        model_class:torch.
                                                        nn.modules.module.
                                                        Module=&lt;class '__m
                                                        ain__.Model_for_Ge
                                                        neric_AASeq_Binary
                                                        Classification_LST
                                                        M'&gt;,
                                                        device:str='gpu',
                                                        **kwargs)</code></pre>
</blockquote>
<p>Provides standardized methods to interact with ml models. Inherit into new class and override the abstract (i.e.&nbsp;not implemented) methods.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>dropout</td>
<td>float</td>
<td>0.1</td>
<td></td>
</tr>
<tr class="even">
<td>model_class</td>
<td>Module</td>
<td>Model_for_Generic_AASeq_BinaryClassification_LSTM</td>
<td>model defined above</td>
</tr>
<tr class="odd">
<td>device</td>
<td>str</td>
<td>gpu</td>
<td></td>
</tr>
<tr class="even">
<td>kwargs</td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="model_for_generic_aaseq_binaryclassification_transformer" class="level3">
<h3 class="anchored" data-anchor-id="model_for_generic_aaseq_binaryclassification_transformer">Model_for_Generic_AASeq_BinaryClassification_Transformer</h3>
<blockquote class="blockquote">
<pre><code> Model_for_Generic_AASeq_BinaryClassification_Transformer (hidden_dim=256,
                                                           nlayers=4, outp
                                                           ut_attentions=F
                                                           alse,
                                                           dropout=0.1,
                                                           **kwargs)</code></pre>
</blockquote>
<p>Base class for all neural network modules.</p>
<p>Your models should also subclass this class.</p>
<p>Modules can also contain other Modules, allowing to nest them in a tree structure. You can assign the submodules as regular attributes::</p>
<pre><code>import torch.nn as nn
import torch.nn.functional as F

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.conv1 = nn.Conv2d(1, 20, 5)
        self.conv2 = nn.Conv2d(20, 20, 5)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        return F.relu(self.conv2(x))</code></pre>
<p>Submodules assigned in this way will be registered, and will have their parameters converted too when you call :meth:<code>to</code>, etc.</p>
<p>:ivar training: Boolean represents whether this module is in training or evaluation mode. :vartype training: bool</p>
<hr>
</section>
<section id="model_for_generic_aaseq_binaryclassification_lstm" class="level3">
<h3 class="anchored" data-anchor-id="model_for_generic_aaseq_binaryclassification_lstm">Model_for_Generic_AASeq_BinaryClassification_LSTM</h3>
<blockquote class="blockquote">
<pre><code> Model_for_Generic_AASeq_BinaryClassification_LSTM (hidden_dim=256,
                                                    n_lstm_layers=4,
                                                    dropout=0.1, **kwargs)</code></pre>
</blockquote>
<p>Base class for all neural network modules.</p>
<p>Your models should also subclass this class.</p>
<p>Modules can also contain other Modules, allowing to nest them in a tree structure. You can assign the submodules as regular attributes::</p>
<pre><code>import torch.nn as nn
import torch.nn.functional as F

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.conv1 = nn.Conv2d(1, 20, 5)
        self.conv2 = nn.Conv2d(20, 20, 5)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        return F.relu(self.conv2(x))</code></pre>
<p>Submodules assigned in this way will be registered, and will have their parameters converted too when you call :meth:<code>to</code>, etc.</p>
<p>:ivar training: Boolean represents whether this module is in training or evaluation mode. :vartype training: bool</p>
</section>
<section id="examples-1" class="level3">
<h3 class="anchored" data-anchor-id="examples-1">Examples</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> create_example_input_dataframe_classification_rt():</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>    rt_df <span class="op">=</span> create_example_input_dataframe_normalized_irt()</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>    rt_df[<span class="st">'is_in_first_half_of_column'</span>] <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>    rt_df.loc[:<span class="dv">5</span>,<span class="st">'is_in_first_half_of_column'</span>]<span class="op">=</span><span class="dv">1</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> rt_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="a-sequence-classification-model-using-model_for_generic_aaseq_binaryclassification_lstm" class="level4">
<h4 class="anchored" data-anchor-id="a-sequence-classification-model-using-model_for_generic_aaseq_binaryclassification_lstm">A sequence classification model using <code>Model_for_Generic_AASeq_BinaryClassification_LSTM</code></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>example_df <span class="op">=</span> create_example_input_dataframe_classification_rt()</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co">#initialize the modelinterface, specify which of the models to use</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> ModelInterface_for_Generic_AASeq_BinaryClassification(</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>    model_class<span class="op">=</span>Model_for_Generic_AASeq_BinaryClassification_LSTM</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="co"># specify the name of the column you want to use for traning</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>model.target_column_to_train <span class="op">=</span> <span class="st">'is_in_first_half_of_column'</span> </span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>model.target_column_to_predict <span class="op">=</span> <span class="st">'predicted_will_be_in_first_half_of_column'</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>model.train(example_df, epoch<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>model.predict(example_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>pep_name</th>
      <th>irt</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>normalized_irt</th>
      <th>is_in_first_half_of_column</th>
      <th>predicted_will_be_in_first_half_of_column</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LGGNEQVTR</td>
      <td>RT-pep a</td>
      <td>-24.92</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>0.000000</td>
      <td>1</td>
      <td>0.993174</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GAGSSEPVTGLDAK</td>
      <td>RT-pep b</td>
      <td>0.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>0.199488</td>
      <td>1</td>
      <td>0.989214</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VEATFGVDESNAK</td>
      <td>RT-pep c</td>
      <td>12.39</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.298671</td>
      <td>1</td>
      <td>0.993532</td>
    </tr>
    <tr>
      <th>3</th>
      <td>YILAGVENSK</td>
      <td>RT-pep d</td>
      <td>19.79</td>
      <td></td>
      <td></td>
      <td>10</td>
      <td>0.357909</td>
      <td>1</td>
      <td>0.992786</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TPVISGGPYEYR</td>
      <td>RT-pep e</td>
      <td>28.71</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.429315</td>
      <td>1</td>
      <td>0.992660</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TPVITGAPYEYR</td>
      <td>RT-pep f</td>
      <td>33.38</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.466699</td>
      <td>1</td>
      <td>0.991030</td>
    </tr>
    <tr>
      <th>6</th>
      <td>DGLDAASYYAPVR</td>
      <td>RT-pep g</td>
      <td>42.26</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.537784</td>
      <td>0</td>
      <td>0.308686</td>
    </tr>
    <tr>
      <th>7</th>
      <td>ADVTPADFSEWSK</td>
      <td>RT-pep h</td>
      <td>54.62</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.636728</td>
      <td>0</td>
      <td>0.317267</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GTFIIDPGGVIR</td>
      <td>RT-pep i</td>
      <td>70.52</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.764009</td>
      <td>0</td>
      <td>0.323006</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GTFIIDPAAVIR</td>
      <td>RT-pep k</td>
      <td>87.23</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.897775</td>
      <td>0</td>
      <td>0.316616</td>
    </tr>
    <tr>
      <th>10</th>
      <td>LFLQFGAQGSPFLK</td>
      <td>RT-pep l</td>
      <td>100.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>1.000000</td>
      <td>0</td>
      <td>0.310942</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
<section id="a-sequence-classification-model-using-model_for_generic_aaseq_binaryclassification_transformer" class="level4">
<h4 class="anchored" data-anchor-id="a-sequence-classification-model-using-model_for_generic_aaseq_binaryclassification_transformer">A sequence classification model using <code>Model_for_Generic_AASeq_BinaryClassification_Transformer</code></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>example_df <span class="op">=</span> create_example_input_dataframe_classification_rt()</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co">#initialize the modelinterface, specify which of the models to use</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> ModelInterface_for_Generic_AASeq_BinaryClassification(</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    model_class<span class="op">=</span>Model_for_Generic_AASeq_BinaryClassification_Transformer</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="co"># specify the name of the column you want to use for traning</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>model.target_column_to_train <span class="op">=</span> <span class="st">'is_in_first_half_of_column'</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>model.target_column_to_predict <span class="op">=</span> <span class="st">'predicted_will_be_in_first_half_of_column'</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>model.train(example_df, epoch<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>model.predict(example_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>pep_name</th>
      <th>irt</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>normalized_irt</th>
      <th>is_in_first_half_of_column</th>
      <th>predicted_will_be_in_first_half_of_column</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LGGNEQVTR</td>
      <td>RT-pep a</td>
      <td>-24.92</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>0.000000</td>
      <td>1</td>
      <td>0.997722</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GAGSSEPVTGLDAK</td>
      <td>RT-pep b</td>
      <td>0.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>0.199488</td>
      <td>1</td>
      <td>0.997216</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VEATFGVDESNAK</td>
      <td>RT-pep c</td>
      <td>12.39</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.298671</td>
      <td>1</td>
      <td>0.996552</td>
    </tr>
    <tr>
      <th>3</th>
      <td>YILAGVENSK</td>
      <td>RT-pep d</td>
      <td>19.79</td>
      <td></td>
      <td></td>
      <td>10</td>
      <td>0.357909</td>
      <td>1</td>
      <td>0.997443</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TPVISGGPYEYR</td>
      <td>RT-pep e</td>
      <td>28.71</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.429315</td>
      <td>1</td>
      <td>0.996104</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TPVITGAPYEYR</td>
      <td>RT-pep f</td>
      <td>33.38</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.466699</td>
      <td>1</td>
      <td>0.995777</td>
    </tr>
    <tr>
      <th>6</th>
      <td>DGLDAASYYAPVR</td>
      <td>RT-pep g</td>
      <td>42.26</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.537784</td>
      <td>0</td>
      <td>0.003511</td>
    </tr>
    <tr>
      <th>7</th>
      <td>ADVTPADFSEWSK</td>
      <td>RT-pep h</td>
      <td>54.62</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.636728</td>
      <td>0</td>
      <td>0.004135</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GTFIIDPGGVIR</td>
      <td>RT-pep i</td>
      <td>70.52</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.764009</td>
      <td>0</td>
      <td>0.004020</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GTFIIDPAAVIR</td>
      <td>RT-pep k</td>
      <td>87.23</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.897775</td>
      <td>0</td>
      <td>0.003701</td>
    </tr>
    <tr>
      <th>10</th>
      <td>LFLQFGAQGSPFLK</td>
      <td>RT-pep l</td>
      <td>100.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>1.000000</td>
      <td>0</td>
      <td>0.003674</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="binary-classification-models-for-given-amino-acid-sequence-and-site-specific-ptms" class="level2">
<h2 class="anchored" data-anchor-id="binary-classification-models-for-given-amino-acid-sequence-and-site-specific-ptms">Binary classification models for given amino acid sequence and site-specific PTMs</h2>
<hr>
<section id="modelinterface_for_generic_modaaseq_binaryclassification" class="level3">
<h3 class="anchored" data-anchor-id="modelinterface_for_generic_modaaseq_binaryclassification">ModelInterface_for_Generic_ModAASeq_BinaryClassification</h3>
<blockquote class="blockquote">
<pre><code> ModelInterface_for_Generic_ModAASeq_BinaryClassification (dropout=0.1,
                                                           model_class:tor
                                                           ch.nn.modules.m
                                                           odule.Module=&lt;c
                                                           lass '__main__.
                                                           Model_for_Gener
                                                           ic_ModAASeq_Bin
                                                           aryClassificati
                                                           on_LSTM'&gt;, devi
                                                           ce:str='gpu',
                                                           **kwargs)</code></pre>
</blockquote>
<p>Provides standardized methods to interact with ml models. Inherit into new class and override the abstract (i.e.&nbsp;not implemented) methods.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>dropout</td>
<td>float</td>
<td>0.1</td>
<td></td>
</tr>
<tr class="even">
<td>model_class</td>
<td>Module</td>
<td>Model_for_Generic_ModAASeq_BinaryClassification_LSTM</td>
<td>model defined above</td>
</tr>
<tr class="odd">
<td>device</td>
<td>str</td>
<td>gpu</td>
<td></td>
</tr>
<tr class="even">
<td>kwargs</td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="model_for_generic_modaaseq_binaryclassification_transformer" class="level3">
<h3 class="anchored" data-anchor-id="model_for_generic_modaaseq_binaryclassification_transformer">Model_for_Generic_ModAASeq_BinaryClassification_Transformer</h3>
<blockquote class="blockquote">
<pre><code> Model_for_Generic_ModAASeq_BinaryClassification_Transformer
                                                              (hidden_dim=
                                                              256,
                                                              nlayers=4, o
                                                              utput_attent
                                                              ions=False,
                                                              dropout=0.1,
                                                              **kwargs)</code></pre>
</blockquote>
<p>Base class for all neural network modules.</p>
<p>Your models should also subclass this class.</p>
<p>Modules can also contain other Modules, allowing to nest them in a tree structure. You can assign the submodules as regular attributes::</p>
<pre><code>import torch.nn as nn
import torch.nn.functional as F

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.conv1 = nn.Conv2d(1, 20, 5)
        self.conv2 = nn.Conv2d(20, 20, 5)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        return F.relu(self.conv2(x))</code></pre>
<p>Submodules assigned in this way will be registered, and will have their parameters converted too when you call :meth:<code>to</code>, etc.</p>
<p>:ivar training: Boolean represents whether this module is in training or evaluation mode. :vartype training: bool</p>
<hr>
</section>
<section id="model_for_generic_modaaseq_binaryclassification_lstm" class="level3">
<h3 class="anchored" data-anchor-id="model_for_generic_modaaseq_binaryclassification_lstm">Model_for_Generic_ModAASeq_BinaryClassification_LSTM</h3>
<blockquote class="blockquote">
<pre><code> Model_for_Generic_ModAASeq_BinaryClassification_LSTM (hidden_dim=256,
                                                       n_lstm_layers=4,
                                                       dropout=0.1,
                                                       **kwargs)</code></pre>
</blockquote>
<p>Base class for all neural network modules.</p>
<p>Your models should also subclass this class.</p>
<p>Modules can also contain other Modules, allowing to nest them in a tree structure. You can assign the submodules as regular attributes::</p>
<pre><code>import torch.nn as nn
import torch.nn.functional as F

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.conv1 = nn.Conv2d(1, 20, 5)
        self.conv2 = nn.Conv2d(20, 20, 5)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        return F.relu(self.conv2(x))</code></pre>
<p>Submodules assigned in this way will be registered, and will have their parameters converted too when you call :meth:<code>to</code>, etc.</p>
<p>:ivar training: Boolean represents whether this module is in training or evaluation mode. :vartype training: bool</p>
</section>
<section id="examples-2" class="level3">
<h3 class="anchored" data-anchor-id="examples-2">Examples</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> create_example_input_dataframe_classification_rt():</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>    rt_df <span class="op">=</span> create_example_input_dataframe_normalized_irt()</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>    rt_df[<span class="st">'is_in_first_half_of_column'</span>] <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>    rt_df.loc[:<span class="dv">5</span>,<span class="st">'is_in_first_half_of_column'</span>]<span class="op">=</span><span class="dv">1</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> rt_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="a-sequence-classification-model-using-model_for_generic_modaaseq_binaryclassification_lstm" class="level4">
<h4 class="anchored" data-anchor-id="a-sequence-classification-model-using-model_for_generic_modaaseq_binaryclassification_lstm">A sequence classification model using <code>Model_for_Generic_ModAASeq_BinaryClassification_LSTM</code></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>example_df <span class="op">=</span> create_example_input_dataframe_classification_rt()</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="co">#initialize the modelinterface, specify which of the models to use</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> ModelInterface_for_Generic_ModAASeq_BinaryClassification(</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    model_class<span class="op">=</span>Model_for_Generic_ModAASeq_BinaryClassification_LSTM</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="co"># specify the name of the column you want to use for traning</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>model.target_column_to_train <span class="op">=</span> <span class="st">'is_in_first_half_of_column'</span> </span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>model.target_column_to_predict <span class="op">=</span> <span class="st">'predicted_will_be_in_first_half_of_column'</span></span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>model.train(example_df, epoch<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>model.predict(example_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>pep_name</th>
      <th>irt</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>normalized_irt</th>
      <th>is_in_first_half_of_column</th>
      <th>predicted_will_be_in_first_half_of_column</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LGGNEQVTR</td>
      <td>RT-pep a</td>
      <td>-24.92</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>0.000000</td>
      <td>1</td>
      <td>0.996521</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GAGSSEPVTGLDAK</td>
      <td>RT-pep b</td>
      <td>0.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>0.199488</td>
      <td>1</td>
      <td>0.996062</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VEATFGVDESNAK</td>
      <td>RT-pep c</td>
      <td>12.39</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.298671</td>
      <td>1</td>
      <td>0.996103</td>
    </tr>
    <tr>
      <th>3</th>
      <td>YILAGVENSK</td>
      <td>RT-pep d</td>
      <td>19.79</td>
      <td></td>
      <td></td>
      <td>10</td>
      <td>0.357909</td>
      <td>1</td>
      <td>0.996884</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TPVISGGPYEYR</td>
      <td>RT-pep e</td>
      <td>28.71</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.429315</td>
      <td>1</td>
      <td>0.996284</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TPVITGAPYEYR</td>
      <td>RT-pep f</td>
      <td>33.38</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.466699</td>
      <td>1</td>
      <td>0.993848</td>
    </tr>
    <tr>
      <th>6</th>
      <td>DGLDAASYYAPVR</td>
      <td>RT-pep g</td>
      <td>42.26</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.537784</td>
      <td>0</td>
      <td>0.406263</td>
    </tr>
    <tr>
      <th>7</th>
      <td>ADVTPADFSEWSK</td>
      <td>RT-pep h</td>
      <td>54.62</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.636728</td>
      <td>0</td>
      <td>0.407196</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GTFIIDPGGVIR</td>
      <td>RT-pep i</td>
      <td>70.52</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.764009</td>
      <td>0</td>
      <td>0.402498</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GTFIIDPAAVIR</td>
      <td>RT-pep k</td>
      <td>87.23</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.897775</td>
      <td>0</td>
      <td>0.397870</td>
    </tr>
    <tr>
      <th>10</th>
      <td>LFLQFGAQGSPFLK</td>
      <td>RT-pep l</td>
      <td>100.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>1.000000</td>
      <td>0</td>
      <td>0.397521</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
<section id="a-sequence-classification-model-using-model_for_generic_modaaseq_binaryclassification_transformer" class="level4">
<h4 class="anchored" data-anchor-id="a-sequence-classification-model-using-model_for_generic_modaaseq_binaryclassification_transformer">A sequence classification model using <code>Model_for_Generic_ModAASeq_BinaryClassification_Transformer</code></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>example_df <span class="op">=</span> create_example_input_dataframe_classification_rt()</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="co">#initialize the modelinterface, specify which of the models to use</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> ModelInterface_for_Generic_ModAASeq_BinaryClassification(</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>    model_class<span class="op">=</span>Model_for_Generic_ModAASeq_BinaryClassification_Transformer</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="co"># specify the name of the column you want to use for traning</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>model.target_column_to_train <span class="op">=</span> <span class="st">'is_in_first_half_of_column'</span> </span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>model.target_column_to_predict <span class="op">=</span> <span class="st">'predicted_will_be_in_first_half_of_column'</span></span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>model.train(example_df, epoch<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>model.predict(example_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>pep_name</th>
      <th>irt</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>normalized_irt</th>
      <th>is_in_first_half_of_column</th>
      <th>predicted_will_be_in_first_half_of_column</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LGGNEQVTR</td>
      <td>RT-pep a</td>
      <td>-24.92</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>0.000000</td>
      <td>1</td>
      <td>0.996638</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GAGSSEPVTGLDAK</td>
      <td>RT-pep b</td>
      <td>0.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>0.199488</td>
      <td>1</td>
      <td>0.995580</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VEATFGVDESNAK</td>
      <td>RT-pep c</td>
      <td>12.39</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.298671</td>
      <td>1</td>
      <td>0.995562</td>
    </tr>
    <tr>
      <th>3</th>
      <td>YILAGVENSK</td>
      <td>RT-pep d</td>
      <td>19.79</td>
      <td></td>
      <td></td>
      <td>10</td>
      <td>0.357909</td>
      <td>1</td>
      <td>0.996606</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TPVISGGPYEYR</td>
      <td>RT-pep e</td>
      <td>28.71</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.429315</td>
      <td>1</td>
      <td>0.995767</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TPVITGAPYEYR</td>
      <td>RT-pep f</td>
      <td>33.38</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.466699</td>
      <td>1</td>
      <td>0.995354</td>
    </tr>
    <tr>
      <th>6</th>
      <td>DGLDAASYYAPVR</td>
      <td>RT-pep g</td>
      <td>42.26</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.537784</td>
      <td>0</td>
      <td>0.003368</td>
    </tr>
    <tr>
      <th>7</th>
      <td>ADVTPADFSEWSK</td>
      <td>RT-pep h</td>
      <td>54.62</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0.636728</td>
      <td>0</td>
      <td>0.003934</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GTFIIDPGGVIR</td>
      <td>RT-pep i</td>
      <td>70.52</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.764009</td>
      <td>0</td>
      <td>0.002995</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GTFIIDPAAVIR</td>
      <td>RT-pep k</td>
      <td>87.23</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>0.897775</td>
      <td>0</td>
      <td>0.002972</td>
    </tr>
    <tr>
      <th>10</th>
      <td>LFLQFGAQGSPFLK</td>
      <td>RT-pep l</td>
      <td>100.00</td>
      <td></td>
      <td></td>
      <td>14</td>
      <td>1.000000</td>
      <td>0</td>
      <td>0.002998</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>